<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="I was reading Feature Selection and Dimension Reduction for Single Cell RNA-Seq based on a Multinomial Model. In the paper, the authors model the scRNAseq counts using a multinomial distribution.
I was using negative binomial distribution for modeling in my last post, so I asked the question on twitter:
for modeling RNAseq counts, what&#39;s the difference/advantages using negative binomial and multinomial distribution? — Ming Tang (@tangming2005) November 26, 2019   some quotes from the answers I get from Matthew">

  
  <link rel="alternate" hreflang="en-us" href="/post/modeling-single-cell-rnaseq-data-with-multinomial-distribution/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/modeling-single-cell-rnaseq-data-with-multinomial-distribution/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/modeling-single-cell-rnaseq-data-with-multinomial-distribution/">
  <meta property="og:title" content="Modeling single cell RNAseq data with multinomial distribution  | DNA confesses Data speak">
  <meta property="og:description" content="I was reading Feature Selection and Dimension Reduction for Single Cell RNA-Seq based on a Multinomial Model. In the paper, the authors model the scRNAseq counts using a multinomial distribution.
I was using negative binomial distribution for modeling in my last post, so I asked the question on twitter:
for modeling RNAseq counts, what&#39;s the difference/advantages using negative binomial and multinomial distribution? — Ming Tang (@tangming2005) November 26, 2019   some quotes from the answers I get from Matthew">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2019-11-26T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2019-11-26T00:00:00&#43;00:00">
  

  
  

  <title>Modeling single cell RNAseq data with multinomial distribution  | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">Modeling single cell RNAseq data with multinomial distribution </h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2019-11-26 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Nov 26, 2019
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    4 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/modeling-single-cell-rnaseq-data-with-multinomial-distribution/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/scrnaseq/">scRNAseq</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Modeling%20single%20cell%20RNAseq%20data%20with%20multinomial%20distribution%20&amp;url=%2fpost%2fmodeling-single-cell-rnaseq-data-with-multinomial-distribution%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fmodeling-single-cell-rnaseq-data-with-multinomial-distribution%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fmodeling-single-cell-rnaseq-data-with-multinomial-distribution%2f&amp;title=Modeling%20single%20cell%20RNAseq%20data%20with%20multinomial%20distribution%20"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fmodeling-single-cell-rnaseq-data-with-multinomial-distribution%2f&amp;title=Modeling%20single%20cell%20RNAseq%20data%20with%20multinomial%20distribution%20"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Modeling%20single%20cell%20RNAseq%20data%20with%20multinomial%20distribution%20&amp;body=%2fpost%2fmodeling-single-cell-rnaseq-data-with-multinomial-distribution%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <p>I was reading <a href="https://www.biorxiv.org/content/10.1101/574574v1">Feature Selection and Dimension Reduction for Single Cell RNA-Seq based on a Multinomial Model</a>. In the paper, the authors model the scRNAseq counts using a multinomial distribution.</p>
<p><img src="/img/multinomial.png" /></p>
<p>I was using negative binomial distribution for modeling in my last <a href="https://divingintogeneticsandgenomics.rbind.io/post/negative-bionomial-distribution-in-single-cell-rnaseq/">post</a>, so I asked the question on twitter:</p>
<blockquote class="twitter-tweet">
<p lang="en" dir="ltr">
for modeling RNAseq counts, what's the difference/advantages using negative binomial and multinomial distribution?
</p>
— Ming Tang (<span class="citation">@tangming2005</span>) <a href="https://twitter.com/tangming2005/status/1199340525188349952?ref_src=twsrc%5Etfw">November 26, 2019</a>
</blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
<p>some quotes from the answers I get from Matthew</p>
<blockquote>
<p>the true distribution is multinomial. The conditional distr has of each gene is Poisson. Since there are so many genes each gene is approximately independent so independent Poissons can be used.</p>
</blockquote>
<blockquote>
<p>the marginal distribution of the true multinomial is binomial, which can be approximated by Poisson.</p>
</blockquote>
<blockquote>
<p>Real data is over dispersed since the Poisson only models technical variance not biological. To model the biological variance we use a mixture of poisons with a gamma prior — the gamma prior accounting for biological variability. The marginal distr of counts is negative binomial</p>
</blockquote>
<p>I am going to use multinomial distribution for the same data I used in my last post.</p>
<pre class="r"><code>library(Seurat)
library(tidyverse)
# There is an error when using this function, need to use the dev branch.
# https://github.com/satijalab/seurat/issues/2060
svensson_data&lt;- ReadH5AD(&quot;~/Downloads/svensson_chromium_control.h5ad&quot;)

raw_counts&lt;- svensson_data@assays$RNA@counts

# there are two datasets, each with 2000 cells
colnames(raw_counts) %&gt;% stringr::str_extract(&quot;[0-9]+_&quot;) %&gt;% table()</code></pre>
<pre><code>## .
## 20311_ 20312_ 
##   2000   2000</code></pre>
<pre class="r"><code># I am going to use only the second dataset sevensson et al 2
raw_counts2&lt;- raw_counts[, grepl(pattern = &quot;20312_&quot;, x = colnames(raw_counts))]</code></pre>
<p>counts from a gene fit a binomial distribution in a cell.</p>
<p>Given
<span class="math inline">\(\displaystyle \Pr(X=k)={\binom {n}{k}}p^{k}(1-p)^{n-k}\)</span> for binomial distribution,</p>
<p>the marginal mean for each gene is <span class="math inline">\(E[y_{ij}]=n_ip_{ij} = \mu_{ij}\)</span></p>
<p>the marginal variance is <span class="math inline">\(Var[y_{ij}] = n_ip_{ij}(1-p_{ij}) = \mu_{ij}- \frac1{n_i}\mu_{ij}^2\)</span></p>
<p>the probability of seeing a 0 count for a gene is: <span class="math inline">\((1-p_{ij})^{n_i} = (1-\frac{\mu_{ij}}{n_i})^{n_i}\)</span></p>
<pre class="r"><code># https://github.com/const-ae/sparseMatrixStats
library(sparseMatrixStats)
library(tidyverse)
gene_means&lt;- sparseMatrixStats::rowMeans2(raw_counts2)


## total counts for each cell
lib_size&lt;- sparseMatrixStats::colSums2(raw_counts2)

## https://github.com/willtownes/scrna2019/blob/master/real/zheng_2017_monocytes/02_exploratory.Rmd#L290
## why median though?
n_i&lt;- median(lib_size)

# probability of 0 for each gene given binomial distribution 
zeros_bn&lt;- (1- gene_means/n_i)^n_i 


## this is copied from last post, probability of 0 given negative binomiral distribution
phi &lt;- 1/0.3725
zeros_nb&lt;- (phi/(gene_means + phi))^phi

zeros_observed&lt;- apply(raw_counts2, 1, function(x) mean(x ==0))

data.frame(zeros_bn = zeros_bn, zeros_nb = zeros_nb, zeros_observed = zeros_observed, 
           gene_means = gene_means) %&gt;%
  ggplot(aes(x =log10(gene_means), y = zeros_observed)) +
  geom_point() +
  geom_line(aes(x = log10(gene_means), y = zeros_nb), color = &quot;red&quot;) +
  geom_line(aes(x = log10(gene_means), y = zeros_bn), color = &quot;blue&quot;) +
  theme_classic(base_size = 14) +
  ggtitle(&quot;Svensson et al 2&quot;)</code></pre>
<p><img src="/post/2019-11-26-modeling-single-cell-rnaseq-data-with-multinomial-distribution_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>At least for this dataset, negative bionomial (red line) seems to fit the observed 0 count better. multinomial with marginal binomial (blue line) seems to support 0 inflated in single-cell RNAseq.</p>
<p><strong>UPDATE</strong> 12/10/2019, as Will pointed out in the comment, we need to downsample the single cell data making each cell has roughly the same number of reads. He replicated my analysis at <a href="https://github.com/willtownes/scrna2019/blob/master/real/svensson_2019/01_exploratory.Rmd" class="uri">https://github.com/willtownes/scrna2019/blob/master/real/svensson_2019/01_exploratory.Rmd</a></p>
<blockquote class="twitter-tweet">
<p lang="en" dir="ltr">
Hi thanks for your interest, in order to make those plots, you have to be able to treat all the cells/droplets as being drawn from same multinomial distribution meaning all the “n_i” terms have to be the same (or at least close). We used downsampling to achieve that…
</p>
— Will (<span class="citation">@sandakano</span>) <a href="https://twitter.com/sandakano/status/1199709577144623106?ref_src=twsrc%5Etfw">November 27, 2019</a>
</blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
<p>I will put them in the same blog post for completeness.</p>
<pre class="r"><code>## from https://github.com/willtownes/scrna2019/blob/master/util/functions.R#L67

Down_Sample_Matrix&lt;-function(expr_mat,min_lib_size=NULL){
  #adapted from https://hemberg-lab.github.io/scRNA.seq.course/cleaning-the-expression-matrix.html#normalisations
  min_sz&lt;-min(colSums(expr_mat))
  if(is.null(min_lib_size)){
    min_lib_size&lt;-min_sz
  } else {
    stopifnot(min_lib_size&lt;=min_sz)
  }
  down_sample&lt;-function(x){
    prob &lt;- min_lib_size/sum(x)
    unlist(lapply(x,function(y){rbinom(1, y, prob)}))
  }
  apply(expr_mat, 2, down_sample)
}


gg&lt;-sparseMatrixStats::rowSums2(raw_counts2)&gt;0 #exclude genes that are zero across all cells
Y&lt;-raw_counts2[gg,]</code></pre>
<p>To make the droplets comparable, we will exclude droplets with total count below 2,000 and downsample all other droplets to have approximately the same total counts.</p>
<pre class="r"><code>total_counts&lt;- sparseMatrixStats::colSums2(Y)

hist(total_counts,breaks=100)</code></pre>
<p><img src="/post/2019-11-26-modeling-single-cell-rnaseq-data-with-multinomial-distribution_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<pre class="r"><code>Yss&lt;-Y[,total_counts&gt;2000]
#downsample to normalize droplet size (total UMI)
Yds&lt;-Down_Sample_Matrix(as.matrix(Yss))</code></pre>
<div id="using-the-downsampled-matrix-yss-for-plotting" class="section level3">
<h3>using the downsampled matrix Yss for plotting</h3>
<pre class="r"><code>Yds&lt;-Yds[rowSums(Yds)&gt;0,]

gene_means&lt;- rowMeans(Yds)
gene_vars&lt;- apply(Yds, 1, var)</code></pre>
<p>after downsampling, the mean and variance now are the same suggesting possion distribution</p>
<pre class="r"><code>df&lt;- bind_cols(gene_means = gene_means, gene_vars = gene_vars)
 
df %&gt;% ggplot(aes(x = log10(gene_means), y = log10(gene_vars))) +
        geom_point() +
        geom_abline(intercept = 0, slope = 1, color = &quot;red&quot;) + 
        theme_classic(base_size = 14) +
        ggtitle(&quot;svensson et al 2 downsample&quot;)</code></pre>
<p><img src="/post/2019-11-26-modeling-single-cell-rnaseq-data-with-multinomial-distribution_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>This is consistent with <a href="https://www.nxn.se/valent/2018/1/30/count-depth-variation-makes-poisson-scrna-seq-data-negative-binomial">Count depth variation makes Poisson scRNA-seq data Negative Binomial</a></p>
<p>Let’s see how the observed 0 counts fit each model:</p>
<pre class="r"><code>## total counts for each cell
lib_size&lt;- colSums(Yds)

## https://github.com/willtownes/scrna2019/blob/master/real/zheng_2017_monocytes/02_exploratory.Rmd#L290

N&lt;-median(colSums(Yds))
predict_zeros_binom&lt;-function(x){(1-exp(x)/N)^N} #binomial
predict_zeros_poi&lt;-function(x){exp(-exp(x))}
predict_zeros_nb&lt;-function(x,phi=100){
  exp(-phi*log1p(exp(x-log(phi))))
}

## note it is natural log here.
data.frame(zeros_observed = rowMeans(Yds==0), 
           x = log(gene_means)) %&gt;%
  ggplot(aes(x , y = zeros_observed), alpha = 0.5) +
  geom_point() +
  stat_function(aes(x,color=&quot;bin&quot;),fun=predict_zeros_binom) +
  stat_function(aes(x,color=&quot;poi&quot;),fun=predict_zeros_poi) +
  stat_function(aes(x,color=&quot;nb&quot;),fun=predict_zeros_nb) +
  scale_color_manual(&quot;model&quot;,breaks=c(&quot;bin&quot;,&quot;poi&quot;,&quot;nb&quot;),values=c(&quot;blue&quot;,&quot;green&quot;,&quot;red&quot;)) +
  theme_classic(base_size = 14) +
  ggtitle(&quot;Svensson et al 2&quot;)</code></pre>
<p><img src="/post/2019-11-26-modeling-single-cell-rnaseq-data-with-multinomial-distribution_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p><strong>After downsampling</strong>, “Poisson, binomial and negative binomial models all fit the data about the same.”</p>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/r/">R</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/statistics/">statistics</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/negative-bionomial-distribution-in-single-cell-rnaseq/">negative bionomial distribution in (single-cell) RNAseq </a></li>
        
        <li><a href="/post/clustering-scatacseq-data-the-tf-idf-way/">clustering scATACseq data: the TF-IDF way</a></li>
        
        <li><a href="/post/plot-10x-scatac-coverage-by-cluster-group/">plot 10x scATAC coverage by cluster/group</a></li>
        
        <li><a href="/post/use-docopt-to-write-command-line-r-utilities/">Use docopt to write command line R utilities </a></li>
        
        <li><a href="/project/evaluating-scrnaseq-cluster/">Evaluating single cell RNAseq cluster stability</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/mixing-mouse-and-human-10x-single-cell-rnaseq-data/" rel="next">Mixing mouse and human 10x single cell RNAseq data</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/negative-bionomial-distribution-in-single-cell-rnaseq/" rel="prev">negative bionomial distribution in (single-cell) RNAseq </a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

