+++
title = "An Integrated Platform for Genome-wide Mapping of Chromatin States Using High-throughput ChIP-sequencing in Tumor Tissues"
date = 2018-04-05T00:00:00

# Authors. Comma separated list, e.g. `["Bob Smith", "David Jones"]`.
authors = ["Christopher Terranova", "Ming Tang et.al"]

# Publication type.
# Legend:
# 0 = Uncategorized
# 1 = Conference proceedings
# 2 = Journal
# 3 = Work in progress
# 4 = Technical report
# 5 = Book
# 6 = Book chapter
publication_types = ["2"]

# Publication name and optional abbreviated version.
publication = "In *Journal of Visualized Experiments*, JOVE."
publication_short = "In *JOVE*"

# Abstract and optional shortened version.
abstract = "Histone modifications constitute a major component of the epigenome and play important regulatory roles in determining the transcriptional status of associated loci. In addition, the presence of specific modifications has been used to determine the position and identity non-coding functional elements such as enhancers. In recent years, chromatin immunoprecipitation followed by next generation sequencing (ChIP-seq) has become a powerful tool in determining the genome-wide profiles of individual histone modifications. However, it has become increasingly clear that the combinatorial patterns of chromatin modifications, referred to as Chromatin States, determine the identity and nature of the associated genomic locus. Therefore, workflows consisting of robust high-throughput (HT) methodologies for profiling a number of histone modification marks, as well as computational analyses pipelines capable of handling myriads of ChIP-Seq profiling datasets, are needed for comprehensive determination of epigenomic states in large number of samples. The HT-ChIP-Seq workflow presented here consists of two modules: 1) an experimental protocol for profiling several histone modifications from small amounts of tumor samples and cell lines in a 96-well format; and 2) a computational data analysis pipeline that combines existing tools to compute both individual mark occupancy and combinatorial chromatin state patterns. Together, these two modules facilitate easy processing of hundreds of ChIP-Seq samples in a fast and efficient manner. The workflow presented here is used to derive chromatin state patterns from 6 histone mark profiles in melanoma tumors and cell lines. Overall, we present a comprehensive ChIP-seq workflow that can be applied to dozens of human tumor samples and cancer cell lines to determine epigenomic aberrations in various malignancies."

# Featured image thumbnail (optional)
image_preview = ""

# Is this a selected publication? (true/false)
selected = false

# Projects (optional).
#   Associate this publication with one or more of your projects.
#   Simply enter the filename of your project in `content/project/`.
#   Otherwise, set `projects = []`.
# projects = ["example-external-project.md"]

# Links (optional).
url_pdf = "files/papers/Uveal.pdf"
# url_preprint = "#"
# url_code = "#"
# url_dataset = "#"
# url_project = "#"
# url_slides = "#"
# url_video = "#"
# url_poster = "#"
# url_source = "#"

# Custom links (optional).
#   Uncomment line below to enable. For multiple links, use the form `[{...}, {...}, {...}]`.
#url_custom = [{name = "Custom Link", url = "http://example.org"}]

# Does the content use math formatting?
math = true

# Does the content use source code highlighting?
highlight = true

# Featured image
# Place your image in the `static/img/` folder and reference its filename below, e.g. `image = "example.jpg"`.
[header]
image = "headers/chromHMM.png"
caption = "My caption :smile:"

+++

More detail can easily be written here using *Markdown* and $\rm \LaTeX$ math code.
