+++
title = "Immunogenomic intertumor heterogeneity across primary and metastatic sites in a patient with lung adenocarcinoma"
date = 2022-05-11T00:00:00

# Authors. Comma separated list, e.g. `["Bob Smith", "David Jones"]`.
authors = ["Runzhe chen ", "Jun Li", "...", "**Ming Tang** et.al."]

# Publication type.
# Legend:
# 0 = Uncategorized
# 1 = Conference proceedings
# 2 = Journal
# 3 = Work in progress
# 4 = Technical report
# 5 = Book
# 6 = Book chapter
publication_types = ["2"]

# Publication name and optional abbreviated version.
publication = "In *Journal of Experimental & Clinical Cancer Research*, Journal of Experimental & Clinical Cancer Research."
publication_short = "In *Journal of Experimental & Clinical Cancer Research*"
# Abstract and optional shortened version.
abstract = "**Background** Lung cancer is the leading cause of cancer death, partially owing to its extensive heterogeneity. The analysis of intertumor heterogeneity has been limited by an inability to concurrently obtain tissue from synchronous metastases unaltered by multiple prior lines of therapy. **Methods** In order to study the relationship between genomic, epigenomic and T cell repertoire heterogeneity in a rare autopsy case from a 32-year-old female never-smoker with left lung primary late-stage lung adenocarcinoma (LUAD), we did whole-exome sequencing (WES), DNA methylation and T cell receptor (TCR) sequencing to characterize the immunogenomic landscape of one primary and 19 synchronous metastatic tumors. **Results** We observed heterogeneous mutation, methylation, and T cell patterns across distinct metastases. Only TP53 mutation was detected in all tumors suggesting an early event while other cancer gene mutations were later events which may have followed subclonal diversification. A set of prevalent T cell clonotypes were completely excluded from left-side thoracic tumors indicating distinct T cell repertoire profiles between left-side and non left-side thoracic tumors. Though a limited number of predicted neoantigens were shared, these were associated with homology of the T cell repertoire across metastases. Lastly, ratio of methylated neoantigen coding mutations was negatively associated with T-cell density, richness and clonality, suggesting neoantigen methylation may partially drive immunosuppression. **Conclusions** Our study demonstrates heterogeneous genomic and T cell profiles across synchronous metastases and how restriction of unique T cell clonotypes within an individual may differentially shape the genomic and epigenomic landscapes of synchronous lung metastases."

# Featured image thumbnail (optional)
image_preview = ""

# Is this a selected publication? (true/false)
selected = false

# Projects (optional).
#   Associate this publication with one or more of your projects.
#   Simply enter the filename of your project in `content/project/`.
#   Otherwise, set `projects = []`.
# projects = ["example-external-project.md"]

# Links (optional).
#url_pdf = "files/papers/Melanoma_chromatin_state.pdf"
# url_preprint = "#"
#url_code = "https://github.com/crazyhottommy/pyflow-ChIPseq"
# url_dataset = "#"
# url_project = "#"
# url_slides = "#"
# url_video = "#"
# url_poster = "#"
# url_source = "#"

# Custom links (optional).
#   Uncomment line below to enable. For multiple links, use the form `[{...}, {...}, {...}]`.
url_custom = [{name = "Paper Link", url = "https://jeccr.biomedcentral.com/articles/10.1186/s13046-022-02361-x"}]

# Does the content use math formatting?
math = true

# Does the content use source code highlighting?
highlight = true

# Featured image
# Place your image in the `static/img/` folder and reference its filename below, e.g. `image = "example.jpg"`.
[header]
image = "headers/extensive_mets.png"
caption = "My caption :smile:"

+++

More detail can easily be written here using *Markdown* and $\rm \LaTeX$ math code.
