+++
title = "Monkeybread: A Python toolkit for the analysis of cellular niches in single-cell resolution spatial transcriptomics data"
date = 2023-09-15T00:00:00

# Authors. Comma separated list, e.g. `["Bob Smith", "David Jones"]`.
authors = ["Matthew N. Bernstein", "Dillon Scott", "Cynthia C. Hession", "...",  "Shruti Malu", "**Ming Tang#**"]

# Publication type.
# Legend:
# 0 = Uncategorized
# 1 = Conference proceedings
# 2 = Journal
# 3 = Work in progress
# 4 = Technical report
# 5 = Book
# 6 = Book chapter
publication_types = ["2"]

# Publication name and optional abbreviated version.
publication = "In *BioRxiv*,BioRxiv."
publication_short = "In *BioRxiv*"
# Abstract and optional shortened version.
abstract = "Spatial transcriptomics technologies enable the spatially resolved measurement of gene expression within a tissue specimen. With these technologies, researchers can investigate how cells organize into cellular niches which are defined as distinct regions in the tissue comprising a specific composition of cell types or phenotypes. While general-purpose software tools for the exploratory analysis of spatial transcriptomics data exist, there is a need for tools that specialize in the analysis of cellular organization into niches. This can further enhance the downstream application of these data towards drug target discovery, target validation, and biomarker development. We present Monkeybread: A Python toolkit for analyzing cellular organization and intercellular communication in single-cell resolution spatial transcriptomics data. We applied Monkeybread to a human melanoma sample to demonstrate its utility in identifying cellular niches with diverse immunogenic compositions in the tumor microenvironment. We found that these niches were differentially enriched for immunogenic and tolerogenic macrophage populations that could be correlated to T cell abundance. These findings highlight how Monkeybread can be used for revealing underlying biology of the tumor microenvironment, and in the future, for understanding the influence of these niches on response to available treatments and discovery of novel drug targets."

# Featured image thumbnail (optional)
image_preview = ""

# Is this a selected publication? (true/false)
selected = false

# Projects (optional).
#   Associate this publication with one or more of your projects.
#   Simply enter the filename of your project in `content/project/`.
#   Otherwise, set `projects = []`.
# projects = ["example-external-project.md"]

# Links (optional).
#url_pdf = "files/papers/Melanoma_chromatin_state.pdf"
url_preprint = "https://www.biorxiv.org/content/10.1101/2023.09.14.557736v1"
#url_code = "https://github.com/crazyhottommy/pyflow-ChIPseq"
# url_dataset = "#"
# url_project = "#"
# url_slides = "#"
# url_video = "#"
# url_poster = "#"
# url_source = "#"

# Custom links (optional).
#   Uncomment line below to enable. For multiple links, use the form `[{...}, {...}, {...}]`.
#url_custom = [{name = "Paper Link", url = "https://aacrjournals.org/cancerimmunolres/article-abstract/10/7/788/705256/Comprehensive-Characterizations-of-Immune-Receptor"}]

# Does the content use math formatting?
math = true

# Does the content use source code highlighting?
highlight = true

# Featured image
# Place your image in the `static/img/` folder and reference its filename below, e.g. `image = "example.jpg"`.
[header]
image = "headers/monkeybread.png"
caption = "My caption :smile:"

+++

More detail can easily be written here using *Markdown* and $\rm \LaTeX$ math code.
