<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="When it comes to make a heatmap, ComplexHeatmap by Zuguang Gu is my favorite. Check it out! You will be amazed on how flexible it is and the documentation is in top niche.
For Single-cell RNAseq, Seurat provides a DoHeatmap function using ggplot2. There are two limitations:
 when your genes are not in the top variable gene list, the scale.data will not have that gene and DoHeatmap will drop those genes.">

  
  <link rel="alternate" hreflang="en-us" href="/post/enhancement-of-scrnaseq-heatmap-using-complexheatmap/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/enhancement-of-scrnaseq-heatmap-using-complexheatmap/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/enhancement-of-scrnaseq-heatmap-using-complexheatmap/">
  <meta property="og:title" content="Enhancement of scRNAseq heatmap using complexheatmap | DNA confesses Data speak">
  <meta property="og:description" content="When it comes to make a heatmap, ComplexHeatmap by Zuguang Gu is my favorite. Check it out! You will be amazed on how flexible it is and the documentation is in top niche.
For Single-cell RNAseq, Seurat provides a DoHeatmap function using ggplot2. There are two limitations:
 when your genes are not in the top variable gene list, the scale.data will not have that gene and DoHeatmap will drop those genes.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2020-09-10T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2020-09-10T00:00:00&#43;00:00">
  

  
  

  <title>Enhancement of scRNAseq heatmap using complexheatmap | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">Enhancement of scRNAseq heatmap using complexheatmap</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2020-09-10 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Sep 10, 2020
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    5 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/enhancement-of-scrnaseq-heatmap-using-complexheatmap/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >, 
    
    <a href="/categories/r/">R</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Enhancement%20of%20scRNAseq%20heatmap%20using%20complexheatmap&amp;url=%2fpost%2fenhancement-of-scrnaseq-heatmap-using-complexheatmap%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fenhancement-of-scrnaseq-heatmap-using-complexheatmap%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fenhancement-of-scrnaseq-heatmap-using-complexheatmap%2f&amp;title=Enhancement%20of%20scRNAseq%20heatmap%20using%20complexheatmap"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fenhancement-of-scrnaseq-heatmap-using-complexheatmap%2f&amp;title=Enhancement%20of%20scRNAseq%20heatmap%20using%20complexheatmap"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Enhancement%20of%20scRNAseq%20heatmap%20using%20complexheatmap&amp;body=%2fpost%2fenhancement-of-scrnaseq-heatmap-using-complexheatmap%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      

<p>When it comes to make a heatmap, <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/" target="_blank"><code>ComplexHeatmap</code></a> by Zuguang Gu is my favorite. Check it out! You will be amazed on how flexible it is and the documentation is in top niche.</p>

<p>For Single-cell RNAseq, <code>Seurat</code> provides a <code>DoHeatmap</code> function using <code>ggplot2</code>. There are two limitations:</p>

<ul>
<li><p>when your genes are not in the top variable gene list, the <code>scale.data</code> will not have that gene and <code>DoHeatmap</code> will drop those genes.</p></li>

<li><p><code>ggplot2</code> does not support clustering of the rows or columns.</p></li>
</ul>

<p>I highly recommend you to read two posts I wrote as well on heatmap:</p>

<ul>
<li><p><a href="https://rpubs.com/crazyhottommy/a-tale-of-two-heatmap-functions" target="_blank">A tale of two heatmap functions</a></p></li>

<li><p><a href="https://rpubs.com/crazyhottommy/heatmap_demystified" target="_blank">Heatmap demystified</a></p></li>
</ul>

<p>Let me walk you through how I replicate and enhance the Seurat version of heatmap using <code>ComplexHeatmap</code>.</p>

<p>follow the tutorial <a href="https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html" target="_blank">https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html</a></p>

<pre><code class="language-r">library(dplyr)
library(Seurat)
library(ComplexHeatmap)
# https://github.com/immunogenomics/presto
library(presto)
library(tictoc)
</code></pre>

<pre><code class="language-r"># Load the PBMC dataset
pbmc.data &lt;- Read10X(data.dir = &quot;~/Downloads/filtered_gene_bc_matrices/hg19/&quot;)
pbmc &lt;- CreateSeuratObject(counts = pbmc.data, project = &quot;pbmc3k&quot;, min.cells = 3, min.features = 200)
</code></pre>

<pre><code>## Warning: Feature names cannot have underscores ('_'), replacing with dashes
## ('-')
</code></pre>

<pre><code class="language-r">pbmc
</code></pre>

<pre><code>## An object of class Seurat 
## 13714 features across 2700 samples within 1 assay 
## Active assay: RNA (13714 features)
</code></pre>

<pre><code class="language-r">pbmc[[&quot;percent.mt&quot;]] &lt;- PercentageFeatureSet(pbmc, pattern = &quot;^MT-&quot;)

## ScaleData uses top variable genes only
pbmc&lt;- pbmc %&gt;% 
  NormalizeData(normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000) %&gt;%
  FindVariableFeatures(selection.method = &quot;vst&quot;, nfeatures = 2000) %&gt;%
  ScaleData() %&gt;%
  RunPCA() %&gt;%
  FindNeighbors(dims = 1:10) %&gt;%
  FindClusters(resolution = 0.5) %&gt;%
  RunUMAP(dims = 1:10)
</code></pre>

<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 2700
## Number of edges: 97958
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8717
## Number of communities: 9
## Elapsed time: 0 seconds
</code></pre>

<pre><code>## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
## This message will be shown once per session
</code></pre>

<pre><code class="language-r">DimPlot(pbmc, reduction = &quot;umap&quot;)  
</code></pre>

<p><img src="/post/2020-09-10-enhancement-of-scrnaseq-heatmap-using-complexheatmap.en_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>

<p>The UMAP plot looks a bit different from the tutorial, but the structure is similar enough (You see how difficult it is to reproduce the exactly the same figure even with the same code:)).</p>

<p>Let&rsquo;s find marker genes for each cluster. I like <code>presto</code> for this purpose. It is much faster than Seurat.</p>

<pre><code class="language-r">tic()
markers&lt;- presto::wilcoxauc(pbmc, 'seurat_clusters', assay = 'data')
toc()
</code></pre>

<pre><code>## 0.419 sec elapsed
</code></pre>

<pre><code class="language-r">markers&lt;- top_markers(markers, n = 10, auc_min = 0.5, pct_in_min = 20, pct_out_max = 20)

markers
</code></pre>

<pre><code>## # A tibble: 10 x 10
##     rank `0`      `1`    `2`    `3`      `4`   `5`         `6`   `7`     `8`    
##    &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  
##  1     1 CCR7     S100A8 AQP3   CD79A    GZMA  FCGR3A      GZMB  FCER1A  PPBP   
##  2     2 PIK3IP1  FCN1   TRAT1  CD79B    CST7  IFITM3      PRF1  CLEC10A NRGN   
##  3     3 PRKCQ-A… LGALS2 SPOCK2 MS4A1    GZMK  RP11-290F2… GNLY  HLA-DQ… PF4    
##  4     4 LEF1     CFD    CD27   HLA-DQA1 LYAR  CFD         FGFB… CPVL    SDPR   
##  5     5 TCF7     GRN    TRADD  HLA-DQB1 GZMM  MS4A7       CST7  HLA-DMB GNG11  
##  6     6 CD27     MS4A6A CD3G   TCL1A    CD8A  CD68        GZMA  CD33    SPARC  
##  7     7 MAL      AP1S2  RGCC   LINC009… KLRG1 SPI1        FCGR… CTSH    RGS18  
##  8     8 RGCC     CD14   CD40LG HLA-DMA  PRF1  RHOC        SPON2 RNASE6  HIST1H…
##  9     9 CD3G     CD68   LAT    VPREB3   GZMH  HCK         CCL4  KLF4    TPM4   
## 10    10 LDLRAP1  LGALS3 FLT3LG HLA-DQA2 HOPX  IFI30       APMAP RNF130  GP9
</code></pre>

<h3 id="doheatmap">DoHeatmap</h3>

<pre><code class="language-r">all_markers&lt;- markers %&gt;%
  select(-rank) %&gt;% 
  unclass() %&gt;% 
  stack() %&gt;%
  pull(values) %&gt;%
  unique() %&gt;%
   .[!is.na(.)]
DoHeatmap(pbmc, features = all_markers) + NoLegend()
</code></pre>

<pre><code>## Warning in DoHeatmap(pbmc, features = all_markers): The following features were
## omitted as they were not found in the scale.data slot for the RNA assay: TPM4,
## RNF130, KLF4, RNASE6, CTSH, CD33, CPVL, APMAP, HCK, RHOC, RP11-290F20.3, CD8A,
## GZMM, VPREB3, FLT3LG, LAT, CD40LG, SPOCK2, TRAT1, CD68, CD14, AP1S2, GRN, CD3G,
## RGCC, CD27, TCF7, LEF1, PRKCQ-AS1, PIK3IP1, CCR7
</code></pre>

<p><img src="/post/2020-09-10-enhancement-of-scrnaseq-heatmap-using-complexheatmap.en_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>

<p>you see warnings:&ldquo;the following features were omitted as they were not found in the scale.data slot for the RNA assay&rdquo;. The color mapping looks different from the tutorial. It could be different Seurat version uses different parameters. I have not check the code base change.</p>

<h3 id="replicate-the-heatmap-using-complexheatmap">replicate the heatmap using Complexheatmap</h3>

<p>We can go back to the <code>data</code> slot to get all the genes</p>

<pre><code class="language-r">mat&lt;- pbmc[[&quot;RNA&quot;]]@data[all_markers, ] %&gt;% as.matrix()

## scale the rows
mat&lt;- t(scale(t(mat)))

cluster_anno&lt;- pbmc@meta.data$seurat_clusters
</code></pre>

<p>We can explicitly map the colors to the scaled expression values</p>

<pre><code class="language-r"># what's the value range in the matrix
quantile(mat, c(0.1, 0.95))
</code></pre>

<pre><code>##        10%        95% 
## -0.5277426  2.3470090
</code></pre>

<pre><code class="language-r">Seurat::PurpleAndYellow()
</code></pre>

<pre><code>##  [1] &quot;#FF00FF&quot; &quot;#F400F4&quot; &quot;#EA00EA&quot; &quot;#DF00DF&quot; &quot;#D500D5&quot; &quot;#CA00CA&quot; &quot;#BF00BF&quot;
##  [8] &quot;#B500B5&quot; &quot;#AA00AA&quot; &quot;#9F009F&quot; &quot;#950095&quot; &quot;#8A008A&quot; &quot;#800080&quot; &quot;#750075&quot;
## [15] &quot;#6A006A&quot; &quot;#600060&quot; &quot;#550055&quot; &quot;#4A004A&quot; &quot;#400040&quot; &quot;#350035&quot; &quot;#2B002B&quot;
## [22] &quot;#200020&quot; &quot;#150015&quot; &quot;#0B000B&quot; &quot;#000000&quot; &quot;#000000&quot; &quot;#0B0B00&quot; &quot;#151500&quot;
## [29] &quot;#202000&quot; &quot;#2B2B00&quot; &quot;#353500&quot; &quot;#404000&quot; &quot;#4A4A00&quot; &quot;#555500&quot; &quot;#606000&quot;
## [36] &quot;#6A6A00&quot; &quot;#757500&quot; &quot;#808000&quot; &quot;#8A8A00&quot; &quot;#959500&quot; &quot;#9F9F00&quot; &quot;#AAAA00&quot;
## [43] &quot;#B5B500&quot; &quot;#BFBF00&quot; &quot;#CACA00&quot; &quot;#D4D400&quot; &quot;#DFDF00&quot; &quot;#EAEA00&quot; &quot;#F4F400&quot;
## [50] &quot;#FFFF00&quot;
</code></pre>

<pre><code class="language-r">## make the black color map to 0. the yellow map to highest and the purle map to the lowest
col_fun = circlize::colorRamp2(c(-1, 0, 3), c(&quot;#FF00FF&quot;, &quot;black&quot;, &quot;#FFFF00&quot;))
</code></pre>

<p>plot the heatmap</p>

<pre><code class="language-r">Heatmap(mat, name = &quot;Expression&quot;,  
        column_split = factor(cluster_anno),
        cluster_columns = TRUE,
        show_column_dend = FALSE,
        cluster_column_slices = TRUE,
        column_title_gp = gpar(fontsize = 8),
        column_gap = unit(0.5, &quot;mm&quot;),
        cluster_rows = TRUE,
        show_row_dend = FALSE,
        col = col_fun,
        row_names_gp = gpar(fontsize = 4),
        column_title_rot = 90,
        top_annotation = HeatmapAnnotation(foo = anno_block(gp = gpar(fill = scales::hue_pal()(9)))),
        show_column_names = FALSE,
        use_raster = TRUE,
        raster_quality = 4)
</code></pre>

<p><img src="/post/2020-09-10-enhancement-of-scrnaseq-heatmap-using-complexheatmap.en_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>

<p>In addition to the capability to plot all the genes, one can cluster the rows (genes) and the columns (cells) within each slice (cell type), and slices can be further clustered as well.</p>

<p>Several other notes:</p>

<ul>
<li>When you have too many cells (&gt; 10,000), the <code>use_raster</code> option really helps. Also consider downsample the Seurat object to a smaller number of cells for plotting the heatmap. Your screen resolution is not as high as 300,000 pixels if you have 300,000 cells (columns).</li>
</ul>

<p>Please read <a href="https://jokergoo.github.io/2020/06/30/rasterization-in-complexheatmap/" target="_blank">https://jokergoo.github.io/2020/06/30/rasterization-in-complexheatmap/</a>
and <a href="https://gdevailly.netlify.app/post/plotting-big-matrices-in-r/" target="_blank">Plotting heatmaps from big matrices in R</a></p>

<ul>
<li><p>check <a href="https://github.com/stemangiola/tidyHeatmap" target="_blank">tidyHeatmap</a> built upon <code>Complexheatmap</code> for tidy dataframe.</p></li>

<li><p><a href="https://github.com/YuLab-SMU/aplot" target="_blank">aplot</a> from Guangchuang Yu.</p></li>

<li><p><a href="https://github.com/rlbarter/superheat" target="_blank">superheat</a> from Rebecca Barter.</p></li>
</ul>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/scrnaseq/">scRNAseq</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/dplyr-count-misses-factor-levels-a-case-in-comparing-scrnaseq-cell-type-abundance/">dplyr::count misses factor levels: a case in comparing scRNAseq cell type abundance </a></li>
        
        <li><a href="/post/stacked-violin-plot-for-visualizing-single-cell-data-in-seurat/">stacked violin plot for visualizing single-cell data in Seurat</a></li>
        
        <li><a href="/post/compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data/">compare kallisto-bustools and cellranger for single nuclei sequencing data</a></li>
        
        <li><a href="/post/cellranger-mk-reference-with-transgenes/">cellranger mk reference with transgenes</a></li>
        
        <li><a href="/post/add-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column/">add pct_in for each cluster for scRNAseq result table using list column</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/compare-slopes-in-linear-regression/" rel="next">compare slopes in linear regression</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/dplyr-count-misses-factor-levels-a-case-in-comparing-scrnaseq-cell-type-abundance/" rel="prev">dplyr::count misses factor levels: a case in comparing scRNAseq cell type abundance </a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

