<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="Dotplot is a nice way to visualize scRNAseq expression data across clusters. It gives information (by color) for the average expression level across cells within the cluster and the percentage (by size of the dot) of the cells express that gene within the cluster.
Seurat has a nice function for that. However, it can not do the clustering for the rows and columns. David McGaughey has written a blog post using ggplot2 and ggtree from Guangchuang Yu.">

  
  <link rel="alternate" hreflang="en-us" href="/post/clustered-dotplot-for-single-cell-rnaseq/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/clustered-dotplot-for-single-cell-rnaseq/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/clustered-dotplot-for-single-cell-rnaseq/">
  <meta property="og:title" content="clustered dotplot for single-cell RNAseq | DNA confesses Data speak">
  <meta property="og:description" content="Dotplot is a nice way to visualize scRNAseq expression data across clusters. It gives information (by color) for the average expression level across cells within the cluster and the percentage (by size of the dot) of the cells express that gene within the cluster.
Seurat has a nice function for that. However, it can not do the clustering for the rows and columns. David McGaughey has written a blog post using ggplot2 and ggtree from Guangchuang Yu.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2021-03-10T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2021-03-10T00:00:00&#43;00:00">
  

  
  

  <title>clustered dotplot for single-cell RNAseq | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">clustered dotplot for single-cell RNAseq</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2021-03-10 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Mar 10, 2021
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    9 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/clustered-dotplot-for-single-cell-rnaseq/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >, 
    
    <a href="/categories/visualization/">visualization</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=clustered%20dotplot%20for%20single-cell%20RNAseq&amp;url=%2fpost%2fclustered-dotplot-for-single-cell-rnaseq%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fclustered-dotplot-for-single-cell-rnaseq%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fclustered-dotplot-for-single-cell-rnaseq%2f&amp;title=clustered%20dotplot%20for%20single-cell%20RNAseq"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fclustered-dotplot-for-single-cell-rnaseq%2f&amp;title=clustered%20dotplot%20for%20single-cell%20RNAseq"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=clustered%20dotplot%20for%20single-cell%20RNAseq&amp;body=%2fpost%2fclustered-dotplot-for-single-cell-rnaseq%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <p>Dotplot is a nice way to visualize scRNAseq expression data across clusters. It gives information (by color) for the average expression level across cells within the cluster and the percentage (by size of the dot) of the cells express that gene within the cluster.</p>
<p><code>Seurat</code> has a nice function for that. However, it can not do the clustering for the rows and columns. David McGaughey has written a <a href="http://davemcg.github.io/post/lets-plot-scrna-dotplots/">blog post</a> using ggplot2 and ggtree from Guangchuang Yu. It looks great!</p>
<p><img src="/img/clustered_dotplot.png" /></p>
<p>In this post, I am trying to reproduce the same effect using <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/">ComplexHeatmap</a></p>
<pre class="r"><code>library(Seurat)
library(tidyverse)
library(presto)
library(ComplexHeatmap)
library(circlize)
# Load the PBMC dataset
pbmc.data &lt;- Read10X(data.dir = &quot;~/Downloads/filtered_gene_bc_matrices/hg19/&quot;)
pbmc &lt;- CreateSeuratObject(counts = pbmc.data, project = &quot;pbmc3k&quot;, min.cells = 3, min.features = 200)</code></pre>
<div id="standard-processing" class="section level3">
<h3>standard processing</h3>
<p>following <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="uri">https://satijalab.org/seurat/articles/pbmc3k_tutorial.html</a></p>
<pre class="r"><code>pbmc[[&quot;percent.mt&quot;]] &lt;- PercentageFeatureSet(pbmc, pattern = &quot;^MT-&quot;)

## ScaleData uses top variable genes only
pbmc&lt;- pbmc %&gt;% 
  NormalizeData(normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000) %&gt;%
  FindVariableFeatures(selection.method = &quot;vst&quot;, nfeatures = 2000) %&gt;%
  ScaleData() %&gt;%
  RunPCA() %&gt;%
  FindNeighbors(dims = 1:10) %&gt;%
  FindClusters(resolution = 0.5) %&gt;%
  RunUMAP(dims = 1:10)</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 2700
## Number of edges: 97958
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8717
## Number of communities: 9
## Elapsed time: 0 seconds</code></pre>
<pre class="r"><code>DimPlot(pbmc, reduction = &quot;umap&quot;)</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-3-1.png" width="1152" /></p>
<pre class="r"><code>new.cluster.ids &lt;- c(&quot;Naive CD4 T&quot;, &quot;CD14+ Mono&quot;, &quot;Memory CD4 T&quot;, &quot;B&quot;, &quot;CD8 T&quot;, &quot;FCGR3A+ Mono&quot;, 
    &quot;NK&quot;, &quot;DC&quot;, &quot;Platelet&quot;)
names(new.cluster.ids) &lt;- levels(pbmc)
pbmc &lt;- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, pt.size = 0.5) + NoLegend()</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-4-1.png" width="1152" /></p>
</div>
<div id="find-marker-genes" class="section level3">
<h3>find marker genes</h3>
<pre class="r"><code>markers&lt;- presto::wilcoxauc(pbmc, &#39;seurat_clusters&#39;, assay = &#39;data&#39;)
markers&lt;- top_markers(markers, n = 10, auc_min = 0.5, pct_in_min = 20, pct_out_max = 20)

markers</code></pre>
<pre><code>## # A tibble: 10 x 10
##     rank `0`      `1`    `2`    `3`      `4`   `5`         `6`   `7`     `8`    
##    &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  
##  1     1 CCR7     S100A8 AQP3   CD79A    GZMA  FCGR3A      GZMB  FCER1A  PPBP   
##  2     2 PIK3IP1  FCN1   TRAT1  CD79B    CST7  IFITM3      PRF1  CLEC10A NRGN   
##  3     3 PRKCQ-A… LGALS2 SPOCK2 MS4A1    GZMK  RP11-290F2… GNLY  HLA-DQ… PF4    
##  4     4 LEF1     CFD    CD27   HLA-DQA1 LYAR  CFD         FGFB… CPVL    SDPR   
##  5     5 TCF7     GRN    TRADD  HLA-DQB1 GZMM  MS4A7       CST7  HLA-DMB GNG11  
##  6     6 CD27     MS4A6A CD3G   TCL1A    CD8A  CD68        GZMA  CD33    SPARC  
##  7     7 MAL      AP1S2  RGCC   LINC009… KLRG1 SPI1        FCGR… CTSH    RGS18  
##  8     8 RGCC     CD14   CD40LG HLA-DMA  PRF1  RHOC        SPON2 RNASE6  HIST1H…
##  9     9 CD3G     CD68   LAT    VPREB3   GZMH  HCK         CCL4  KLF4    TPM4   
## 10    10 LDLRAP1  LGALS3 FLT3LG HLA-DQA2 HOPX  IFI30       APMAP RNF130  GP9</code></pre>
<pre class="r"><code>all_markers&lt;- markers %&gt;%
  select(-rank) %&gt;% 
  unclass() %&gt;% 
  stack() %&gt;%
  pull(values) %&gt;%
  unique() %&gt;%
   .[!is.na(.)]</code></pre>
<p>Seurat’s dot plot</p>
<pre class="r"><code>p&lt;- DotPlot(object = pbmc, features = all_markers)
p</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-6-1.png" width="1152" /></p>
</div>
<div id="lets-reproduce-it-using-complexheatmap" class="section level3">
<h3>Let’s reproduce it using Complexheatmap</h3>
<pre class="r"><code>df&lt;- p$data
head(df)</code></pre>
<pre><code>##            avg.exp  pct.exp features.plot          id avg.exp.scaled
## CCR7      2.993317 43.41534          CCR7 Naive CD4 T       2.337223
## PIK3IP1   2.674179 42.11288       PIK3IP1 Naive CD4 T       1.844182
## PRKCQ-AS1 2.167078 33.28509     PRKCQ-AS1 Naive CD4 T       2.123716
## LEF1      2.003597 31.83792          LEF1 Naive CD4 T       2.027023
## TCF7      2.253631 37.48191          TCF7 Naive CD4 T       1.773285
## CD27      2.326533 39.07381          CD27 Naive CD4 T       1.210923</code></pre>
<pre class="r"><code>### the matrix for the scaled expression 
exp_mat&lt;-df %&gt;% 
  select(-pct.exp, -avg.exp) %&gt;%  
  pivot_wider(names_from = id, values_from = avg.exp.scaled) %&gt;% 
  as.data.frame() 
    
row.names(exp_mat) &lt;- exp_mat$features.plot  
exp_mat &lt;- exp_mat[,-1] %&gt;% as.matrix()

head(exp_mat)</code></pre>
<pre><code>##           Naive CD4 T          B Memory CD4 T FCGR3A+ Mono         NK
## CCR7         2.337223  0.4065916    0.6473101   -0.5357480 -0.6116367
## PIK3IP1      1.844182 -0.2277953    1.4179422   -0.6531667 -0.1884663
## PRKCQ-AS1    2.123716 -0.6830185    1.0323842   -0.5907663  0.0349381
## LEF1         2.027023 -0.5746819    1.4350437   -0.5707971 -0.4876044
## TCF7         1.773285 -0.4479782    1.5832831   -0.5337908 -0.6645233
## CD27         1.210923 -0.2029260    1.3852365   -0.7374133 -0.6415774
##                CD8 T CD14+ Mono         DC   Platelet
## CCR7      -0.4272161 -0.5827381 -0.5508353 -0.6829502
## PIK3IP1    0.2223521 -0.8180393 -0.9991901 -0.5978184
## PRKCQ-AS1  0.2375921 -0.6732668 -0.7407892 -0.7407892
## LEF1      -0.3271094 -0.4091405 -0.7288104 -0.3639229
## TCF7       0.2798941 -0.5843243 -0.6207981 -0.7850477
## CD27       1.3328521 -0.7261509 -0.7919650 -0.8289795</code></pre>
<pre class="r"><code>## the matrix for the percentage of cells express a gene

percent_mat&lt;-df %&gt;% 
  select(-avg.exp, -avg.exp.scaled) %&gt;%  
  pivot_wider(names_from = id, values_from = pct.exp) %&gt;% 
  as.data.frame() 
    
row.names(percent_mat) &lt;- percent_mat$features.plot  
percent_mat &lt;- percent_mat[,-1] %&gt;% as.matrix()

head(percent_mat)</code></pre>
<pre><code>##           Naive CD4 T         B Memory CD4 T FCGR3A+ Mono        NK     CD8 T
## CCR7         43.41534 15.759312     26.37131     5.095541  2.702703  4.747774
## PIK3IP1      42.11288 13.180516     40.92827    12.738854 13.513514 18.694362
## PRKCQ-AS1    33.28509  1.146132     29.53586     4.458599 11.486486 13.353116
## LEF1         31.83792  2.578797     32.27848     3.821656  4.054054  5.341246
## TCF7         37.48191  9.169054     36.91983    17.197452  8.108108 18.991098
## CD27         39.07381 12.034384     45.14768     4.458599  4.729730 25.816024
##           CD14+ Mono        DC Platelet
## CCR7        2.434077  8.333333 0.000000
## PIK3IP1     6.693712  2.777778 6.666667
## PRKCQ-AS1   1.419878  0.000000 0.000000
## LEF1        5.070994  0.000000 6.666667
## TCF7        9.736308 16.666667 6.666667
## CD27        3.042596  2.777778 0.000000</code></pre>
<pre class="r"><code>## the range is from 0 - 100
range(percent_mat)</code></pre>
<pre><code>## [1]   0 100</code></pre>
<pre class="r"><code>## these two matrix have the same dimension
dim(exp_mat)</code></pre>
<pre><code>## [1] 80  9</code></pre>
<pre class="r"><code>dim(percent_mat)</code></pre>
<pre><code>## [1] 80  9</code></pre>
<pre class="r"><code>library(viridis)
library(Polychrome)

Polychrome::swatch(viridis(20))</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-8-1.png" width="1152" /></p>
<pre class="r"><code>## get an idea of the ranges of the matrix
quantile(exp_mat, c(0.1, 0.5, 0.9, 0.99))</code></pre>
<pre><code>##        10%        50%        90%        99% 
## -0.6695752 -0.3567907  1.7156356  2.5000000</code></pre>
<pre class="r"><code>## any value that is greater than 2 will be mapped to yellow
col_fun = circlize::colorRamp2(c(-1, 0, 2), viridis(20)[c(1,10, 20)])


cell_fun = function(j, i, x, y, w, h, fill){
          grid.rect(x = x, y = y, width = w, height = h, 
                    gp = gpar(col = NA, fill = NA))
          grid.circle(x=x,y=y,r= percent_mat[i, j]/100 * min(unit.c(w, h)),
                      gp = gpar(fill = col_fun(exp_mat[i, j]), col = NA))}</code></pre>
<pre class="r"><code>## also do a kmeans clustering for the genes with k = 4
Heatmap(exp_mat,
        heatmap_legend_param=list(title=&quot;expression&quot;),
        column_title = &quot;clustered dotplot&quot;, 
        col=col_fun,
        rect_gp = gpar(type = &quot;none&quot;),
        cell_fun = cell_fun,
        row_names_gp = gpar(fontsize = 5),
        row_km = 4,
        border = &quot;black&quot;)</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-9-1.png" width="768" /></p>
<p>It looks pretty decent. One can now take advantage of all the awesome annotation function to add annotation bars in complexheatmap.</p>
<pre class="r"><code>colnames(exp_mat)</code></pre>
<pre><code>## [1] &quot;Naive CD4 T&quot;  &quot;B&quot;            &quot;Memory CD4 T&quot; &quot;FCGR3A+ Mono&quot; &quot;NK&quot;          
## [6] &quot;CD8 T&quot;        &quot;CD14+ Mono&quot;   &quot;DC&quot;           &quot;Platelet&quot;</code></pre>
<pre class="r"><code>library(RColorBrewer)
cluster_anno&lt;-  c(&quot;CD4T&quot;, &quot;B&quot;, &quot;CD4T&quot;, &quot;Mono&quot;, &quot;NK&quot;, &quot;CD8T&quot;, &quot;CD14_Mono&quot;, &quot;DC&quot;, &quot;Platelet&quot;)

column_ha&lt;- HeatmapAnnotation(
    cluster_anno = cluster_anno,
    col = list(cluster_anno = setNames(brewer.pal(8, &quot;Paired&quot;), unique(cluster_anno))
    ),
    na_col = &quot;grey&quot;
)

Heatmap(exp_mat,
        heatmap_legend_param=list(title=&quot;expression&quot;),
        column_title = &quot;clustered dotplot&quot;, 
        col=col_fun,
        rect_gp = gpar(type = &quot;none&quot;),
        cell_fun = cell_fun,
        row_names_gp = gpar(fontsize = 5),
        row_km = 4,
        border = &quot;black&quot;,
        top_annotation = column_ha)</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-10-1.png" width="1152" /></p>
</div>
<div id="update" class="section level3">
<h3>UPDATE</h3>
<p>Zuguang mentioned: <a href="https://twitter.com/jokergoo_gu/status/1372941312622743558" class="uri">https://twitter.com/jokergoo_gu/status/1372941312622743558</a></p>
<blockquote>
<p>Alternatively you can use <code>layer_fun</code> which is a vectorized version of <code>cell_fun</code>. It will speed up plotting in interactive graphics devices. Just change <code>percent_mat[i, j]</code> to <code>pindex(percent_mat, i, j)</code> and keep other code unchanged, and assign to <code>layer_fun</code> argument.</p>
</blockquote>
<blockquote>
<p>One can also create a legend for the dot size for the percentages. <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/legends.html#add-customized-legends" class="uri">https://jokergoo.github.io/ComplexHeatmap-reference/book/legends.html#add-customized-legends</a></p>
</blockquote>
<blockquote>
<p>I think for points, it might be the diameter. <code>size</code> is directly passed to <code>grid.points()</code> so it should mean the size of an arbitrary symbol.</p>
</blockquote>
<p>Later, he confirmed that the size of points generated by grid.points(size = 10) is different from that from grid.circle(r=5)</p>
<blockquote>
<p>For your case, I would suggest to also use grid.points in your layer_fun.
If you use grid.circle in layer_fun, then you also need to use grid.circle to construct legend, which is a little bit complicated. You need to set <code>graphics</code> argument in Legend()</p>
</blockquote>
<blockquote>
<p>See the last example in Section 5.2. <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/legends.html#discrete-legends" class="uri">https://jokergoo.github.io/ComplexHeatmap-reference/book/legends.html#discrete-legends</a></p>
</blockquote>
</div>
<div id="use-grid.circle-in-both-the-heatmap-body-and-the-legend" class="section level3">
<h3>use grid.circle() in both the heatmap body and the legend</h3>
<pre class="r"><code>## To make the size of the dot in the heatmap body comparable to the legend, I used fixed
## size unit.(2, &quot;mm) rather than min(unit.c(w, h).

layer_fun = function(j, i, x, y, w, h, fill){
          grid.rect(x = x, y = y, width = w, height = h, 
                    gp = gpar(col = NA, fill = NA))
          grid.circle(x=x,y=y,r= pindex(percent_mat, i, j)/100 * unit(2, &quot;mm&quot;),
                      gp = gpar(fill = col_fun(pindex(exp_mat, i, j)), col = NA))}


lgd_list = list(
    Legend( labels = c(0,0.25,0.5,0.75,1), title = &quot;pt&quot;,
            graphics = list(
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0 * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0.25 * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0.5 * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0.75 * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 1 * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)))
            ))

set.seed(123)    
hp&lt;- Heatmap(exp_mat,
        heatmap_legend_param=list(title=&quot;expression&quot;),
        column_title = &quot;clustered dotplot&quot;, 
        col=col_fun,
        rect_gp = gpar(type = &quot;none&quot;),
        layer_fun = layer_fun,
        row_names_gp = gpar(fontsize = 5),
        row_km = 4,
        border = &quot;black&quot;,
        top_annotation = column_ha)

draw( hp, annotation_legend_list = lgd_list)</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-11-1.png" width="1152" /></p>
<p>Note that, we use the percentage as the radius of the circle. The area of the dot is <code>pi *r^2</code>.
It will appear 4 times different if the original percentage is 0.5 versus 1 (2 times difference) which may not be desirable.
see <a href="https://stackoverflow.com/questions/64067450/does-size-for-ggplot2geom-point-refer-to-radius-diameter-area-or-somethin" class="uri">https://stackoverflow.com/questions/64067450/does-size-for-ggplot2geom-point-refer-to-radius-diameter-area-or-somethin</a></p>
<p>We can do a square root of the radius:</p>
<pre class="r"><code>layer_fun1 = function(j, i, x, y, w, h, fill){
          grid.rect(x = x, y = y, width = w, height = h, 
                    gp = gpar(col = NA, fill = NA))
          grid.circle(x=x,y=y,r= sqrt(pindex(percent_mat, i, j)/100)  * unit(2, &quot;mm&quot;),
                      gp = gpar(fill = col_fun(pindex(exp_mat, i, j)), col = NA))}

lgd_list1 = list(
    Legend( labels = c(0,0.25,0.5,0.75,1), title = &quot;pt&quot;,
            graphics = list(
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0  * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = sqrt(0.25) * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = sqrt(0.5) * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = sqrt(0.75) * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 1 * unit(2, &quot;mm&quot;),
                                               gp = gpar(fill = &quot;black&quot;)))
            ))

set.seed(123)    
hp1&lt;- Heatmap(exp_mat,
        heatmap_legend_param=list(title=&quot;expression&quot;),
        column_title = &quot;clustered dotplot&quot;, 
        col=col_fun,
        rect_gp = gpar(type = &quot;none&quot;),
        layer_fun = layer_fun1,
        row_names_gp = gpar(fontsize = 5),
        row_km = 4,
        border = &quot;black&quot;,
        top_annotation = column_ha)

draw( hp1, annotation_legend_list = lgd_list1)</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-12-1.png" width="1152" /></p>
</div>
<div id="use-grid.points-in-both-the-heatmap-body-and-the-legend" class="section level3">
<h3>use grid.points() in both the heatmap body and the legend</h3>
<pre class="r"><code>## note for grid.points, use col for the filling color of the points, while in grid.circle, use fill for the filling color of the circle. I should learn more about {grid}

layer_fun2 = function(j, i, x, y, w, h, fill){
          grid.rect(x = x, y = y, width = w, height = h, 
                    gp = gpar(col = NA, fill = NA))
          grid.points(x = x, y = y, 
                     gp = gpar(col = col_fun(pindex(exp_mat, i, j))),
                     size = pindex(percent_mat, i, j)/100 * unit(4, &quot;mm&quot;),
                     pch = 16
                    )
}


lgd_list2 = list(
    Legend( labels = c(0,0.25,0.5,0.75,1), title = &quot;pt&quot;, type = &quot;points&quot;, pch = 16, size = c(0,0.25,0.5,0.75,1) * unit(4,&quot;mm&quot;),
        legend_gp = gpar(col = &quot;black&quot;)))

set.seed(123)
hp2&lt;- Heatmap(exp_mat,
        heatmap_legend_param=list(title=&quot;expression&quot;),
        column_title = &quot;clustered dotplot&quot;, 
        col=col_fun,
        rect_gp = gpar(type = &quot;none&quot;),
        layer_fun = layer_fun2,
        row_names_gp = gpar(fontsize = 5),
        row_km = 4,
        border = &quot;black&quot;,
        top_annotation = column_ha)

draw( hp2, annotation_legend_list = lgd_list2)</code></pre>
<p><img src="/post/2021-03-10-clustered-dotplot-for-single-cell-rnaseq.en_files/figure-html/unnamed-chunk-13-1.png" width="1152" /></p>
</div>
<div id="conclusion" class="section level3">
<h3>Conclusion</h3>
<p>I would use <code>grid.circle</code> to draw the dots in the legend and the heatmap body since I can fine-tune the looking of the dot size. I am not entirely sure how the size argument determines the size of the dots in <code>grid.points(size)</code>.</p>
<p>Again, thanks <a href="https://twitter.com/jokergoo_gu">Zuguang Gu</a> for this awesome package!!</p>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/scrnaseq/">scRNAseq</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/enhancement-of-scrnaseq-heatmap-using-complexheatmap/">Enhancement of scRNAseq heatmap using complexheatmap</a></li>
        
        <li><a href="/post/dplyr-count-misses-factor-levels-a-case-in-comparing-scrnaseq-cell-type-abundance/">dplyr::count misses factor levels: a case in comparing scRNAseq cell type abundance </a></li>
        
        <li><a href="/post/stacked-violin-plot-for-visualizing-single-cell-data-in-seurat/">stacked violin plot for visualizing single-cell data in Seurat</a></li>
        
        <li><a href="/post/compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data/">compare kallisto-bustools and cellranger for single nuclei sequencing data</a></li>
        
        <li><a href="/post/cellranger-mk-reference-with-transgenes/">cellranger mk reference with transgenes</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/how-to-test-if-two-distributions-are-different/" rel="next">How to test if two distributions are different </a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/precious-lessons-i-learned-from-my-mom/" rel="prev">Precious lessons I learned from my Mom</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

