<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="In this blog post, I am going to show you how to make a correlation heatmap with p-values and significant values labeled in the heatmap body. Let’s use the PBMC single cell data as an example.
You may want to read my previous blog post How to do gene correlation for single-cell RNAseq data.
Load libraries library(dplyr) library(Seurat) library(patchwork) library(ggplot2) library(ComplexHeatmap) library(SeuratData) library(hdWGCNA) library(WGCNA) set.seed(1234)  prepare the data data(&quot;pbmc3k&quot;) pbmc3k #&gt; An object of class Seurat #&gt; 13714 features across 2700 samples within 1 assay #&gt; Active assay: RNA (13714 features, 0 variable features) ## routine processing pbmc3k&lt;- pbmc3k %&gt;% NormalizeData(normalization.">

  
  <link rel="alternate" hreflang="en-us" href="/post/how-to-make-a-triangle-correlation-heatmap-with-p-values-labeled/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Chatomics">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Chatomics">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/how-to-make-a-triangle-correlation-heatmap-with-p-values-labeled/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="Chatomics">
  <meta property="og:url" content="/post/how-to-make-a-triangle-correlation-heatmap-with-p-values-labeled/">
  <meta property="og:title" content="How to make a triangle correlation heatmap with p-values labeled  | Chatomics">
  <meta property="og:description" content="In this blog post, I am going to show you how to make a correlation heatmap with p-values and significant values labeled in the heatmap body. Let’s use the PBMC single cell data as an example.
You may want to read my previous blog post How to do gene correlation for single-cell RNAseq data.
Load libraries library(dplyr) library(Seurat) library(patchwork) library(ggplot2) library(ComplexHeatmap) library(SeuratData) library(hdWGCNA) library(WGCNA) set.seed(1234)  prepare the data data(&quot;pbmc3k&quot;) pbmc3k #&gt; An object of class Seurat #&gt; 13714 features across 2700 samples within 1 assay #&gt; Active assay: RNA (13714 features, 0 variable features) ## routine processing pbmc3k&lt;- pbmc3k %&gt;% NormalizeData(normalization.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2023-02-05T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2023-02-05T00:00:00&#43;00:00">
  

  
  

  <title>How to make a triangle correlation heatmap with p-values labeled  | Chatomics</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">Chatomics</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>Newsletter</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">How to make a triangle correlation heatmap with p-values labeled </h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2023-02-05 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Feb 5, 2023
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    4 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/how-to-make-a-triangle-correlation-heatmap-with-p-values-labeled/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >, 
    
    <a href="/categories/genomics/">genomics</a
    >, 
    
    <a href="/categories/r/">R</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=How%20to%20make%20a%20triangle%20correlation%20heatmap%20with%20p-values%20labeled%20&amp;url=%2fpost%2fhow-to-make-a-triangle-correlation-heatmap-with-p-values-labeled%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fhow-to-make-a-triangle-correlation-heatmap-with-p-values-labeled%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fhow-to-make-a-triangle-correlation-heatmap-with-p-values-labeled%2f&amp;title=How%20to%20make%20a%20triangle%20correlation%20heatmap%20with%20p-values%20labeled%20"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fhow-to-make-a-triangle-correlation-heatmap-with-p-values-labeled%2f&amp;title=How%20to%20make%20a%20triangle%20correlation%20heatmap%20with%20p-values%20labeled%20"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=How%20to%20make%20a%20triangle%20correlation%20heatmap%20with%20p-values%20labeled%20&amp;body=%2fpost%2fhow-to-make-a-triangle-correlation-heatmap-with-p-values-labeled%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<p>In this blog post, I am going to show you how to make a correlation heatmap with p-values
and significant values labeled in the heatmap body. Let’s use the PBMC single cell data
as an example.</p>
<p>You may want to read my previous blog post <a href="https://divingintogeneticsandgenomics.rbind.io/post/how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-2-using-meta-cell/">How to do gene correlation for single-cell RNAseq data</a>.</p>
<div id="load-libraries" class="section level3">
<h3>Load libraries</h3>
<pre class="r"><code>library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(ComplexHeatmap)
library(SeuratData)
library(hdWGCNA)
library(WGCNA)
set.seed(1234)</code></pre>
</div>
<div id="prepare-the-data" class="section level3">
<h3>prepare the data</h3>
<pre class="r"><code>data(&quot;pbmc3k&quot;)

pbmc3k</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 13714 features across 2700 samples within 1 assay 
#&gt; Active assay: RNA (13714 features, 0 variable features)</code></pre>
<pre class="r"><code>## routine processing
pbmc3k&lt;- pbmc3k %&gt;% 
  NormalizeData(normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000) %&gt;%
  FindVariableFeatures(selection.method = &quot;vst&quot;, nfeatures = 2000) %&gt;%
  ScaleData() %&gt;%
  RunPCA(verbose = FALSE) %&gt;%
  FindNeighbors(dims = 1:10, verbose = FALSE) %&gt;%
  FindClusters(resolution = 0.5, verbose = FALSE) %&gt;%
  RunUMAP(dims = 1:10, verbose = FALSE)

Idents(pbmc3k)&lt;- pbmc3k$seurat_annotations

pbmc3k&lt;- pbmc3k[, !is.na(pbmc3k$seurat_annotations)]</code></pre>
</div>
<div id="construct-metacell" class="section level3">
<h3>construct metacell</h3>
<pre class="r"><code>pbmc3k &lt;- SetupForWGCNA(
  pbmc3k,
  gene_select = &quot;fraction&quot;, # the gene selection approach
  fraction = 0.05, # fraction of cells that a gene needs to be expressed in order to be included
  wgcna_name = &quot;tutorial&quot; # the name of the hdWGCNA experiment
)


# construct metacells in each group.
# choosing K is an art. 
pbmc3k &lt;- MetacellsByGroups(
  seurat_obj = pbmc3k,
  group.by = c(&quot;seurat_annotations&quot;, &quot;orig.ident&quot;), # specify the columns in seurat_obj@meta.data to group by
  k = 10, # nearest-neighbors parameter
  max_shared = 5, # maximum number of shared cells between two metacells
  ident.group = &#39;seurat_annotations&#39; # set the Idents of the metacell seurat object
)

# extract the metacell seurat object 
pbmc_metacell &lt;- GetMetacellObject(pbmc3k)

pbmc_metacell@meta.data %&gt;% head()</code></pre>
<pre><code>#&gt;            orig.ident nCount_RNA nFeature_RNA
#&gt; B#pbmc3k_1     pbmc3k     1794.8         3237
#&gt; B#pbmc3k_2     pbmc3k     4170.4         4962
#&gt; B#pbmc3k_3     pbmc3k     1765.7         3233
#&gt; B#pbmc3k_4     pbmc3k     1731.3         3060
#&gt; B#pbmc3k_5     pbmc3k     1982.2         3468
#&gt; B#pbmc3k_6     pbmc3k     3084.5         4351
#&gt;                                                                                                                                                     cells_merged
#&gt; B#pbmc3k_1 CGCCATTGGAGCAG,TGCCGACTCTCCCA,GGTACATGAAAGCA,GATTGGACGGTGTT,ATCCCGTGGCTGAT,CCAGCTACCAGCTA,TTGGAGACTATGGC,GAAGTCACCCTGTC,CGAGGGCTACGACT,TTATCCGACTAGTG
#&gt; B#pbmc3k_2 AAACATTGAGCTAC,TATGTCACGGAACG,AAACTTGAAAAACG,GGAGCCACCTTCTA,CATTTGTGGGATCT,GCTAGAACAGAGGC,CAGGTTGAGGATCT,ACTGAGACGTTGGT,ACGAAGCTCTGAGT,CGCGAGACAGGTCT
#&gt; B#pbmc3k_3 CTACTATGTAAAGG,CCAGGTCTACACCA,TTGGTACTGAATCC,ATCGCGCTTTTCGT,GCACTAGAAGATGA,ATGACGTGACGACT,GGGCCAACGCGTTA,CGCCATTGGAGCAG,GAAAGCCTACGTTG,CATCATACGGAGCA
#&gt; B#pbmc3k_4 GAGAAATGTTCTCA,ACCCAAGATTCACT,GGGCCAACGCGTTA,GAGTGGGAGTCTTT,TATTGCTGCCGTTC,GGCAATACGCTAAC,CGAGGGCTACGACT,CAGTGATGTACGCA,TCGATTTGTCGTGA,TGATCACTTCTACT
#&gt; B#pbmc3k_5 CAGTTACTAAGGTA,AAGTGGCTTGGAGG,TGATTCACTGTCAG,TGCGAAACAGTCAC,TATACCACCTGATG,TTATTCCTGGTACT,TTGACACTGATAAG,CGAGGGCTACGACT,TACTCAACTGCTAG,TAGCTACTGAATAG
#&gt; B#pbmc3k_6 TTCCAAACCTCCCA,CGCGATCTTTCTTG,AAATCAACAATGCC,ATCCCGTGGCTGAT,ACGAAGCTCTGAGT,TTTAGAGATCCTCG,AAACATTGAGCTAC,ACGACCCTTGACAC,GTAGCAACCATTTC,GAATTAACTGAAGA
#&gt;            seurat_annotations
#&gt; B#pbmc3k_1                  B
#&gt; B#pbmc3k_2                  B
#&gt; B#pbmc3k_3                  B
#&gt; B#pbmc3k_4                  B
#&gt; B#pbmc3k_5                  B
#&gt; B#pbmc3k_6                  B</code></pre>
<p><code>Hmisc::rcorr</code> function gives the pair-wise correlation coefficient and p-value</p>
<pre class="r"><code>genes&lt;- c(&quot;GNLY&quot;, &quot;NKG7&quot;, &quot;KLRD1&quot;, &quot;ITGB2&quot;, &quot;PRF1&quot;, &quot;IFNG&quot;, &quot;NCAM1&quot;, &quot;FCGR3A&quot;, &quot;CCR7&quot;, &quot;CXCR3&quot;, &quot;IL2RB&quot;)
cells&lt;- WhichCells(pbmc_metacell, expression = seurat_annotations == &quot;NK&quot;)

mat&lt;- pbmc_metacell@assays$RNA@data[genes, cells] %&gt;% 
  as.matrix() %&gt;%
  t()

# Hmisc package gives the pair-wise correlation coefficient and p-value
cor_res&lt;- Hmisc::rcorr(mat) 
cor_mat&lt;- cor_res$r

# sometimes, you may have NA in the matrix, and clustering does not play well with it
# a simple hack is to turn the NA to 0
cor_mat[is.na(cor_mat)]&lt;- 0


cor_p&lt;- cor_res$P
cor_p</code></pre>
<pre><code>#&gt;                GNLY         NKG7      KLRD1     ITGB2         PRF1       IFNG
#&gt; GNLY             NA 1.034720e-02 0.01446758 0.2740351 5.598974e-06 0.13474999
#&gt; NKG7   1.034720e-02           NA 0.09914756 0.2480082 7.635844e-05 0.59502786
#&gt; KLRD1  1.446758e-02 9.914756e-02         NA 0.1174652 1.951340e-01 0.54200187
#&gt; ITGB2  2.740351e-01 2.480082e-01 0.11746519        NA 1.956304e-01 0.77673864
#&gt; PRF1   5.598974e-06 7.635844e-05 0.19513395 0.1956304           NA 0.27047205
#&gt; IFNG   1.347500e-01 5.950279e-01 0.54200187 0.7767386 2.704720e-01         NA
#&gt; NCAM1  8.662654e-03 6.572153e-01 0.39486873 0.3767390 3.435796e-04 0.60012618
#&gt; FCGR3A 2.259989e-05 4.024618e-07 0.92716018 0.3407721 3.353928e-10 0.45453683
#&gt; CCR7   4.702197e-02 1.452948e-02 0.34889863 0.2674199 2.461824e-04 0.78811106
#&gt; CXCR3  9.232615e-13 8.725519e-04 0.06138810 0.1045795 6.629722e-05 0.02079542
#&gt; IL2RB  1.199037e-02 7.699923e-02 0.16780117 0.1016665 9.815965e-04 0.09288220
#&gt;               NCAM1       FCGR3A         CCR7        CXCR3        IL2RB
#&gt; GNLY   0.0086626537 2.259989e-05 4.702197e-02 9.232615e-13 0.0119903711
#&gt; NKG7   0.6572153099 4.024618e-07 1.452948e-02 8.725519e-04 0.0769992304
#&gt; KLRD1  0.3948687306 9.271602e-01 3.488986e-01 6.138810e-02 0.1678011737
#&gt; ITGB2  0.3767389762 3.407721e-01 2.674199e-01 1.045795e-01 0.1016664987
#&gt; PRF1   0.0003435796 3.353928e-10 2.461824e-04 6.629722e-05 0.0009815965
#&gt; IFNG   0.6001261787 4.545368e-01 7.881111e-01 2.079542e-02 0.0928821980
#&gt; NCAM1            NA 4.384662e-01 2.119770e-02 1.787253e-01 0.7278303881
#&gt; FCGR3A 0.4384662288           NA 1.065590e-01 1.939171e-06 0.0523142134
#&gt; CCR7   0.0211977006 1.065590e-01           NA 5.449032e-07 0.0040966177
#&gt; CXCR3  0.1787252622 1.939171e-06 5.449032e-07           NA 0.0001233260
#&gt; IL2RB  0.7278303881 5.231421e-02 4.096618e-03 1.233260e-04           NA</code></pre>
<pre class="r"><code>## add significant **
cor_p[is.nan(cor_p)]&lt;- 1

## the diagonal are NA, make them 1 
cor_p[is.na(cor_p)]&lt;- 1

col_fun&lt;- circlize::colorRamp2(c(-1, 0, 1), c(&quot;green&quot;, &quot;white&quot;, &quot;red&quot;))</code></pre>
<p>Only lable the correlation coefficients with p-values that are &lt;=0.05;
add * for p value &lt;=0.05 and ** for p value &lt;=0.01</p>
<pre class="r"><code>cell_fun = function(j, i, x, y, w, h, fill){
    if(as.numeric(x) &lt;= 1 - as.numeric(y) + 1e-6) {
            grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
    }
  
    if (cor_p[i, j]  &lt; 0.01 &amp; as.numeric(x) &lt;= 1 - as.numeric(y) + 1e-6){
      grid.text(paste0(sprintf(&quot;%.2f&quot;, cor_mat[i, j]),&quot;**&quot;), x, y, gp = gpar(fontsize = 10))
    } else if (cor_p[i, j]  &lt;= 0.05 &amp; as.numeric(x) &lt;= 1 - as.numeric(y) + 1e-6){
      grid.text(paste0(sprintf(&quot;%.2f&quot;, cor_mat[i, j]),&quot;*&quot;), x, y, gp = gpar(fontsize = 10))
    }
}


hp&lt;- ComplexHeatmap::Heatmap(cor_mat,
                        rect_gp = gpar(type = &quot;none&quot;),
                        column_dend_side = &quot;bottom&quot;,
                        column_title = &quot;NK cells&quot;,
                        name = &quot;correlation&quot;, col = col_fun,
                        cell_fun = cell_fun,
                        cluster_rows = T, cluster_columns = T,
                        row_names_side = &quot;left&quot;)

lgd_list = list(
    Legend( labels = c(&quot;&lt;0.01&quot;, &quot;&lt;0.05&quot;), title = &quot;pvalue&quot;,
            graphics = list(
              function(x, y, w, h) grid.text(&quot;**&quot;, x = x, y = y,
                                               gp = gpar(fill = &quot;black&quot;)),
              function(x, y, w, h) grid.text(&quot;*&quot;, x = x, y = y,
                                               gp = gpar(fill = &quot;black&quot;)))
            ))


draw(hp, annotation_legend_list = lgd_list, ht_gap = unit(1, &quot;cm&quot;) )</code></pre>
<p><img src="/post/2023-02-05-how-to-make-a-triangle-correlation-heatmap-with-p-values-labeled_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
</div>
<div id="futher-reading" class="section level3">
<h3>Futher reading</h3>
<ul>
<li><a href="https://github.com/jokergoo/ComplexHeatmap/issues/675" class="uri">https://github.com/jokergoo/ComplexHeatmap/issues/675</a></li>
<li><a href="https://jokergoo.github.io/2021/07/22/make-triangle-heatmap/" class="uri">https://jokergoo.github.io/2021/07/22/make-triangle-heatmap/</a></li>
</ul>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/bioconductor/">Bioconductor</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/heatmap/">heatmap</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-2-using-meta-cell/">How to do gene correlation for single-cell RNAseq data (part 2) using meta-cell </a></li>
        
        <li><a href="/post/partial-least-square-regression-for-marker-gene-identification-in-scrnaseq-data/">Partial least square regression for marker gene identification in scRNAseq data</a></li>
        
        <li><a href="/post/obtain-metadata-for-public-datasets-in-geo/">Obtain metadata for public datasets in GEO</a></li>
        
        <li><a href="/post/develop-bioconductor-packages-with-docker-container/">Develop Bioconductor packages with docker container</a></li>
        
        <li><a href="/post/a-tale-of-two-heatmap-functions/">A tale of two heatmap functions</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/how-to-commit-changes-to-a-docker-image/" rel="next">How to commit changes to a docker image</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-2-using-meta-cell/" rel="prev">How to do gene correlation for single-cell RNAseq data (part 2) using meta-cell </a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

