---
title: The end of 2019
author: Ming Tang
date: '2019-12-29'
slug: the-end-of-2019
categories:
  - bioinformatics
tags: []
header:
  caption: ''
  image: ''
---

It is the end of 2019. How time flies! It is a good time to reflect what I have 
achieved during the past year and what to look forward in 2020. I wrote a post for 
2018 [here](https://divingintogeneticsandgenomics.rbind.io/post/the-end-of-2018/).
I am not the only one who has impostor syndrome :) It is important to celebrate your 
small successes/achievements by writing them down.

1. I taught a snakemake and scRNAseq workshop during the FAS informatics 2-week [nanocourse](https://divingintogeneticsandgenomics.rbind.io/talk/2019-harvard-fas-workshop/). I love teaching biologists computing skills that I have learned from scratch. Nowadays, almost [every biology is computational biology](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2002050). I also helped/taught lab members in Dulac lab with programming. 


2. I had my first R package [scclusteval](https://github.com/crazyhottommy/scclusteval) for estimating cluster stability in single-cell RNAseq data in github and am writing it up for a small paper. I had some changes that have not been committed to github yet. I presented it in the [2019 bioconductor annual meeting](http://bioc2019.bioconductor.org/) which was held in NYC. I met many wonderful tweeps during the conference including [Rob Patro](https://twitter.com/nomad421), [Mike Love](https://twitter.com/mikelove), [Ansul kundaje](https://twitter.com/anshulkundaje), [Qian liu](https://twitter.com/QianLiu28878838) (congrats for the new professorship!), [Lihua Zhu](https://profiles.umassmed.edu/display/129880), [Dave Tang](https://twitter.com/davetang31), [Avi Srivastava](https://twitter.com/k3yavi) and many others. Thanks bioconductor for the travel award! bioconductor 2020 will be in local Boston, so I will see everyone here!

3. My first co-first author computational paper is still under review after revision. I have my fingers crossed for the reviewers' comments back. It has tons of experimental data and computational analysis of both in-house and public data. You can have a look in biorxiv https://www.biorxiv.org/content/10.1101/507202v1. I promise to update the README and pipeline once it is accepted :) https://github.com/crazyhottommy/pyflow-ChIPseq.

4. I started working on scATACseq in March 2019. It is a brand new field and I have learned quite a bit during the journey. I learned the sparsity of the matrix (more sparse than scRNAseq given the nature of the experiment: we only 2 copies of diploid DNA). I wrote some [blog posts](https://divingintogeneticsandgenomics.rbind.io/#posts) on it. I am developing an R/bioconductor package to deal with scATACseq data https://github.com/crazyhottommy/scATACutils. I plan to add more functions and improve the documentations. I presented it in the cold spring harbor [Single Cell Analyses](https://meetings.cshl.edu/meetings.aspx?meet=SINGLE&year=19) meeting.It is amazing to see the development of the single-cell field in the past 10 years! I met some awesome people there including [Luciano Martelotto](https://twitter.com/LGMartelotto) and got to see some of my old MD Anderson Colleagues. 

5. I said I would start learning some deep learning in 2019. That did not happen much. Instead, I started watching some linear algebra courses by MIT Gilbert Strang [18.06](https://ocw.mit.edu/courses/mathematics/18-06-linear-algebra-spring-2010/index.htm) and [18.065](https://ocw.mit.edu/courses/mathematics/18-065-matrix-methods-in-data-analysis-signal-processing-and-machine-learning-spring-2018/index.htm). I have finished watching 70% of the videos and really enjoyed them. I got a much better idea of matrix dimentionality (along with space and subspace) and matrix factorization(e.g. SVD). I highly recommend youtube videos from 3blue1brown: [Essense of linear algebra](https://www.3blue1brown.com/essence-of-linear-algebra-page) as well. Thanks [Yi Zhang](https://twitter.com/zhiiiyang) for recommending. At the same time, I slowly picked up [Deep learning with R](https://www.manning.com/books/deep-learning-with-r).

6. I started [using docker and singularity](https://divingintogeneticsandgenomics.rbind.io/post/run-rstudio-server-with-singularity-on-hpc/) for reproducible computing. It is a life-saver for me. Thanks Nathan for helping me along the way.

7. I finally met [Jason Willimas](https://twitter.com/JasonWilliamsNY) in person together with [Damien](https://twitter.com/dccc_phd). Jason is a very nice person. Next time we will treat you when you are in Boston again. I am sure this is in the right category of my success list :) 

I am grateful that I am in a such supporting position and I have been learning new things. Thanks everyone who has helped me along the way.

In the coming 2020, I should

1. learn more stats. I registered one class from Harvard Extension school and will see how it goes. Lacking statistics background hurts me. 

2. do better in data and project management. I will start using [dtool](https://pypi.org/project/dtool/).

2. write some papers up. A good paper is a finished paper. I hear you.

3. try to be a better human, a better husband and father and then a better researcher.


