---
title: 'My odyssey of obtaining scRNAseq metadata '
author: Ming Tang
date: '2022-06-08'
slug: my-odyssey-of-obtaining-scrnaseq-metadata
categories:
  - bioinformatics
  - scRNAseq
  - R
tags:
  - scRNAseq
header:
  caption: ''
  image: ''
---

### ffq failed me 

```bash
(base) âžœ  ~ ffq GSE169246
[2022-06-06 14:32:22,716]    INFO Parsing GEO GSE169246
[2022-06-06 14:32:24,418]    INFO Finding supplementary files for GEO GSE169246
[2022-06-06 14:32:28,001] WARNING There are 83 samples for GSE169246
[2022-06-06 14:32:28,001]    INFO Parsing GSM GSM5188367
[2022-06-06 14:32:33,047]    INFO Finding supplementary files for GSM GSM5188367
[2022-06-06 14:32:34,409]    INFO No supplementary files found for GSM5188367
[2022-06-06 14:32:39,659]    INFO Getting sample for GSM5188367
[2022-06-06 14:32:41,847] WARNING No sample found. Either the provided GSM accession is invalid or raw data was not provided for this record
```

### get the metadata for this dataset

other methods: pysradb https://github.com/saketkc/pysradb

```{r}
library(GEOquery)
library(tidyverse)
library(here)
#getGEOSuppFiles(GEO ="GSE176021")
meta<- getGEO(GEO= "GSE169246",GSEMatrix=FALSE)


## metadata 
meta@gsms$GSM5188440@header$characteristics_ch1

## sample ID
meta@gsms$GSM5188440@header$title 

## patient ID
meta@gsms$GSM5188440@header$source_name_ch1


meta1<- purrr::map(meta@gsms, ~.x@header$characteristics_ch1) %>%
        stack() %>%
        separate(values, into = c("condition", "value"), sep= ": ")%>%
        pivot_wider(names_from= condition, values_from = value) %>%
        janitor::clean_names()
        

meta2<- purrr::map(meta@gsms, ~.x@header$title) %>%
        stack() %>%
        dplyr::rename(sample_id = values)


meta3<- purrr::map(meta@gsms, ~.x@header$source_name_ch1) %>%
        stack() %>%
        dplyr::rename(patient_id = values)



head(meta1)
head(meta2)
head(meta3)

##mereg all of them

meta<- left_join(meta1, meta2) %>%
        left_join(meta3)

head(meta)

## remove the scATACseq samples
meta<- meta %>%
        filter(!str_detect(sample_id, "ATAC-seq"))
```


### Get the response data

Download the data 
```{r}
clinical_data<- readxl::read_xlsx(here("data/1-s2.0-S1535610821004992-mmc2.xlsx"), skip = 1)

##
clinical_data<- clinical_data %>%
        janitor::clean_names() %>%
        slice(2:23) %>%
        tidyr::fill(treatment) %>%
        dplyr::select(!(tumor:x15)) 


## yeah, and N for not available , and sometimes - means not available.let's fix that
clinical_data<- clinical_data %>%
        #trim off the white spaces. this will help the later step too
        mutate_all(~str_trim(.x, side ="both")) %>%
        mutate_at(vars(-pd_l1), ~str_replace(.x , "^-$", NA_character_)) %>%
        mutate_all( ~str_replace(.x , "^N$", NA_character_))


```


```{r}
meta<- meta %>%
        left_join(clinical_data) %>%
        arrange(patient_id, biopsied_lesion) %>%
        select(-biopsied_lesion)

write_tsv(meta, here("results/GSE169246_sample_level_meta.tsv"))
```


