---
title: How to construct a spatial object in Seurat
author: Ming Tang
date: '2023-03-30'
slug: how-to-construct-a-spatial-object-in-seurat
categories:
  - bioinformatics
  - spatial
  - R
  - single-cell
tags:
  - bioinformatics
  - R
  - spatial
  - single-cell
header:
  caption: ''
  image: ''
editor_options: 
  chunk_output_type: console
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<div id="sign-up-for-my-newsletter-to-not-miss-a-post-like-this" class="section level3">
<h3>Sign up for my newsletter to not miss a post like this</h3>
<p><a href="https://divingintogeneticsandgenomics.ck.page/newsletter" class="uri">https://divingintogeneticsandgenomics.ck.page/newsletter</a></p>
<p>Single-cell spatial transcriptome data is a new and advanced technology that combines the study of individual cells’ genes and their location in a tissue to understand the complex cellular and molecular differences within it. This allows scientists to investigate how genes are expressed and how cells interact with each other with much greater detail than before.</p>
<p>Do not be intimidated by the complexity of the data. If you already know how to analyze regular single cell data (e.g., 10x data), it is not hard to understand and analyze spatial data. In this blog post, I will walk you through the basic analsyis steps.</p>
</div>
<div id="download-the-demo-data" class="section level3">
<h3>download the demo data</h3>
<p>Following <a href="https://satijalab.org/seurat/articles/spatial_vignette_2.html" class="uri">https://satijalab.org/seurat/articles/spatial_vignette_2.html</a></p>
<p>Download the public breast <code>vizgen</code> cancer data.</p>
<pre class="bash"><code>gsutil -m cp -n gs://vz-ffpe-showcase/HumanOvarianCancerPatient2Slice2/cell_by_gene.csv \
    gs://vz-ffpe-showcase/HumanOvarianCancerPatient2Slice2/cell_metadata.csv \
    ./spatial_data/
</code></pre>
<p>It is critical to understand the data files before you do anything.</p>
<blockquote>
<p>The data for each sample is split up across a few different files. <code>cell_by_gene.csv</code> is the standard file containing cells as rows and genes as columns. <code>cell_metadata.csv</code> contains additional information for cells relating to spatial data, namely x-y coordinates for each cell (min, center, and max). These files are both preprocessed. To access more raw data, we can examine the cell_bounds/ folder and detected_transcripts.csv. The former contains many files, named feature_data_{fov}.hdf5, corresponding to the fov column in cell_metadata.csv. These hdf5 files contain boundary coordinates for each cell, giving a less boxy outline than provided in the processed data. detected_transcripts.csv contains transcripts in each row with their respective x-y coordinates, which are then assigned to cells based on the cell boundaries.</p>
</blockquote>
</div>
<div id="read-in-the-data-and-pre-process" class="section level3">
<h3>read in the data and pre-process</h3>
<pre class="r"><code>library(Seurat)
library(here)
library(ggplot2)
library(dplyr)

# the LoadVizgen function requires the raw segmentation files which is too big. We will only use the x,y coordinates 
# vizgen.obj &lt;- LoadVizgen(data.dir = here(&quot;data&quot;))

vizgen.input &lt;- ReadVizgen(data.dir = &quot;~/blog_data/spatial_data&quot;, type = &quot;centroids&quot;)</code></pre>
<p>For most of the analysis, we will only need the x,y coordinates (the center of the cell). You can also read in
the raw segmentation file( which gives you more detailed cell shape information), or set <code>type = "box"</code> to get the
rectangular information of a cell (xmin, xmax, ymin and ymax).</p>
<p><code>vizgen.input</code> is a list containing the count matrix and the spatial centrioids.</p>
<pre class="r"><code>vizgen.input$centroids %&gt;% head()</code></pre>
<pre><code>#&gt;           x        y cell
#&gt; 1 10145.793 5611.687 7650
#&gt; 2  9975.309 5626.726 7654
#&gt; 3 10129.183 5630.601 7655
#&gt; 4 10112.692 5635.038 7656
#&gt; 5 10151.574 5634.673 7657
#&gt; 6 10099.624 5636.831 7658</code></pre>
<pre class="r"><code>## This gives you the image 
ggplot(vizgen.input$centroids, aes(x= x, y = y))+
        geom_point(size = 0.1, color = &quot;grey&quot;) +
        theme_classic()</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>Create a seurat object. The documentation for making a spatial object is sparse.
I went to the <a href="https://github.com/satijalab/seurat/blob/763259d05991d40721dee99c9919ec6d4491d15e/R/convenience.R#L139">source code</a>
of <code>LoadVizgen</code> and came up with the code below.</p>
<p>You can read the code from the same link and see how other types of spatial data (10x Xenium, nanostring) are read into Seurat.</p>
<pre class="r"><code>## remove the Blank-* control probes
blank_index&lt;- which(stringr::str_detect(rownames(vizgen.input$transcripts), &quot;^Blank&quot;))

transcripts&lt;-vizgen.input$transcripts[-blank_index, ]

dim(vizgen.input$transcripts)</code></pre>
<pre><code>#&gt; [1]   550 71381</code></pre>
<pre class="r"><code>dim(transcripts)</code></pre>
<pre><code>#&gt; [1]   500 71381</code></pre>
<p>There are 50 such probes.</p>
</div>
<div id="create-a-seurat-spatial-object" class="section level3">
<h3>create a Seurat spatial object</h3>
<pre class="r"><code>vizgen.obj&lt;- CreateSeuratObject(counts = transcripts, assay = &quot;Vizgen&quot;)

cents &lt;- CreateCentroids(vizgen.input$centroids)
segmentations.data &lt;- list(
    &quot;centroids&quot; = cents,
    &quot;segmentation&quot; = NULL
  )

coords &lt;- CreateFOV(
    coords = segmentations.data,
    type = c(&quot;segmentation&quot;, &quot;centroids&quot;),
    molecules = NULL,
    assay = &quot;Vizgen&quot;
  )
 
vizgen.obj[[&quot;p2s2&quot;]] &lt;- coords
 
GetTissueCoordinates(vizgen.obj[[&quot;p2s2&quot;]][[&quot;centroids&quot;]]) %&gt;%
        head()</code></pre>
<pre><code>#&gt;           x        y cell
#&gt; 1 10145.793 5611.687 7650
#&gt; 2  9975.309 5626.726 7654
#&gt; 3 10129.183 5630.601 7655
#&gt; 4 10112.692 5635.038 7656
#&gt; 5 10151.574 5634.673 7657
#&gt; 6 10099.624 5636.831 7658</code></pre>
<pre class="r"><code>ImageDimPlot(vizgen.obj, fov = &quot;p2s2&quot;, cols = &quot;polychrome&quot;, axes = TRUE)</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>Note, from this recent paper <a href="https://www.nature.com/articles/s41467-023-37168-7">A comprehensive benchmarking with practical guidelines for cellular deconvolution of spatial transcriptomics</a>, <code>sctransform normalization</code> works worse than log normalization for deconvolution.</p>
<p>This part is the same as regular single-cell RNAseq data. For clustering, one can also incorporate the spatial information, but that’s
something out of this tutorial.</p>
<p>I will use log normalization.</p>
<pre class="r"><code>vizgen.obj &lt;- NormalizeData(vizgen.obj, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000) %&gt;%
  ScaleData() 
vizgen.obj &lt;- RunPCA(vizgen.obj, npcs = 30, features = rownames(vizgen.obj))
vizgen.obj &lt;- RunUMAP(vizgen.obj, dims = 1:30)
vizgen.obj &lt;- FindNeighbors(vizgen.obj, reduction = &quot;pca&quot;, dims = 1:30)
vizgen.obj &lt;- FindClusters(vizgen.obj, resolution = 0.3)</code></pre>
<pre><code>#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
#&gt; 
#&gt; Number of nodes: 71381
#&gt; Number of edges: 2100209
#&gt; 
#&gt; Running Louvain algorithm...
#&gt; Maximum modularity in 10 random starts: 0.9109
#&gt; Number of communities: 14
#&gt; Elapsed time: 23 seconds</code></pre>
</div>
<div id="umap-by-cluster" class="section level3">
<h3>UMAP by cluster</h3>
<pre class="r"><code>DimPlot(vizgen.obj, reduction = &quot;umap&quot;, label = TRUE)</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="image-plot-by-cluster" class="section level3">
<h3>Image plot by cluster</h3>
<pre class="r"><code>ImageDimPlot(vizgen.obj, fov = &quot;p2s2&quot;, cols = &quot;polychrome&quot;, axes = TRUE)</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="find-cell-markers" class="section level3">
<h3>Find cell markers</h3>
<pre class="r"><code>library(tictoc)
tic()
markers&lt;- presto::wilcoxauc(vizgen.obj, &#39;seurat_clusters&#39;, assay = &#39;data&#39;,seurat_assay= &quot;Vizgen&quot; )
toc()</code></pre>
<pre><code>#&gt; 1.137 sec elapsed</code></pre>
<pre class="r"><code>top_markers&lt;- presto::top_markers(markers, n =3)

top_markers</code></pre>
<pre><code>#&gt; # A tibble: 3 × 13
#&gt;    rank `0`   `1`    `10`  `11`  `2`   `3`   `4`   `5`   `6`   `7`   `8`   `9`  
#&gt;   &lt;int&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
#&gt; 1     1 MUC1  LRP1   KIT   MZB1  FN1   COL1… C1QC  CDKN… COL4… LAMC2 ETS1  IL21 
#&gt; 2     2 PKM   ENG    CTSG  CD79A SFRP2 FFAR2 CD14  LAMC2 VWF   MKI67 IL2RB IDO2 
#&gt; 3     3 SOX9  COL1A1 ITGAM HLA-B COL1… CXCR1 CSF1R SERP… PECA… FOXM1 CD3E  XCL1</code></pre>
<pre class="r"><code>top_markers&lt;- top_markers %&gt;% select(-rank) %&gt;% stack() %&gt;%
        pull(values) %&gt;%
        unique()

top_markers</code></pre>
<pre><code>#&gt;  [1] &quot;MUC1&quot;     &quot;PKM&quot;      &quot;SOX9&quot;     &quot;LRP1&quot;     &quot;ENG&quot;      &quot;COL1A1&quot;  
#&gt;  [7] &quot;KIT&quot;      &quot;CTSG&quot;     &quot;ITGAM&quot;    &quot;MZB1&quot;     &quot;CD79A&quot;    &quot;HLA-B&quot;   
#&gt; [13] &quot;FN1&quot;      &quot;SFRP2&quot;    &quot;FFAR2&quot;    &quot;CXCR1&quot;    &quot;C1QC&quot;     &quot;CD14&quot;    
#&gt; [19] &quot;CSF1R&quot;    &quot;CDKN1A&quot;   &quot;LAMC2&quot;    &quot;SERPINA1&quot; &quot;COL4A1&quot;   &quot;VWF&quot;     
#&gt; [25] &quot;PECAM1&quot;   &quot;MKI67&quot;    &quot;FOXM1&quot;    &quot;ETS1&quot;     &quot;IL2RB&quot;    &quot;CD3E&quot;    
#&gt; [31] &quot;IL21&quot;     &quot;IDO2&quot;     &quot;XCL1&quot;</code></pre>
<pre class="r"><code>manual_markers&lt;- c(&quot;CD3D&quot;, &quot;CD4&quot;, &quot;CD8A&quot;, &quot;CD8B&quot;, &quot;CD14&quot;,
                                            &quot;MS4A1&quot;, &quot;FCGR3A&quot;, &quot;PTPRC&quot;, &quot;EPCAM&quot;, 
                                            &quot;KIT&quot;, &quot;PDGFA&quot;, &quot;CCR7&quot;,&quot;GNLY&quot;, 
                                            &quot;PRF1&quot;, &quot;GZMB&quot;, &quot;COL1A1&quot;, &quot;PECAM1&quot;, 
                                            &quot;NKG7&quot;,&quot;XCL1&quot;, &quot;ICOS&quot;, &quot;PDCD1&quot;, &quot;TIGIT&quot;, 
                                            &quot;HAVCR2&quot;, &quot;NCAM1&quot;, &quot;CD79A&quot;, &quot;ITGAM&quot;, 
                                            &quot;ITGAX&quot;, &quot;FCER1A&quot;, &quot;CD86&quot;)
scCustomize::Clustered_DotPlot(vizgen.obj,  
                               features = c(top_markers, manual_markers), 
                               plot_km_elbow = FALSE,
                               row_label_size = 8)</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-10-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>table(vizgen.obj$seurat_clusters)</code></pre>
<pre><code>#&gt; 
#&gt;     0     1     2     3     4     5     6     7     8     9    10    11 
#&gt; 34147 12200  6251  5505  5324  3122  1656  1164  1065   647   183   117</code></pre>
<p>The cancer cell clusters (0,5,7) marked by EPCAM is the most abundant cells in this sample.
cluster 8 is the T cells; cluster 4 is mono/mac, cluster 10 seems to be the MAST (KIT) cells;
cluster 11 is the B (CD79A, MSA41) cells; cluster 2 is fibroblasts (SFRP2);
cluster 1 is a different fibroblasts cluster; cluster 6 is Endothelial cells (PECAM1)</p>
<p>cluster 9 probably is an artifact cluster. It is the strange circle in the UMAP.
Let’s take a look at the count distribution for all clusters.</p>
<pre class="r"><code>library(ggplot2)

vizgen.obj@meta.data %&gt;%
        ggplot(aes(x=  nCount_Vizgen)) +
        geom_histogram() +
        facet_wrap(~seurat_clusters, scales = &quot;free&quot;)</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-11-1.png" width="672" />
cluster 9 have very low counts for all the cells. Those cells should be removed from the pre-processing steps by:</p>
<p><code>CreateSeuratObject(min.features = 5)</code>.</p>
<p>min.features: Include cells where at least this many features are detected.</p>
<pre class="r"><code>p1 &lt;- ImageFeaturePlot(vizgen.obj, features = c(&quot;EPCAM&quot;, &quot;CD79A&quot;))
p1 </code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>We plotted the feature expression on the image using the <code>ImageFeaturePlot</code>. Let’s reproduce it using <code>ggplot2</code>:</p>
<pre class="r"><code>vizgen.input$centroids %&gt;%
        ggplot(aes(x=y, y = x)) +
        geom_point(color = &quot;grey&quot;, size = 0.2) +
        theme_classic(base_size = 14)</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Now, we need to highlight the points using the expression value of <code>EPCAM</code>:</p>
<p>Some helper function:</p>
<pre class="r"><code>matrix_to_expression_df&lt;- function(x, obj){
        df&lt;- x %&gt;%
                as.matrix() %&gt;% 
                as.data.frame() %&gt;%
                tibble::rownames_to_column(var= &quot;gene&quot;) %&gt;%
                tidyr::pivot_longer(cols = -1, names_to = &quot;cell&quot;, values_to = &quot;expression&quot;) %&gt;%
                tidyr::pivot_wider(names_from = &quot;gene&quot;, values_from = expression) %&gt;%
                left_join(obj@meta.data %&gt;% 
                                  tibble::rownames_to_column(var = &quot;cell&quot;))
        return(df)
}


get_expression_data&lt;- function(obj, assay = &quot;RNA&quot;, slot = &quot;data&quot;, 
                               genes = NULL, cells = NULL){
        if (is.null(genes) &amp; !is.null(cells)){
                df&lt;- GetAssayData(obj, assay = assay, slot = slot)[, cells, drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        } else if (!is.null(genes) &amp; is.null(cells)){
                df &lt;- GetAssayData(obj, assay = assay, slot = slot)[genes, , drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        } else if (is.null(genes &amp; is.null(cells))){
                df &lt;- GetAssayData(obj, assay = assay, slot = slot)[, , drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        } else {
                df&lt;- GetAssayData(obj, assay = assay, slot = slot)[genes, cells, drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        }
        return(df)
}</code></pre>
<p>Get the expression data and merge with the spatial information.</p>
<pre class="r"><code>df&lt;- get_expression_data(vizgen.obj, assay=&quot;Vizgen&quot;, slot = &quot;data&quot;, genes = &quot;EPCAM&quot;)

head(df)</code></pre>
<pre><code>#&gt; # A tibble: 6 × 7
#&gt;   cell  EPCAM orig.ident    nCount_Vizgen nFeature_Vizgen Vizgen_snn_res.0.3
#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;fct&gt;                 &lt;dbl&gt;           &lt;int&gt; &lt;fct&gt;             
#&gt; 1 0      0    SeuratProject             0               0 9                 
#&gt; 2 1      0    SeuratProject             0               0 9                 
#&gt; 3 2      0    SeuratProject            17              11 3                 
#&gt; 4 3      3.93 SeuratProject           400             108 0                 
#&gt; 5 4      0    SeuratProject             9               9 3                 
#&gt; 6 5      0    SeuratProject             3               3 3                 
#&gt; # … with 1 more variable: seurat_clusters &lt;fct&gt;</code></pre>
<pre class="r"><code>vizgen.input$centroids %&gt;% head()</code></pre>
<pre><code>#&gt;           x        y cell
#&gt; 1 10145.793 5611.687 7650
#&gt; 2  9975.309 5626.726 7654
#&gt; 3 10129.183 5630.601 7655
#&gt; 4 10112.692 5635.038 7656
#&gt; 5 10151.574 5634.673 7657
#&gt; 6 10099.624 5636.831 7658</code></pre>
<pre class="r"><code>df&lt;- vizgen.input$centroids %&gt;%
        left_join(df)

head(df)</code></pre>
<pre><code>#&gt;           x        y cell    EPCAM    orig.ident nCount_Vizgen nFeature_Vizgen
#&gt; 1 10145.793 5611.687 7650 4.091399 SeuratProject           680             173
#&gt; 2  9975.309 5626.726 7654 4.087071 SeuratProject           683             166
#&gt; 3 10129.183 5630.601 7655 4.087276 SeuratProject          1195             225
#&gt; 4 10112.692 5635.038 7656 4.255680 SeuratProject          1151             216
#&gt; 5 10151.574 5634.673 7657 0.000000 SeuratProject             0               0
#&gt; 6 10099.624 5636.831 7658 4.336921 SeuratProject          1325             218
#&gt;   Vizgen_snn_res.0.3 seurat_clusters
#&gt; 1                  4               4
#&gt; 2                  0               0
#&gt; 3                  0               0
#&gt; 4                  0               0
#&gt; 5                  9               9
#&gt; 6                  0               0</code></pre>
<p>Now, we have a dataframe with the spatial coordinates and the gene expression value.</p>
<p>Ready to plot:</p>
<pre class="r"><code>p1&lt;- ggplot(df, aes(x= y, y=x)) +
        geom_point(aes(color = EPCAM), size = 0.1) +
        scale_color_gradient(low=&quot;grey&quot;, high=&quot;red&quot;) +
        coord_fixed()+
        theme_classic(base_size = 14)
        

p2&lt;- ImageFeaturePlot(vizgen.obj, features =&quot;EPCAM&quot;)

p1 + p2</code></pre>
<p><img src="/post/2023-03-30-how-to-construct-a-spatial-object-in-seurat_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>Looks pretty similar!</p>
<p>Since it is spatial data, we want to explore the cell-cell contact information.
In a future post, I will show you how to find the closest cells within a fixed radius of a cell.</p>
</div>
<div id="take-home-messages" class="section level3">
<h3>Take home messages</h3>
<ul>
<li><p>Spatial data is similar to regular single-cell data for the count matrix, but with each cell, we have
additional x,y coordinates for the spatial information.</p></li>
<li><p>It is not hard to implement spatial visualization functions if you know the basics of <code>R</code> and <code>tidyverse</code>. <code>Seurat</code>
nicely integrated the spatial information to the Seurat object, so we can plot conveniently.</p></li>
<li><p>Bioconductor has a <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html">spatial experiment object</a> which is extended from the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a> object.</p></li>
<li><p>If you use python, check <a href="https://squidpy.readthedocs.io/en/stable/"><code>squidpy</code></a> and the <a href="https://monkeybread.readthedocs.io/en/latest/notebooks/tutorial.html"><code>monkeybread</code></a> package developed in our compbio group at <a href="https://www.immunitastx.com/">Immunitas</a>.</p></li>
</ul>
</div>
