---
title: Reuse the single cell data! How to create a seurat object from GEO datasets
author: Ming Tang
date: '2023-07-14'
slug: reuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets
categories:
  - bioinformatics
  - single-cell
tags:
  - single-cell
  - Seurat
  - bioinformatics
header:
  caption: ''
  image: ''
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
knitr::opts_chunk$set(
  comment = "#>", echo = TRUE, message= FALSE, warning = FALSE
)
knitr::opts_chunk$set(fig.width = 6, fig.height = 4)  # Set the desired width and height
```

### Download the data 

https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116256

```{bash eval=FALSE}
cd data/GSE116256
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE116nnn/GSE116256/suppl/GSE116256_RAW.tar
tar xvf GSE116256_RAW.tar
rm  GSE116256_RAW.tar
```

Depending on how the authors upload their data. 

Some authors may just upload the merged count matrix file. This is the easiest situation. 

In this dataset, each sample has a separate set of matrix (*dem.txt.gz), features and barcodes. 

Total, there are 43 samples. For each sample, it has an associated metadata file (*anno.txt.gz) too.


![](/img/geo_data.png)
You can inspect the files in command line:
```{bash eval=FALSE}
zless -S GSM3587923_AML1012-D0.dem.txt.gz
```


Go back to `R`.
Load the libraries:
```{r}
library(here)
library(stringr)
library(dplyr)
library(ggplot2)
library(Seurat)
library(purrr)
library(readr)
library(harmony)
library(scCustomize)
library(SeuratDisk)
```

read in the count matrix 

```{r}
# a small function to read the counts
read_counts<- function(file){
  x<- read_tsv(file)
  x<- as.data.frame(x)
  genes<- x$Gene
  x<- x[, -1]
  rownames(x)<- genes
  return(as.matrix(x))
}


counts_files<- list.files(here("data/GSE116256"), full.names = TRUE, pattern = "*dem.txt.gz")

samples<- map_chr(counts_files, basename) 

samples<- str_replace(samples, "(GSM[0-9]+_.+).dem.txt.gz", "\\1")

names(counts_files)<- samples

# for demonstration, only read in the first 4 samples
counts<- purrr::map(counts_files[1:4], read_counts)

```

read in the meta data 
```{r}
read_meta<- function(file){
  y<- read_tsv(file)
  y<- as.data.frame(y)
  cells<- y$Cell
  y<- y[,-1]
  rownames(y)<- cells
  return(y)
}


meta_files<- list.files(here("data/GSE116256"), full.names = TRUE, pattern = "*anno.txt.gz")
meta_names<- map_chr(meta_files, basename)

meta_names<- str_replace(meta_names, "(GSM[0-9]+_.+).anno.txt.gz", "\\1")
names(meta_files)<- meta_names

meta<- purrr::map(meta_files[1:4], read_meta)
```

### create a seurat object

```{r}
library(Matrix) #for sparse matrix
objs<- purrr::map2(counts, meta,  
                   ~CreateSeuratObject(counts = as(.x, "sparseMatrix"), 
                                       meta.data = .y))


# merge to a single object 
merged_seurat<- purrr::reduce(objs, function(x,y) {merge(x,y)})

## free memory
rm(counts)
rm(objs)
rm(meta)
gc()
```

### standard preprocess of the data 

```{r}
merged_seurat<- merged_seurat %>%
  NormalizeData(normalization.method = "LogNormalize", scale.factor = 10000) %>%
  FindVariableFeatures( selection.method = "vst", nfeatures = 2000) %>%
  ScaleData() %>%
  RunPCA() %>%
  RunHarmony(group.by.vars = "orig.ident", dims.use = 1:30) %>%
  RunUMAP(reduction = "harmony", dims = 1:30) %>%
  FindNeighbors(reduction = "harmony", dims = 1:30) %>% 
  FindClusters(resolution = 0.6)
```

### visualization

```{r}
DimPlot_scCustom(seurat_object = merged_seurat)
```

I have a youtube video on this too! subscribe to chatomics!

```{r echo=FALSE}
library("vembedr")

embed_url("https://www.youtube.com/watch?v=5e_8wr5Xx_Q")
```


