---
title: How to do gene correlation for single-cell RNAseq data (part 1)
author: Ming Tang
date: '2023-01-13'
slug: how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1
categories:
  - bioinformatics
  - scRNAseq
tags:
  - bioinformatics
  - tidyverse
header:
  caption: ''
  image: ''
  preview: yes
editor_options: 
  chunk_output_type: inline
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<div id="load-libraries" class="section level3">
<h3>Load libraries</h3>
<pre class="r"><code>library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(ComplexHeatmap)
library(SeuratData)
set.seed(1234)</code></pre>
</div>
<div id="prepare-the-data" class="section level3">
<h3>prepare the data</h3>
<pre class="r"><code>data(&quot;pbmc3k&quot;)

pbmc3k</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 13714 features across 2700 samples within 1 assay 
#&gt; Active assay: RNA (13714 features, 0 variable features)</code></pre>
<pre class="r"><code>## routine processing
pbmc3k&lt;- pbmc3k %&gt;% 
  NormalizeData(normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000) %&gt;%
  FindVariableFeatures(selection.method = &quot;vst&quot;, nfeatures = 2000) %&gt;%
  ScaleData() %&gt;%
  RunPCA(verbose = FALSE) %&gt;%
  FindNeighbors(dims = 1:10, verbose = FALSE) %&gt;%
  FindClusters(resolution = 0.5, verbose = FALSE) %&gt;%
  RunUMAP(dims = 1:10, verbose = FALSE)

Idents(pbmc3k)&lt;- pbmc3k$seurat_annotations

pbmc3k&lt;- pbmc3k[, !is.na(pbmc3k$seurat_annotations)]

DimPlot(pbmc3k, reduction = &quot;umap&quot;, label = TRUE)</code></pre>
<p><img src="/post/2023-01-13-how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1.en_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
</div>
<div id="some-functions-we-will-use" class="section level3">
<h3>some functions we will use</h3>
<pre class="r"><code>matrix_to_expression_df&lt;- function(x, obj){
        df&lt;- x %&gt;%
                as.matrix() %&gt;% 
                as.data.frame() %&gt;%
                tibble::rownames_to_column(var= &quot;gene&quot;) %&gt;%
                tidyr::pivot_longer(cols = -1, names_to = &quot;cell&quot;, values_to = &quot;expression&quot;) %&gt;%
                tidyr::pivot_wider(names_from = &quot;gene&quot;, values_from = expression) %&gt;%
                left_join(obj@meta.data %&gt;% 
                                  tibble::rownames_to_column(var = &quot;cell&quot;))
        return(df)
}


get_expression_data&lt;- function(obj, assay = &quot;RNA&quot;, slot = &quot;data&quot;, 
                               genes = NULL, cells = NULL){
        if (is.null(genes) &amp; !is.null(cells)){
                df&lt;- GetAssayData(obj, assay = assay, slot = slot)[, cells, drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        } else if (!is.null(genes) &amp; is.null(cells)){
                df &lt;- GetAssayData(obj, assay = assay, slot = slot)[genes, , drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        } else if (is.null(genes &amp; is.null(cells))){
                df &lt;- GetAssayData(obj, assay = assay, slot = slot)[, , drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        } else {
                df&lt;- GetAssayData(obj, assay = assay, slot = slot)[genes, cells, drop = FALSE] %&gt;%
                        matrix_to_expression_df(obj = obj)
        }
        return(df)
}</code></pre>
</div>
<div id="check-cd4-and-cd3d-correlation" class="section level3">
<h3>check CD4 and CD3D correlation</h3>
<pre class="r"><code>genes&lt;- c(&quot;CD3D&quot;, &quot;CD4&quot;)

expression_data&lt;- get_expression_data(pbmc3k, genes = genes)

head(expression_data)</code></pre>
<pre><code>#&gt; # A tibble: 6 × 9
#&gt;   cell            CD3D   CD4 orig.ident nCount_RNA nFeature_RNA seurat_annotati…
#&gt;   &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           
#&gt; 1 AAACATACAACCAC  2.86     0 pbmc3k           2419          779 Memory CD4 T    
#&gt; 2 AAACATTGAGCTAC  0        0 pbmc3k           4903         1352 B               
#&gt; 3 AAACATTGATCAGC  3.49     0 pbmc3k           3147         1129 Memory CD4 T    
#&gt; 4 AAACCGTGCTTCCG  0        0 pbmc3k           2639          960 CD14+ Mono      
#&gt; 5 AAACCGTGTATGCG  0        0 pbmc3k            980          521 NK              
#&gt; 6 AAACGCACTGGTAC  1.73     0 pbmc3k           2163          781 Memory CD4 T    
#&gt; # … with 2 more variables: RNA_snn_res.0.5 &lt;fct&gt;, seurat_clusters &lt;fct&gt;</code></pre>
<p>Let’s make a scatter plot and check the correlation between CD3D and CD4</p>
<pre class="r"><code># https://github.com/LKremer/ggpointdensity
# ggpubr to add the correlation

ggplot(expression_data, aes(x= CD3D, y = CD4)) + 
        #geom_smooth(method=&quot;lm&quot;) +
        geom_point(size = 0.8) +
        facet_wrap(~seurat_annotations) +
        ggpubr::stat_cor(method = &quot;pearson&quot;)</code></pre>
<p><img src="/post/2023-01-13-how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1.en_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>Let’s focus on the Naive CD4 T cells and Memory CD4 T cells.</p>
<p>Did you also notice the patches of dots on the diagonal lines? Also, there are many 0s in the x-axis and y-axis. This is very typical for single-cell RNAseq data. The data are very spare meaning there are many 0s in the count matrix.</p>
<p>Note, read my <a href="https://divingintogeneticsandgenomics.rbind.io/post/negative-bionomial-distribution-in-single-cell-rnaseq/">previous post</a> on zero-inflation on scRNAseq data.</p>
<p>What’s going on with the cells on diagonal lines?</p>
<pre class="r"><code>expression_data %&gt;%
        filter(seurat_annotations == &quot;Naive CD4 T&quot;, CD3D == CD4, CD3D !=0) </code></pre>
<pre><code>#&gt; # A tibble: 9 × 9
#&gt;   cell            CD3D   CD4 orig.ident nCount_RNA nFeature_RNA seurat_annotati…
#&gt;   &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           
#&gt; 1 CCACCATGATCGGT  1.61  1.61 pbmc3k           2494          732 Naive CD4 T     
#&gt; 2 CGCTAAGACAACTG  1.53  1.53 pbmc3k           2771          858 Naive CD4 T     
#&gt; 3 CTTACAACTAACGC  1.65  1.65 pbmc3k           2368          854 Naive CD4 T     
#&gt; 4 GGGCAGCTTTTCTG  1.67  1.67 pbmc3k           2333          752 Naive CD4 T     
#&gt; 5 GTACCCTGTGAACC  1.61  1.61 pbmc3k           2510          787 Naive CD4 T     
#&gt; 6 GTCACCTGCCTCCA  1.74  1.74 pbmc3k           2126          742 Naive CD4 T     
#&gt; 7 TACTACTGTATGGC  1.51  1.51 pbmc3k           2843          843 Naive CD4 T     
#&gt; 8 TTACACACTCCTAT  1.61  1.61 pbmc3k           2507          815 Naive CD4 T     
#&gt; 9 TTCAACACAACAGA  1.63  1.63 pbmc3k           2429          847 Naive CD4 T     
#&gt; # … with 2 more variables: RNA_snn_res.0.5 &lt;fct&gt;, seurat_clusters &lt;fct&gt;</code></pre>
<p>There are 9 cells on the <code>y=x</code> diagonal line.</p>
<p>Let’s take a look at their raw counts</p>
<pre class="r"><code>cells&lt;- expression_data %&gt;%
        filter(seurat_annotations == &quot;Naive CD4 T&quot;, CD3D == CD4, CD3D !=0) %&gt;%
        pull(cell)

pbmc3k@assays$RNA@counts[genes, cells]</code></pre>
<pre><code>#&gt; 2 x 9 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;      CCACCATGATCGGT CGCTAAGACAACTG CTTACAACTAACGC GGGCAGCTTTTCTG GTACCCTGTGAACC
#&gt; CD3D              1              1              1              1              1
#&gt; CD4               1              1              1              1              1
#&gt;      GTCACCTGCCTCCA TACTACTGTATGGC TTACACACTCCTAT TTCAACACAACAGA
#&gt; CD3D              1              1              1              1
#&gt; CD4               1              1              1              1</code></pre>
<p>They all have counts of 1! the log normalization by library size make them on the <code>y=x</code> line.</p>
<p>Let’s highlight them:</p>
<pre class="r"><code>ggplot(expression_data %&gt;% 
               mutate(highlight = if_else(cell%in% cells, &quot;yes&quot;, &quot;no&quot;)) %&gt;%
               filter(seurat_annotations == &quot;Naive CD4 T&quot;), 
       aes(x= CD3D, y = CD4)) + 
        geom_point(aes(col = highlight), size = 0.8) +
        scale_color_manual(values = c(&quot;black&quot;, &quot;red&quot;))+
        facet_wrap(~seurat_annotations) +
        ggpubr::stat_cor(method = &quot;pearson&quot;)</code></pre>
<p><img src="/post/2023-01-13-how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1.en_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>Let’s look at some other cells on the diagonal line with slop not equal to 1:</p>
<pre class="r"><code>cells2&lt;- expression_data %&gt;% 
        filter(seurat_annotations == &quot;Naive CD4 T&quot;, CD3D !=0, CD4 !=0, CD3D != CD4) %&gt;%
        pull(cell)

pbmc3k@assays$RNA@counts[genes, cells2]</code></pre>
<pre><code>#&gt; 2 x 33 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                                                        
#&gt; CD3D 3 2 2 11 4 3 3 4 2 2 4 3 2 4 2 3 2 2 1 1 2 2 2 3 4 3 2 4 3 4 5 3 4
#&gt; CD4  1 1 1  1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 2 2 1 1 1 1 1 2 1 1 1 1 1 1 1</code></pre>
<p>extract cells with CD4 counts of 1 and CD3D counts of 3:</p>
<pre class="r"><code>other_cells&lt;- pbmc3k@assays$RNA@counts[genes, cells2, drop = FALSE] %&gt;% 
        as.matrix() %&gt;% 
        as.data.frame() %&gt;%
        tibble::rownames_to_column(var= &quot;gene&quot;) %&gt;%
        tidyr::pivot_longer(cols = -1, names_to = &quot;cell&quot;, values_to = &quot;expression&quot;) %&gt;%
        tidyr::pivot_wider(names_from = &quot;gene&quot;, values_from = expression) %&gt;%
        filter(CD3D == 3, CD4 == 1) %&gt;%
        pull(cell)</code></pre>
<p>Let’s see where are they:</p>
<pre class="r"><code>ggplot(expression_data %&gt;% 
               mutate(highlight = if_else(cell%in% other_cells, &quot;yes&quot;, &quot;no&quot;)) %&gt;%
               filter(seurat_annotations == &quot;Naive CD4 T&quot;), 
       aes(x= CD3D, y = CD4)) + 
        geom_point(aes(col = highlight), size = 0.8) +
        scale_color_manual(values = c(&quot;black&quot;, &quot;red&quot;))+
        facet_wrap(~seurat_annotations) +
        ggpubr::stat_cor(method = &quot;pearson&quot;)</code></pre>
<p><img src="/post/2023-01-13-how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1.en_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>So, all those stripes of dots are of the same discrete counts for each gene, respectively. They are on the same line just because the library size (sequencing depth) for each cell is different.</p>
<p>It has to do with the discrete distribution of the counts.</p>
<pre class="r"><code>CD4_naive_cells&lt;- WhichCells(pbmc3k, expression = seurat_annotations == &quot;Naive CD4 T&quot;)

get_expression_data(pbmc3k, slot = &quot;count&quot;, cells= CD4_naive_cells) %&gt;%
        ggplot(aes(x = CD4)) +
        geom_histogram(binwidth = 1) +
        coord_cartesian(xlim = c(0,5))</code></pre>
<p><img src="/post/2023-01-13-how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1.en_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>One way to get around the 0s is to do imputation using tools such as SAVER.</p>
<p>Read <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02132-x">A systematic evaluation of single-cell RNA-sequencing imputation methods</a>.</p>
<blockquote>
<p>We found that the majority of scRNA-seq imputation methods outperformed no imputation in recovering gene expression observed in bulk RNA-seq. <strong>However, the majority of the methods did not improve performance in downstream analyses compared to no imputation, in particular for clustering and trajectory analysis, and thus should be used with caution</strong>. In addition, we found substantial variability in the performance of the methods within each evaluation aspect. Overall, MAGIC, kNN-smoothing, and SAVER were found to outperform the other methods most consistently</p>
</blockquote>
<p>Be careful that you may have false-positives.</p>
</div>
<div id="lets-try-dino" class="section level3">
<h3>Let’s try DINO</h3>
<p><a href="https://academic.oup.com/bioinformatics/article/37/22/4123/6306403"><code>DINO</code></a> is more of a normalization methods rather than an imputation tool. We use it because it was developed by our team member <a href="https://twitter.com/Matthew_N_B">Matthew Bernstein</a>’s old colleague. Matt did a journal club for us.</p>
<blockquote>
<p>Dino utilizes a flexible mixture of Negative Binomials model of gene expression to reconstruct full gene-specific expression distributions which are independent of sequencing depth. By giving exact zeros positive probability, the Negative Binomial components are applicable to shallow sequencing (high proportions of zeros)</p>
</blockquote>
<pre class="r"><code>#BiocManager::install(&quot;Dino&quot;)
library(Dino)

pbmc3k_norm&lt;- Dino(pbmc3k@assays$RNA@counts, nCores = 4)</code></pre>
<pre><code>#&gt; Some genes have expression non-zero expression below  &#39;minNZ&#39; when subsampled
#&gt; and will be normalized via scale-factor</code></pre>
<pre class="r"><code>pbmc3k_dino&lt;- SeuratFromDino(pbmc3k_norm, doNorm = FALSE, doLog = TRUE)

# no need to run the normalization again.
pbmc3k_dino&lt;- pbmc3k_dino %&gt;%
  FindVariableFeatures(selection.method = &quot;vst&quot;, nfeatures = 2000) %&gt;%
  ScaleData() %&gt;%
  RunPCA(verbose = FALSE) %&gt;%
  FindNeighbors(dims = 1:10, verbose = FALSE) %&gt;%
  FindClusters(resolution = 0.5, verbose = FALSE) %&gt;%
  RunUMAP(dims = 1:10, verbose = FALSE)

## borrow the cell annotation
cell_annotations&lt;- pbmc3k@meta.data %&gt;% 
        tibble::rownames_to_column(var = &quot;cell&quot;) %&gt;%
        select(cell, seurat_annotations)

new_meta&lt;- pbmc3k_dino@meta.data %&gt;%
        tibble::rownames_to_column(var = &quot;cell&quot;) %&gt;%
        left_join(cell_annotations)

new_meta&lt;- as.data.frame(new_meta)
rownames(new_meta)&lt;- new_meta$cell
pbmc3k_dino@meta.data&lt;- new_meta %&gt;% select(-cell)
        
Idents(pbmc3k_dino)&lt;- pbmc3k_dino$seurat_annotations

pbmc3k_dino&lt;- pbmc3k_dino[, !is.na(pbmc3k_dino$seurat_annotations)]

DimPlot(pbmc3k_dino, reduction = &quot;umap&quot;, label = TRUE)</code></pre>
<p><img src="/post/2023-01-13-how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1.en_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
</div>
<div id="visualization" class="section level3">
<h3>visualization</h3>
<pre class="r"><code>expression_data2&lt;- get_expression_data(pbmc3k_dino, genes = genes)

head(expression_data2)</code></pre>
<pre><code>#&gt; # A tibble: 6 × 7
#&gt;   cell            CD3D     CD4 orig.ident    RNA_snn_res.0.5 seurat_clusters
#&gt;   &lt;chr&gt;          &lt;dbl&gt;   &lt;dbl&gt; &lt;fct&gt;         &lt;fct&gt;           &lt;fct&gt;          
#&gt; 1 AAACATACAACCAC 1.51  0       SeuratProject 2               2              
#&gt; 2 AAACATTGAGCTAC 0     0.00797 SeuratProject 3               3              
#&gt; 3 AAACATTGATCAGC 2.11  0       SeuratProject 0               0              
#&gt; 4 AAACCGTGCTTCCG 0.105 0.0788  SeuratProject 1               1              
#&gt; 5 AAACCGTGTATGCG 0.473 0.0305  SeuratProject 6               6              
#&gt; 6 AAACGCACTGGTAC 0.781 0.00598 SeuratProject 0               0              
#&gt; # … with 1 more variable: seurat_annotations &lt;fct&gt;</code></pre>
<pre class="r"><code>ggplot(expression_data2, aes(x= CD3D, y = CD4)) + 
        #geom_smooth(method=&quot;lm&quot;) +
        geom_point(size = 0.8) +
        facet_wrap(~seurat_annotations) +
        ggpubr::stat_cor(method = &quot;pearson&quot;)</code></pre>
<p><img src="/post/2023-01-13-how-to-do-gene-correlation-for-single-cell-rnaseq-data-part-1.en_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>It did not help much for CD4’s expression. Many of the cells still have 0 or very low expression of CD4 in CD4 T cells. The strips of dots pattern does disappear.</p>
<p>It is known that CD4 mRNA and protein levels are not always correlated. read this twitter thread <a href="https://twitter.com/tangming2005/status/1501766040686108678" class="uri">https://twitter.com/tangming2005/status/1501766040686108678</a>. Similarly, CD56/NCAM (NK cell marker) mRNA is not expressed highly in NK cells.</p>
<p>It is the same with <a href="https://cite-seq.com/">CITE-seq</a> data:</p>
<blockquote>
<p>All the CITE-seq data I have worked with have the same phenomena. Low CD4 RNA and moderate CD4 ADT signal.</p>
</blockquote>
<p>Note that CITE-seq has its noise too.</p>
<blockquote>
<p>Since a major component of ADT noise in CITE-seq data is due to ambient capture of unbound ADTs nearly ALL of the cells in this experiment are ‘positive’ for CD4 protein.</p>
</blockquote>
<p>read this blog post <a href="https://rpubs.com/MattPM/cd4" class="uri">https://rpubs.com/MattPM/cd4</a> and <a href="https://www.biorxiv.org/content/10.1101/2020.02.24.963603v3">the paper</a> for more details.</p>
<p>In another study <a href="https://www.nature.com/articles/s41592-022-01684-z">Quantification of extracellular proteins, protein complexes and mRNAs in single cells by proximity sequencing</a>:</p>
<p>It shows poor correlation of mRNA and protein for some well-known marker genes:</p>
<p><img src="/img/posts_img/protein_mRNA_cor.jpeg" />
In the coming post, let’s see if <code>DINO</code> helps for other correlated genes.</p>
<p>Until next post, take care of others, take care of yourself!</p>
</div>
<div id="further-reading" class="section level3">
<h3>Further reading</h3>
<ul>
<li><p><a href="http://michaelskinnider.com/files/Nat%20Meth%202019%20-%20Evaluating%20measures%20of%20association%20for%20single-cell%20transcriptomics.pdf">Evaluating measures of association for single-cell transcriptomics</a></p></li>
<li><p>Bulk RNA-seq has many 0s too! read <a href="https://onlinelibrary.wiley.com/doi/10.1002/ggn2.202200024">Transcriptomics for Clinical and Experimental Biology Research: Hang on a Seq</a>.</p></li>
</ul>
</div>
<div id="why-i-love-sharing" class="section level3">
<h3>Why I love sharing</h3>
<p>Because I get to know more from the replies.</p>
<blockquote class="twitter-tweet">
<p lang="en" dir="ltr">
There is the "zero inflated" Kendall tau and Spearman correlation metrics as described in <a href="https://t.co/Guh26SWGET">https://t.co/Guh26SWGET</a> and implemented with sample-specific weights in the scHOT package <a href="https://t.co/BcsHVl5tKk">https://t.co/BcsHVl5tKk</a> see the weightedZIKendall() and weightedZISpearman() functions
</p>
— Shila Ghazanfar (<span class="citation">@shazanfar</span>) <a href="https://twitter.com/shazanfar/status/1615538897722961921?ref_src=twsrc%5Etfw">January 18, 2023</a>
</blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
<p>Let me try the the “zero inflated” Kendall tau and Spearman described in <a href="https://www.sciencedirect.com/science/article/abs/pii/S0167715214003137">Association of zero-inflated continuous variables</a>.</p>
<p>Implemented in <code>weightedZIKendall()</code> and <code>weightedZISpearman()</code> in the <a href="http://www.bioconductor.org/packages/release/bioc/html/scHOT.html">scHOT</a> package</p>
<pre class="r"><code>#BiocManager::install(&quot;scHOT&quot;)
library(scHOT)
CD3D&lt;- expression_data %&gt;% 
        filter(seurat_annotations == &quot;Naive CD4 T&quot;) %&gt;%
        pull(CD3D)

CD4&lt;- expression_data %&gt;% 
        filter(seurat_annotations == &quot;Naive CD4 T&quot;) %&gt;%
        pull(CD4)

scHOT::weightedZIKendall(CD3D, CD4)</code></pre>
<pre><code>#&gt; [1] 0.009568571</code></pre>
<pre class="r"><code>scHOT::weightedSpearman(CD3D, CD4)</code></pre>
<pre><code>#&gt; [1] -0.004406211</code></pre>
<p>The coefficient is very small. I will try it with other highly correlated genes in my next post.</p>
</div>
