---
title: Part 3 Centered log ratio (CLR) normalization for CITE-seq protein count data
author: Ming Tang
date: '2023-12-11'
slug: details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data
categories:
  - CITE-seq
  - bioinformatics
tags:
  - bioinformatics
  - Seurat
header:
  caption: ''
  image: ''
editor_options: 
  chunk_output_type: console
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<p>Following my <a href="https://divingintogeneticsandgenomics.com/post/cite-seq-downstream-analysis-from-alevin-output-to-seurat-visualization/">last blog post</a>, download the CITE-seq protein and RNA count data at <a href="https://osf.io/43v5a/">here</a>.</p>
<pre class="r"><code>library(Seurat)
library(ggplot2)
library(dplyr)

pbmc&lt;- readRDS(&quot;~/blog_data/CITEseq/pbmc1k_adt.rds&quot;)</code></pre>
<div id="how-to-normalize-protein-adt-count-data" class="section level3">
<h3>How to normalize protein ADT count data?</h3>
<p>Seurat uses the centered log ratio (CLR) to normalize protein count data.</p>
<p>In the Seurat github page:</p>
<pre class="r"><code># https://github.com/satijalab/seurat/blob/fc4a4f5203227832477a576bfe01bc6efeb23f51/R/preprocessing.R#L1768-L1769
clr_function &lt;- function(x) {
          return(log1p(x = x / (exp(x = sum(log1p(x = x[x &gt; 0]), na.rm = TRUE) / length(x = x)))))
}</code></pre>
<p>log1p(x) computes log(1+x) accurately also for |x| &lt;&lt; 1.</p>
<p>It addes a pseduo-count in the last step by using <code>log1p</code> and it effectively makes the transformed values &gt;=0.</p>
<p>Let’s see it in action.</p>
<pre class="r"><code>pbmc &lt;- NormalizeData(pbmc, normalization.method = &quot;CLR&quot;, margin = 2, assay = &quot;ADT&quot;)

# Now, we will visualize CD14 levels for RNA and protein By setting the default assay, we can
# visualize one or the other
DefaultAssay(pbmc) &lt;- &quot;ADT&quot;
FeaturePlot(pbmc, &quot;CD19&quot;, cols = c(&quot;lightgrey&quot;, &quot;darkgreen&quot;)) + ggtitle(&quot;CD19 protein&quot;)</code></pre>
<p><img src="/post/2023-12-11-details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data_files/figure-html/unnamed-chunk-4-1.png" width="576" /></p>
<pre class="r"><code>RidgePlot(pbmc, features = c(&quot;CD19&quot;, &quot;CD3&quot;))</code></pre>
<p><img src="/post/2023-12-11-details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data_files/figure-html/unnamed-chunk-4-2.png" width="576" />
We do see the minimal value is 0.</p>
<p>Second, what does<code>margin = 2</code> mean?</p>
<ul>
<li>margin=1 means “perform CLR normalization for each feature independently”</li>
<li>margin=2 means “perform CLR normalization within a cell”</li>
</ul>
<p>Let’s verify it.</p>
<pre class="r"><code># raw counts
adt_count&lt;- pbmc@assays$ADT@counts

# first cell 
adt_count[,1]</code></pre>
<pre><code>#&gt;    CD3    CD4   CD8a   CD14   CD15   CD16   CD56   CD19   CD25 CD45RA CD45RO 
#&gt;      4    483     16   1148     36     12     18      1      4    248    290 
#&gt;   PD-1  TIGIT  CD127  IgG2a   IgG1  IgG2b 
#&gt;     12      4      5      4      6      2</code></pre>
<pre class="r"><code>#normalized data
adt_norm&lt;- pbmc@assays$ADT@data

# first cell 
adt_norm[,1]</code></pre>
<pre><code>#&gt;        CD3        CD4       CD8a       CD14       CD15       CD16       CD56 
#&gt; 0.18254021 3.22611701 0.58836964 4.06860466 1.03046254 0.47049554 0.64247520 
#&gt;       CD19       CD25     CD45RA     CD45RO       PD-1      TIGIT      CD127 
#&gt; 0.04885264 0.18254021 2.59646813 2.74206669 0.47049554 0.18254021 0.22340593 
#&gt;      IgG2a       IgG1      IgG2b 
#&gt; 0.18254021 0.26266700 0.09542945</code></pre>
<pre class="r"><code>all.equal(clr_function(adt_count[,1]), adt_norm[,1])</code></pre>
<pre><code>#&gt; [1] TRUE</code></pre>
<p>The result shows that when specifying “margin = 2”, seurat is normalizing all the features within a cell.</p>
<p>However, the help page of the normalization is a little bit confusing.</p>
<blockquote>
<p>margin
If performing CLR normalization, normalize across features (1) or cells (2)</p>
</blockquote>
<p>and when you invoke the function, if you specify marigin =1
it will message: “Normalizing across features”; if you specify margin = 2, it will
message “Normalizing across cells”.</p>
<p>This is oppostite to what seurat is doing internally. Read this issue and someone else had the same intepretation <a href="https://github.com/satijalab/seurat/issues/3605" class="uri">https://github.com/satijalab/seurat/issues/3605</a>.</p>
</div>
<div id="clr-normalization-in-detail" class="section level3">
<h3>CLR normalization in detail</h3>
<p>Now, let’s take a closer look at centered log-ratio (CLR) normalization. CLR was initially used in this paper <a href="https://sci-hub.ru/https://www.nature.com/articles/nmeth.4380">Simultaneous epitope and transcriptome measurement in single cells</a> on compositional data.</p>
<p><img src="/img/clr_norm.png" /></p>
<p>The first function below is what the orginal paper uses.</p>
<pre class="r"><code>## the usual CLR adding pseduo-count for each protein first
clr_original &lt;- function(x) {
  return(log((x + 1) / mean(x + 1, na.rm = TRUE)))
}


## clr from seurat 
clr_function &lt;- function(x) {
  return(log1p(x = x / (exp(x = sum(log1p(x = x[x &gt; 0]), na.rm = TRUE) / length(x = x)))))
}</code></pre>
<p>Seurat CLR removes 0 counts first by <code>x[x&gt;0]</code> and then <code>log1p</code> transform the raw counts, sum them up, calculate the average of the log counts, <code>exp</code> it back, and then divided the raw counts by this average and finally <code>log1p</code> again…</p>
<p>Seurat CLR normalization.</p>
<pre class="r"><code>clr_norm&lt;- apply(adt_count, MARGIN = 2,  FUN= clr_function)

# indeed, it is the same with the normalized ADT data in the seurat object
all.equal(clr_norm, pbmc@assays$ADT@data)</code></pre>
<pre><code>#&gt; [1] TRUE</code></pre>
<pre class="r"><code>range(clr_norm)</code></pre>
<pre><code>#&gt; [1] 0.000000 5.754108</code></pre>
<pre class="r"><code>hist(clr_norm)</code></pre>
<p><img src="/post/2023-12-11-details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data_files/figure-html/unnamed-chunk-7-1.png" width="576" />
We do not see negative values.</p>
<p>The original CLR</p>
<pre class="r"><code>clr_original_norm&lt;- apply(adt_count, MARGIN = 2, clr_original)

range(clr_original_norm)</code></pre>
<pre><code>#&gt; [1] -6.363028  2.786000</code></pre>
<pre class="r"><code>hist(clr_original_norm)</code></pre>
<p><img src="/post/2023-12-11-details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data_files/figure-html/unnamed-chunk-8-1.png" width="576" />
We do see negative values.</p>
<p>Visualize in Ridge plots:</p>
<pre class="r"><code>get_expression_meta&lt;- function(obj, mat){
  meta&lt;- obj@meta.data %&gt;%
  tibble::rownames_to_column(var = &quot;cell&quot;)
  
  df&lt;- t(mat) %&gt;%
    as.data.frame() %&gt;% 
    tibble::rownames_to_column(var = &quot;cell&quot;) %&gt;%
    left_join(meta)
  
  return(df)
}


p1&lt;- get_expression_meta(pbmc, clr_norm ) %&gt;% 
  ggplot(aes(x = CD19, y = celltypes)) +
  ggridges::geom_density_ridges(aes(fill = celltypes)) + 
  theme_bw(base_size = 14) +
  xlab(&quot;&quot;) +
  ylab(&quot;&quot;) +
  ggtitle(&quot;Seurat CLR&quot;)


p2&lt;- get_expression_meta(pbmc, clr_original_norm ) %&gt;% 
  ggplot(aes(x = CD19, y = celltypes)) +
  ggridges::geom_density_ridges(aes(fill = celltypes)) + 
  theme_bw(base_size = 14) +
  xlab(&quot;&quot;) +
  ylab(&quot;&quot;) +
  ggtitle(&quot;original CLR&quot;)

p1 + p2 + patchwork::plot_layout(guides = &quot;collect&quot;)</code></pre>
<p><img src="/post/2023-12-11-details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data_files/figure-html/unnamed-chunk-9-1.png" width="960" /></p>
</div>
<div id="visualize-in-heatmap" class="section level3">
<h3>visualize in Heatmap</h3>
<pre class="r"><code>library(ComplexHeatmap)
hp1&lt;- Heatmap(clr_norm, 
              show_column_names = FALSE,
              name = &quot;Seurat CLR&quot;)

row_index&lt;- row_order(hp1)
col_index&lt;- column_order(hp1)


hp2&lt;- Heatmap(clr_original_norm, 
              show_column_names = FALSE,
              row_order = row_index,
              column_order = col_index,
              name = &quot;original CLR&quot;)
hp1 + hp2 </code></pre>
<p><img src="/post/2023-12-11-details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data_files/figure-html/unnamed-chunk-10-1.png" width="576" /></p>
<p>They both look similar in terms of patterns. The Seurat CLR makes it 0 bound and the authors argue it is better for visualization.</p>
</div>
<div id="take-home-messages" class="section level3">
<h3>Take home messages</h3>
<ol style="list-style-type: decimal">
<li><p>Details matter. If you blindly uses the functions in a package, you may get the wrong interpretation of your results.</p></li>
<li><p>Documentation matters. If the documentation is confusing, it may lead to unintended usages of the options.</p></li>
</ol>
</div>
<div id="references" class="section level3">
<h3>References</h3>
<p><a href="https://github.com/satijalab/seurat/issues/2624" class="uri">https://github.com/satijalab/seurat/issues/2624</a>
<a href="https://github.com/satijalab/seurat/issues/1268" class="uri">https://github.com/satijalab/seurat/issues/1268</a>
<a href="https://github.com/scverse/muon/pull/28" class="uri">https://github.com/scverse/muon/pull/28</a></p>
</div>
