---
title: neighborhood/cellular niches analysis with spatial transcriptome data in Seurat
  and Bioconductor
author: Ming Tang
date: '2023-08-08'
slug: neighborhood-cellular-niches-analysis-with-spatial-transcriptome-data-in-seurat-and-bioconductor
categories:
  - bioinformatics
  - scRNAseq
  - spatial
tags:
  - bioinformatics
  - scRNAseq
  - spatial
header:
  caption: ''
  image: ''
---

---
title: "Untitled"
author: "Ming Tang"
date: '2023-08-07'
output: html_document
---


```{r, include=FALSE}
# https://bookdown.org/yihui/rmarkdown-cookbook/cache-lazy.html
knitr::opts_chunk$set(
  comment = "#>", echo = TRUE, message= FALSE, warning = FALSE,
  cache = TRUE, cache.lazy= FALSE
)
knitr::opts_chunk$set(fig.width = 12, fig.height = 8)  # Set the desired width and height

```


### Spatial transcriptome cellular niche analysis using 10x xenium data 

go to https://www.10xgenomics.com/resources/datasets

There is a lung cancer and a breast cancer dataset. Let's work on the lung cancer one.

https://www.10xgenomics.com/resources/datasets/xenium-human-lung-preview-data-1-standard

37G zipped file!

```{bash eval=FALSE}
wget https://s3-us-west-2.amazonaws.com/10x.files/samples/xenium/1.3.0/Xenium_Preview_Human_Lung_Cancer_With_Add_on_2_FFPE/Xenium_Preview_Human_Lung_Cancer_With_Add_on_2_FFPE_outs.zip

unzip Xenium_Preview_Human_Lung_Cancer_With_Add_on_2_FFPE_outs.zip 

sudo tar xvzf cell_feature_matrix.tar.gz
```

`cell_feature_matrix/` 
`cell_feature_matrix/barcodes.tsv.gz`
`cell_feature_matrix/features.tsv.gz`
`cell_feature_matrix/matrix.mtx.gz`

### read in the data with Seurat 

We really only care about the cell by gene count matrix which is inside the `cell_feature_matrix` folder, 
and the cell location x,y coordinates: `cells.csv.gz`.

The `data/xenium` folder below contains the `cells.csv.gz` file and the `cell_feature_matrix` folder.

The transcripts.csv.gz contains the molecule location information which is not needed for 
most of the analysis if you do not work on subcelluar locations of the RNAs.

```{r}
library(Seurat)
library(dplyr)
library(here)
library(tictoc)

xenium_input<- ReadXenium(
  data.dir = "~/github_repos/compbio_tutorials/data/xenium",
  outs = "matrix",
  type = "centroids",
  mols.qv.threshold = 20
)
```

construct a seurat object.
https://github.com/satijalab/seurat/blob/763259d05991d40721dee99c9919ec6d4491d15e/R/convenience.R#L186

```{r}
segmentations.data <- list(
    "centroids" = CreateCentroids(xenium_input$centroids)
    #"segmentation" = CreateSegmentation(data$segmentations)
  )

coords <- CreateFOV(
    coords = segmentations.data,
    type = "centroids",
    molecules = NULL,
    assay = "Xenium"
  )

xenium.obj <- CreateSeuratObject(counts = xenium_input$matrix[["Gene Expression"]], assay = "Xenium")
xenium.obj[["NegativeControlCodeword"]] <- CreateAssayObject(counts = xenium_input$matrix[["Negative Control Codeword"]])

xenium.obj[["NegativeControlProbe"]] <- CreateAssayObject(counts = xenium_input$matrix[["Negative Control Probe"]])

xenium.obj[["fov"]] <- coords
```

filter the cells

```{r}
# remove cells with 0 counts
xenium.obj <- subset(xenium.obj, subset = nCount_Xenium > 0)

VlnPlot(xenium.obj, features = c("nFeature_Xenium", "nCount_Xenium"), ncol = 2, pt.size = 0)
```


```{r}
median(xenium.obj$nCount_Xenium)
median(xenium.obj$nFeature_Xenium)
```
median number of detected genes is around 62, and the median counts per cell is 105.


```{r}
ImageFeaturePlot(xenium.obj, features = c("EPCAM", "CD4", "CD8A", "CD79A"), max.cutoff = c(20,
    15, 6, 6), size = 0.75, cols = c("white", "red"))
```

`EPCAM` mark the  epithelial/tumor cells, and we also see some CD4, CD8 T cells and B cells.


standard processing 

This takes 30mins
```{r}
tic()
xenium.obj <- SCTransform(xenium.obj, assay = "Xenium")
xenium.obj <- RunPCA(xenium.obj, npcs = 30, features = rownames(xenium.obj))
xenium.obj <- RunUMAP(xenium.obj, dims = 1:30)
xenium.obj <- FindNeighbors(xenium.obj, reduction = "pca", dims = 1:30)
xenium.obj <- FindClusters(xenium.obj, resolution = 0.3)
toc()
```

```{r plot}
DimPlot(xenium.obj)

FeaturePlot(xenium.obj, features = c("CD4", "CD8A", "EPCAM", "NKG7", "GZMB",
                                     "FOXP3"))

ImageDimPlot(xenium.obj, cols = "polychrome", size = 0.75)
```

### Find all markers of the clusters and annotate the clusters with marker genes

```{r}
# devtools::install_github("immunogenomics/presto")
library(presto)

all_markers<- presto::wilcoxauc(xenium.obj, assay = "data",
                                seurat_assay = "Xenium")

all_markers %>% head()

# presto is much faster than FindAllMarkers
presto::top_markers(all_markers, n = 10, auc_min = 0.5, pct_in_min = 20)
```
Based on the marker genes, I manually annotate the cell types.

* cluster 0: fibroblast. 
* cluster 1: M2-like macrophage. 
* cluster 10: ciliated epithelial cells. 
* cluster 11: M2-like macrophage. 
* cluster 12: M2-like macrophage. 
* cluster 13: endothelia-1. 
* cluster 14: mast cell. 
* cluster 15: B cells. 
* cluster 2: alveolar epithelial cells. 
* cluster 3: endothelial-2. 
* cluster 4: T/NK cells. 
* cluster 5: alveolar epithelial cells. 
* cluster 6: smooth muscle cells. 
* cluster 7: M2-like macrophage. 
* cluster 8: M1-like macrophage. 
* cluster 9: plasma cells. 

Add the cell annotations 
```{r}
annotations<- c("fibroblast", "M2-like macrophage-1", "ciliated epithelial cells", "M2-like macrophage-2", "M2-like macrophage-3", "endothelial-1", "mast cell", "B cells-1", "alveolar epithelial cells-1", "endothelial-2", "T/NK cells", "alveolar epithelial cells-2", "smooth muscle cells", "M2-like macrophage-4", "M1-like macrophage", "plasma cells")

names(annotations)<- c(0,1,10,11,12,13,14,15,2,3,4,5,6,7,8,9)

xenium.obj$annotations<- annotations[as.character(xenium.obj$seurat_clusters)]
```

cluster 4 is the T/NK cells, I want to subcluster it to more fine-grained clusters
```{r}
table(xenium.obj$seurat_clusters)
xenium.obj<- FindSubCluster(xenium.obj, cluster = 4, graph.name = "SCT_snn",
                            resolution = 0.2)

table(xenium.obj$sub.cluster)

Idents(xenium.obj)<- xenium.obj$sub.cluster
scCustomize::Clustered_DotPlot(xenium.obj[, xenium.obj$sub.cluster %>%
                                            stringr::str_detect("^4_")], 
                               features = c("CD4", "CD8A", "NKG7", "GZMB", 
                               "PDCD1", "FOXP3"), plot_km_elbow = FALSE)

all_markers2<- presto::wilcoxauc(xenium.obj[,xenium.obj$sub.cluster %>%
                                            stringr::str_detect("^4_")],
                                 assay = "data", seurat_assay = "Xenium")

presto::top_markers(all_markers2, n = 20, auc_min = 0.5, pct_in_min = 20)
```

more fine-grained annotation 

* 4_0: CD4 Tcells. 
* 4_1: NK cells. 
* 4_2: fibroblast. 
* 4_3: CD8 Tcells. 
* 4_5: PRG4 is a fibroblast marker we also see CD4 in this cluster.  
* 4_6: macrophage. 

Interesting, subcluster the NK/T cells revealed fibroblasts and macrophage (4_6) populations.
My experience with spatial data from 10x, nanostring and Vizgen are similar, they all have spill out transcripts of cells that maybe contaminating the nearby cells. Also, there is no clear cut of CD4 T cells and CD8 T cells.

### neighorhoods analysis in Seurat 

Cellular neighborhood/niche is aggregation of cells can result in ‘cellular neighbourhoods’. A neighbourhood is defined as a group of cells that with a similar composition together. These can be homotypic, containing cells of a single class (e.g. immune cells), or heterotypic (e.g. a mixture of tumour and immune cells).

In one of my [previous post](https://divingintogeneticsandgenomics.com/post/how-to-do-neighborhood-cellular-niches-analysis-with-spatial-transcriptome-data/), I showed you how to identify cellular neighborhood/niches within seurat.

Let's repeat it.
```{r}
xenium_input$centroids %>% head()
mat<- xenium_input$centroids[,1:2] 
mat<- as.matrix(mat)
rownames(mat)<- xenium_input$centroids$cell
```

Reorder the cells in the coordinates matrix as the same order as in the Seurat object.

```{r}
Cells(xenium.obj) %>% head()
## make sure the order of the cells is the same 
all.equal(Cells(xenium.obj), rownames(mat))

head(mat)

# reorder the matrix rows
mat<- mat[Cells(xenium.obj), ]
head(mat)

all.equal(Cells(xenium.obj), rownames(mat))
```

Find the closest cells for each cell within 50 um radius.

```{r}
library(dbscan)
eps <- 50
tic()
nn <- frNN(x= mat, eps = eps)
toc()

nn$id %>%
  head(n=3)
```
It is a list of list containing the indices of the cells that are within 50um radius. 

Create the neighborhood count matrix.
```{r}
# the same cell order
all.equal(colnames(xenium.obj), names(nn$id))

nn$id %>%
  stack() %>%
  head()

## make it to a dataframe 
nn_df<- nn$id %>%
  stack()
```

get the annotations for cells that are within the 50 um of each cell.

```{r}
cluster_ids<- xenium.obj$annotations %>% unname()

## vectorization, so much faster than the solution in my old post
nn_df$cluster_id<- cluster_ids[nn_df$values]

head(nn_df)
nn_df$cluster_id<- factor(nn_df$cluster_id)
tic()
nn_count<- nn_df %>%
  group_by(ind) %>%
  count(cluster_id, .drop = FALSE)
toc()

head(nn_count)
```

pivot it to a wide format and create a `cell x cluster_id` matrix 

```{r}
nn_count<- nn_count %>%
  tidyr::pivot_wider(names_from = cluster_id, values_from = n)

head(nn_count)

# convert to a matrix
nn_mat<- nn_count[,-1] %>% as.matrix()
rownames(nn_mat)<- nn_count$ind

head(nn_mat)

all.equal(rownames(nn_mat), Cells(xenium.obj))
```

#### k-means clustering 

```{r}
set.seed(123)

# I cheated here to make k=13 so I can compare it with the Louvian clustering later
k_means_res<- kmeans(nn_mat, centers = 13)

k_means_id<- k_means_res$cluster %>%
  tibble::enframe(name = "cell_id", value = "kmeans_cluster")

head(k_means_id)

# add it to the metadata solot for nn_obj below
k_means_df<- as.data.frame(k_means_id)
rownames(k_means_df)<- k_means_id$cell_id
```

#### Louvian clustering just like scRANseq data 

Create a Seurat object and do a regular single-cell count matrix analysis, but now we only have 16 features (clusters) instead of 20,000 genes.

```{r}
nn_obj<- CreateSeuratObject(counts = t(nn_mat),  min.features = 5)
```

routine processing

```{r}
nn_obj<- SCTransform(nn_obj, vst.flavor = "v2")

## we only have 16 clusters/celltypes, max pc 16
nn_obj <- RunPCA(nn_obj, npcs = 15, features = rownames(nn_obj))
ElbowPlot(nn_obj)

nn_obj <- FindNeighbors(nn_obj, reduction = "pca", dims = 1:15)
nn_obj <- FindClusters(nn_obj, resolution = 0.3)
nn_obj <- RunUMAP(nn_obj, dims = 1:10)


DimPlot(nn_obj)
```

#### Suerat V5 has a function to detect neighborhood

https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2

I do not have Seurat V5 installed, so the code below is not run.
```{r eval=FALSE}
xenium.obj <- BuildNicheAssay(object = xenium.obj, group.by = "annotations",
    niches.k = 5, neighbors.k = 30)
```

#### compare k-means and the Louvian clustering 

```{r}
nn_obj<- AddMetaData(nn_obj, metadata = k_means_df )

nn_obj@meta.data %>%
  head()

table(nn_obj$kmeans_cluster, nn_obj$seurat_clusters)
```

Some help functions from `scclusteval` to calculate the jaccard distances for two difference clustering methods:

```{r}
#devtools::install_github("crazyhottommy/scclusteval")

library(scclusteval)
PairWiseJaccardSetsHeatmap(Idents(nn_obj), k_means_res$cluster, best_match = TRUE)
```
x-axis is the k-means clustering: cluster1-13.
y-axis is the Louvian clustering: cluster0-12.

There are some 1-1 cluster mapping for the two different clustering methods.

Add the niche information to the original seurat object.
```{r}
nn_meta<- nn_obj@meta.data %>%
  select(cell_id, SCT_snn_res.0.3, kmeans_cluster) 

xenium.obj<- AddMetaData(xenium.obj, nn_meta)
```

characterize the neighborhoods/niches by looking at the composition of cell types 

```{r}
library(ComplexHeatmap)

# calculate the average abundance of cell types per niche
average_cell_type_abundance<- AverageExpression(
  nn_obj,
  assays = NULL,
  features = rownames(nn_obj),
  return.seurat = FALSE,
  group.by = "SCT_snn_res.0.3",
)

cell_fun = function(j, i, x, y, width, height, fill) {
                grid::grid.rect(x = x, y = y, width = width *0.99, 
                                height = height *0.99,
                                gp = grid::gpar(col = "grey", 
                                                fill = fill, lty = 1, lwd = 0.5))
}

col_fun=circlize::colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))


Heatmap(t(scale(t(average_cell_type_abundance$SCT))),
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        rect_gp = grid::gpar(type = "none"),
        cell_fun = cell_fun,
        col = col_fun,
        column_names_rot = 45)
```

T/NK cells are mainly in niche 8 and 12.

```{r}
scCustomize::Clustered_DotPlot(nn_obj, features= rownames(nn_obj), plot_km_elbow = FALSE)

scCustomize::Stacked_VlnPlot(nn_obj, features= rownames(nn_obj))
```

Niche 1 contains most of the endothelia-1 cells and looks likely there is a blood vessel there.
```{r}
p1<- ImageDimPlot(xenium.obj, fov = "fov", cols = "polychrome", axes = TRUE, 
                  group.by = "annotations")


## the cells are colored by the clustering of the cells by neighborhood 
p2<- ImageDimPlot(xenium.obj, fov = "fov", cols = "polychrome", axes = TRUE, 
                  group.by = "SCT_snn_res.0.3" )
        
p1 + p2
```

or table it using the tally counts

```{r}

mat2<- table(xenium.obj$SCT_snn_res.0.3, xenium.obj$annotations) 

head(mat2)


Heatmap(t(scale(mat2)),
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        rect_gp = grid::gpar(type = "none"),
        cell_fun = cell_fun,
        col = col_fun,
        column_names_rot = 45)


mat3<- table(xenium.obj$kmeans_cluster, xenium.obj$annotations)
Heatmap(t(scale(mat3)),
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        rect_gp = grid::gpar(type = "none"),
        cell_fun = cell_fun)
```

Tally the counts may obscure the fine structures of the niches. For example, 
it only shows T/NK cells in niche 8 but not so clear in niche 12.

### neigbhorhood analysis in SpatialExperiment

The [SPIAT bioconductor package](https://trigosteam.github.io/SPIAT/articles/neighborhood.html) provides functions to do it and more functions to analyze the spatial data.
But it only works with `SpatialExperiment` object. [Paper](https://www.nature.com/articles/s41467-023-37822-0) describing it.

Convert the seurat object to [SpatialExperiment](https://lmweber.org/OSTA-book/spatialexperiment.html) object.

https://github.com/drighelli/SpatialExperiment/issues/115

```{r eval=TRUE}
library(SpatialExperiment)
library(Seurat)
# BiocManager::install("SpatialExperiment")
library(dplyr)
# BiocManager::install("SPIAT")
library(SPIAT)


## Function
seurat_to_spe <- function(seu, sample_id, img_id) {
    ## Convert to SCE
    sce <- Seurat::as.SingleCellExperiment(seu)
    
    ## Extract spatial coordinates
    spatialCoords <- as.matrix(
        seu@images[[img_id]]@boundaries$centroids@coords[, c("x", "y")])
    
    # the column names are specific for SPAIT
    colnames(spatialCoords)<- c("Cell.X.Position", "Cell.Y.Position")
    
    ## Extract and process image data
    #img <- SpatialExperiment::SpatialImage(
    #    x = as.raster(seu@images[[img_id]]@image))
    
    #imgData <- DataFrame(
    #    sample_id = sample_id,
    #    image_id = img_id,
    #    data = I(list(img)),
    #    scaleFactor = seu@images[[img_id]]@scale.factors$lowres)
    
    # Convert to SpatialExperiment
    spe <- SpatialExperiment(
        assays = assays(sce),
        rowData = rowData(sce),
        colData = colData(sce),
        metadata = metadata(sce),
        reducedDims = reducedDims(sce),
        altExps = altExps(sce),
        sample_id = sample_id,
        spatialCoords = spatialCoords,
        #imgData = imgData
    )
    # indicate all spots are on the tissue
    spe$in_tissue <- 1
    spe$sample_id <- sample_id
    # Return Spatial Experiment object
    spe
}


spe<- seurat_to_spe(seu = xenium.obj, sample_id = "1", img_id = 1)

```

inspect the data 
```{r eval=TRUE}
# what's the gene and cell number
dim(spe)

assay(spe)[1:5, 1:5]

colData(spe)
spe$seurat_clusters %>% head()

spatialCoords(spe)[1:5, ]
```


```{r eval=TRUE}
average_minimum_distance(spe)
```


```{r eval=TRUE}
tic()
spe <- identify_neighborhoods(
  spe, method = "dbscan", min_neighborhood_size = 100,
  cell_types_of_interest = annotations, radius = 50, 
  feature_colname = "annotations")

toc()
```

It adds a new column of the niche ID in the `colData`:
```{r eval=TRUE}
colData(spe) %>% head(n = 3)
```

plot the cell type composition:
```{r eval=TRUE}
neighorhoods_vis <- 
  composition_of_neighborhoods(spe, feature_colname = "annotations")

neighorhoods_vis <- 
  neighorhoods_vis[neighorhoods_vis$Total_number_of_cells >=5,]

plot_composition_heatmap(neighorhoods_vis,
                         feature_colname="annotations")
```

There is no parameters exposed to fine-tune the number of neighborhoods with `dbscan` in the `identify_neighborhoods` function.

There are many other functions you can use from the `SPAIT` package. Make sure you go through the tutorial.

That's it for today! Happy Learning.

