---
title: 'How to preprocess GEO bulk RNAseq data with salmon '
author: Ming Tommy Tang
date: '2024-03-13'
slug: how-to-preprocess-geo-bulk-rnaseq-data-with-salmon
categories:
  - bioinformatics
  - RNAseq
tags:
  - bioinformatics
  - RNAseq
header:
  caption: ''
  image: ''
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<link href="/rmarkdown-libs/vembedr/css/vembedr.css" rel="stylesheet" />


<div id="install-fastq-dl" class="section level3">
<h3>Install fastq-dl</h3>
<p>To easily download fastq from GEO or ENA, use <a href="https://github.com/rpetit3/fastq-dl">fastq-dl</a></p>
<p>Assume you already have <a href="https://conda.io/projects/conda/en/latest/user-guide/install/index.html"><code>conda</code></a> installed, do the following:</p>
<pre class="bash"><code>conda config --add channels conda-forge
conda config --add channels bioconda
conda create -n fastq_download -c conda-forge -c bioconda fastq-dl
conda activate fastq_download </code></pre>
<p><strong>Tip</strong>: use <a href="https://github.com/mamba-org/mamba"><code>mamba</code></a> if <code>conda</code> is too slow for you.
They are all big snakes!!</p>
<p>We will use bulk RNAseq data from this GEO accession ID:
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576</a></p>
<pre class="bash"><code>mkdir hypoxia_RNAseq
cd hypoxia_RNAseq

mkdir data
cd data
## some times the same experiment may have multiple runs (SRR)
## use the --group-by-experiment to merge them, this example only has one SRR
## per sample   

## normoxia sgCTRL
fastq-dl --accession SRX14311105  --group-by-experiment
fastq-dl --accession SRX14311106  --group-by-experiment

## hypoxia sgCTRL
fastq-dl --accession SRX14311111  --group-by-experiment
fastq-dl --accession SRX14311112  --group-by-experiment</code></pre>
<p>We only have 4 samples here and we typed in the command one by one which is fine.
What if we want to process all of them and you do not want to do it manually.</p>
<p>Get the metadata using <a href="https://github.com/NCBI-Hackathons/ncbi-cloud-tutorials/blob/master/SRA%20tutorials/tutorial_SRA_run_selector.md">SRA Run selector</a> or use <a href="https://www.youtube.com/watch?v=NSoU58IqA3Y">GEOquery</a> or <a href="https://www.bioconductor.org/packages/release/bioc/html/GEOmetadb.html">GEOmetadb</a>.</p>
<p>You may want to do a <code>fastqc</code> for quality control of the reads and trimming with
<code>fastp</code> for the sequencing adapters. We will skip it in this tutorial.</p>
</div>
<div id="install-salmon" class="section level3">
<h3>install salmon</h3>
<pre class="bash"><code>conda create -n salmon salmon</code></pre>
</div>
<div id="build-an-index" class="section level3">
<h3>build an index</h3>
<p>We need a transcriptome fasta file to build the index, we also need the gtf file that we can
use to map gene ids.</p>
<p>Let’s go to gencode <a href="https://www.gencodegenes.org/human/" class="uri">https://www.gencodegenes.org/human/</a></p>
<p>There are multiple sources to get the reference files. If you are confused,
watch this video.</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/eIVlSG11umQ" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
<p>Download the reference fasta:</p>
<pre class="bash"><code>cd hypoxia_RNAseq
mkdir reference
cd reference

wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_45/gencode.v45.annotation.gtf.gz


wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_45/gencode.v45.basic.annotation.gtf.gz</code></pre>
<p>A quick look at the gtf file:</p>
<pre class="bash"><code>zless -S gencode.v45.annotation.gtf.gz | grep -v &quot;#&quot; | \
awk &#39;$3==&quot;gene&quot;&#39; | cut -f9 | head -3

gene_id &quot;ENSG00000290825.1&quot;; gene_type &quot;lncRNA&quot;; gene_name &quot;DDX11L2&quot;; level 2; tag &quot;overlaps_pseudogene&quot;;
gene_id &quot;ENSG00000223972.6&quot;; gene_type &quot;transcribed_unprocessed_pseudogene&quot;; gene_name &quot;DDX11L1&quot;; level 2; hgnc_id &quot;HGNC:37102&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;;
gene_id &quot;ENSG00000227232.6&quot;; gene_type &quot;unprocessed_pseudogene&quot;; gene_name &quot;WASH7P&quot;; level 2; hgnc_id &quot;HGNC:38034&quot;; havana_gene &quot;OTTHUMG00000000958.1&quot;;</code></pre>
<p>Let’s tally the gene_type</p>
<pre class="bash"><code>zless -S gencode.v45.annotation.gtf.gz | grep -v &quot;#&quot; | \
awk &#39;$3==&quot;gene&quot;&#39; | cut -f9 | \
cut -f2 -d &quot;;&quot; | sort | uniq -c | sort -k1,1nr

  20049  gene_type &quot;protein_coding&quot;
  19370  gene_type &quot;lncRNA&quot;
  10143  gene_type &quot;processed_pseudogene&quot;
   2604  gene_type &quot;unprocessed_pseudogene&quot;
   2208  gene_type &quot;misc_RNA&quot;
   1901  gene_type &quot;snRNA&quot;
   1879  gene_type &quot;miRNA&quot;
   1054  gene_type &quot;TEC&quot;
    962  gene_type &quot;transcribed_unprocessed_pseudogene&quot;
    942  gene_type &quot;snoRNA&quot;
    513  gene_type &quot;transcribed_processed_pseudogene&quot;
    497  gene_type &quot;rRNA_pseudogene&quot;
    187  gene_type &quot;IG_V_pseudogene&quot;
    158  gene_type &quot;transcribed_unitary_pseudogene&quot;
    145  gene_type &quot;IG_V_gene&quot;
    107  gene_type &quot;TR_V_gene&quot;
    100  gene_type &quot;unitary_pseudogene&quot;
     79  gene_type &quot;TR_J_gene&quot;
     49  gene_type &quot;scaRNA&quot;
     47  gene_type &quot;rRNA&quot;
     37  gene_type &quot;IG_D_gene&quot;
     33  gene_type &quot;TR_V_pseudogene&quot;
     22  gene_type &quot;Mt_tRNA&quot;
     19  gene_type &quot;artifact&quot;
     18  gene_type &quot;IG_J_gene&quot;
     14  gene_type &quot;IG_C_gene&quot;
      9  gene_type &quot;IG_C_pseudogene&quot;
      8  gene_type &quot;ribozyme&quot;
      6  gene_type &quot;TR_C_gene&quot;
      5  gene_type &quot;TR_D_gene&quot;
      5  gene_type &quot;sRNA&quot;
      4  gene_type &quot;TR_J_pseudogene&quot;
      4  gene_type &quot;vault_RNA&quot;
      3  gene_type &quot;IG_J_pseudogene&quot;
      2  gene_type &quot;Mt_rRNA&quot;
      2  gene_type &quot;translated_processed_pseudogene&quot;
      1  gene_type &quot;IG_pseudogene&quot;
      1  gene_type &quot;scRNA&quot;</code></pre>
<p>We will use the <code>gtf</code> file for downstream analysis.</p>
<p>Download the transcript fasta file</p>
<pre class="bash"><code>wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_45/gencode.v45.transcripts.fa.gz
</code></pre>
<p>index the transcriptome</p>
<pre class="bash"><code># list all the conda env
conda env list

# activate the salmon env
conda activate salmon

salmon index -t gencode.v45.transcripts.fa.gz -i gencode.v45_human_index -k 31 --gencode</code></pre>
</div>
<div id="quantification" class="section level3">
<h3>quantification</h3>
<p>This is a single end read dataset, so I used <code>-r</code> to feed in the fastq. If you
have paired-end reads, you need to use <code>-1</code> and <code>-2</code> for the <code>R1</code> and <code>R2</code> reads.</p>
<p>There are also different library types for paired end reads. read <a href="https://salmon.readthedocs.io/en/latest/library_type.html">the documentation</a> for details.</p>
<p><img src="/img/salmon_lib.png" /></p>
<p>Make sure you know how the library is constructed. I do not know if this library
is stranded or unstranded, so I used <code>-l A</code> to automatically let
salmon to detect the library type. salmon detected it as <code>U</code> unstranded.</p>
<pre class="bash"><code>cd ..

salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311105.fastq.gz -o SRX14311105_quant  --validateMappings</code></pre>
<p>This finished in a couple of mins, so fast!</p>
<pre class="bash"><code>cat SRX14311105_quant/logs/salmon_quant.log | grep &quot;Mapping rate&quot;
[jointLog] [info] Mapping rate = 93.0329%</code></pre>
<p>93% mapping rate!</p>
<p>Let’s do the same for the other 3 samples</p>
<pre class="bash"><code>salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311106.fastq.gz -o SRX14311106_quant  --validateMappings

salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311111.fastq.gz -o SRX14311111_quant  --validateMappings

salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311112.fastq.gz -o SRX14311112_quant  --validateMappings</code></pre>
<p>Now we have the salmon quantification output:</p>
<pre class="bash"><code> find . -name &quot;*sf&quot;
./SRX14311111_quant/quant.sf
./SRX14311105_quant/quant.sf
./SRX14311106_quant/quant.sf
./SRX14311112_quant/quant.sf</code></pre>
<p>and the mapping rates are all pretty high:</p>
<pre class="bash"><code>find . -name &quot;salmon_quant.log&quot; | xargs grep &quot;Mapping rate&quot;
./SRX14311111_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 91.5163%
./SRX14311105_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 93.0329%
./SRX14311106_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 92.8808%
./SRX14311112_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 91.4357%</code></pre>
<p>we can load it into R and use <code>DESeq2</code>for downstream analysis. That’s for another post!
Happy Learning!</p>
<p>Watch the video here:</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/_Q8fYokTCTs" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
