---
title: 'How to separate a comma delimited string into multiple lines in R and python '
author: Ming Tommy Tang
date: '2024-05-09'
slug: how-to-separate-a-comma-delimited-string-into-multiple-lines-in-r-and-python
categories:
  - bioinformatics
  - rstats
  - python
tags:
  - python
  - rstats
header:
  caption: ''
  image: ''
editor_options: 
  chunk_output_type: console
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<p><strong>To not miss a post like this, sign up for my <a href="https://divingintogeneticsandgenomics.ck.page/profile">newsletter</a> to learn computational
biology and bioinformatics.</strong></p>
<div id="the-problem" class="section level3">
<h3>The problem</h3>
<pre class="r"><code>df&lt;- data.frame(id = c(1,2,3), value = c(&#39;x,y&#39;, &#39;z,w&#39;, &#39;a&#39;))
df</code></pre>
<pre><code>#&gt;   id value
#&gt; 1  1   x,y
#&gt; 2  2   z,w
#&gt; 3  3     a</code></pre>
<p>we want to put x,y in the first row into two rows:</p>
<p>1, x</p>
<p>1, y</p>
<p>and put z,w into two rows too.</p>
</div>
<div id="solution-with-r" class="section level3">
<h3>solution with R</h3>
<p>There is a neat function <code>separate_rows</code> that does exactly this in <code>tidyr</code> package:</p>
<pre class="r"><code>library(tidyr)
library(dplyr)

df %&gt;%
        tidyr::separate_rows(value, sep=&quot;,&quot;)</code></pre>
<pre><code>#&gt; # A tibble: 5 × 2
#&gt;      id value
#&gt;   &lt;dbl&gt; &lt;chr&gt;
#&gt; 1     1 x    
#&gt; 2     1 y    
#&gt; 3     2 z    
#&gt; 4     2 w    
#&gt; 5     3 a</code></pre>
<p>We can also convert it back using some purrr tricks.</p>
<pre class="r"><code>library(tidyr)
library(purrr)

df_nest&lt;- df %&gt;%
        tidyr::separate_rows(value, sep=&quot;,&quot;) %&gt;%
        group_by(id) %&gt;%
        nest()

df_nest</code></pre>
<pre><code>#&gt; # A tibble: 3 × 2
#&gt; # Groups:   id [3]
#&gt;      id data            
#&gt;   &lt;dbl&gt; &lt;list&gt;          
#&gt; 1     1 &lt;tibble [2 × 1]&gt;
#&gt; 2     2 &lt;tibble [2 × 1]&gt;
#&gt; 3     3 &lt;tibble [1 × 1]&gt;</code></pre>
<pre class="r"><code>df_nest$data[[1]]</code></pre>
<pre><code>#&gt; # A tibble: 2 × 1
#&gt;   value
#&gt;   &lt;chr&gt;
#&gt; 1 x    
#&gt; 2 y</code></pre>
<pre class="r"><code># convert it back 
df_nest %&gt;%
        mutate(new_data = map_chr(data, ~paste(.x$value, collapse = &quot;,&quot;)))</code></pre>
<pre><code>#&gt; # A tibble: 3 × 3
#&gt; # Groups:   id [3]
#&gt;      id data             new_data
#&gt;   &lt;dbl&gt; &lt;list&gt;           &lt;chr&gt;   
#&gt; 1     1 &lt;tibble [2 × 1]&gt; x,y     
#&gt; 2     2 &lt;tibble [2 × 1]&gt; z,w     
#&gt; 3     3 &lt;tibble [1 × 1]&gt; a</code></pre>
</div>
<div id="solution-with-python" class="section level3">
<h3>solution with python</h3>
<p>Note, how to use python in Rmarkdown, read <a href="https://rstudio.github.io/reticulate/articles/versions.html" class="uri">https://rstudio.github.io/reticulate/articles/versions.html</a></p>
<pre class="r"><code>library(reticulate)
use_condaenv(&quot;r-reticulate&quot;)</code></pre>
<pre class="python"><code>import pandas as pd

df = pd.DataFrame({&#39;id&#39;: [1, 2, 3], &#39;data&#39;: [&#39;x, y&#39;, &#39;z, w&#39;, &#39;a&#39;]})

print(df)

# Split text into a list</code></pre>
<pre><code>#&gt;    id  data
#&gt; 0   1  x, y
#&gt; 1   2  z, w
#&gt; 2   3     a</code></pre>
<pre class="python"><code>df[&#39;data&#39;] = df[&#39;data&#39;].str.split(&#39;,&#39;)
print(df)

# Convert list into multiple rows</code></pre>
<pre><code>#&gt;    id     data
#&gt; 0   1  [x,  y]
#&gt; 1   2  [z,  w]
#&gt; 2   3      [a]</code></pre>
<pre class="python"><code>df = df.explode(&#39;data&#39;)
print(df)</code></pre>
<pre><code>#&gt;    id data
#&gt; 0   1    x
#&gt; 0   1    y
#&gt; 1   2    z
#&gt; 1   2    w
#&gt; 2   3    a</code></pre>
<p>we used <code>explode</code> function in pandas. Of course, you can do it manually:
read in the dataframe line by line, split the last column to a list using str.split(‘,’) and open an empty file and write the str.splited entries into new lines.</p>
</div>
<div id="use-case-in-bioinformatics" class="section level3">
<h3>use case in bioinformatics?</h3>
<p>follow <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-go.html" class="uri">https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-go.html</a>
for gene set enrichment analysis.</p>
<pre class="r"><code>library(clusterProfiler)
data(geneList, package=&quot;DOSE&quot;)
gene &lt;- names(geneList)[abs(geneList) &gt; 2]

# Entrez gene ID
head(gene)</code></pre>
<pre><code>#&gt; [1] &quot;4312&quot;  &quot;8318&quot;  &quot;10874&quot; &quot;55143&quot; &quot;55388&quot; &quot;991&quot;</code></pre>
<p>Let’s do a go-term over-representation analysis</p>
<pre class="r"><code>library(org.Hs.eg.db)

ego &lt;- enrichGO(gene          = gene,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = &quot;CC&quot;,
                pAdjustMethod = &quot;BH&quot;,
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable      = TRUE)
head(ego)</code></pre>
<pre><code>#&gt;                    ID                              Description GeneRatio
#&gt; GO:0005819 GO:0005819                                  spindle    26/201
#&gt; GO:0000779 GO:0000779 condensed chromosome, centromeric region    16/201
#&gt; GO:0072686 GO:0072686                          mitotic spindle    17/201
#&gt; GO:0000775 GO:0000775           chromosome, centromeric region    18/201
#&gt; GO:0098687 GO:0098687                       chromosomal region    23/201
#&gt; GO:0000776 GO:0000776                              kinetochore    15/201
#&gt;              BgRatio       pvalue     p.adjust       qvalue
#&gt; GO:0005819 306/11853 1.072029e-11 3.151766e-09 2.888837e-09
#&gt; GO:0000779 114/11853 7.709944e-11 8.659125e-09 7.936756e-09
#&gt; GO:0072686 133/11853 8.835841e-11 8.659125e-09 7.936756e-09
#&gt; GO:0000775 158/11853 1.684987e-10 1.179661e-08 1.081250e-08
#&gt; GO:0098687 272/11853 2.006225e-10 1.179661e-08 1.081250e-08
#&gt; GO:0000776 106/11853 2.733425e-10 1.339378e-08 1.227644e-08
#&gt;                                                                                                                                                               geneID
#&gt; GO:0005819 CDCA8/CDC20/KIF23/CENPE/ASPM/DLGAP5/SKA1/NUSAP1/TPX2/TACC3/NEK2/CDK1/MAD2L1/KIF18A/BIRC5/KIF11/TRAT1/TTK/AURKB/PRC1/KIFC1/KIF18B/KIF20A/AURKA/CCNB1/KIF4A
#&gt; GO:0000779                                                             CENPE/NDC80/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/TTK/NCAPG/AURKB/CCNB1
#&gt; GO:0072686                                                      KIF23/CENPE/ASPM/SKA1/NUSAP1/TPX2/TACC3/CDK1/MAD2L1/KIF18A/KIF11/TRAT1/AURKB/PRC1/KIFC1/KIF18B/AURKA
#&gt; GO:0000775                                                 CDCA8/CENPE/NDC80/TOP2A/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/TTK/NCAPG/AURKB/CCNB1
#&gt; GO:0098687                   CDCA8/CENPE/NDC80/TOP2A/HJURP/SKA1/NEK2/CENPM/RAD51AP1/CENPN/CDK1/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/EZH2/TTK/NCAPG/AURKB/CHEK1/CCNB1/MCM5
#&gt; GO:0000776                                                                   CENPE/NDC80/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/TTK/AURKB/CCNB1
#&gt;            Count
#&gt; GO:0005819    26
#&gt; GO:0000779    16
#&gt; GO:0072686    17
#&gt; GO:0000775    18
#&gt; GO:0098687    23
#&gt; GO:0000776    15</code></pre>
<p>The last <code>geneID</code> column is a list of genes separated by <code>/</code>. If I want to
get a tidy dataframe, I can do</p>
<pre class="r"><code>ego@result %&gt;% 
        tidyr::separate_rows(geneID, sep=&quot;/&quot;)</code></pre>
<pre><code>#&gt; # A tibble: 918 × 9
#&gt;    ID       Description GeneRatio BgRatio   pvalue p.adjust  qvalue geneID Count
#&gt;    &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;  &lt;int&gt;
#&gt;  1 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 CDCA8     26
#&gt;  2 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 CDC20     26
#&gt;  3 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 KIF23     26
#&gt;  4 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 CENPE     26
#&gt;  5 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 ASPM      26
#&gt;  6 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 DLGAP5    26
#&gt;  7 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 SKA1      26
#&gt;  8 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 NUSAP1    26
#&gt;  9 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 TPX2      26
#&gt; 10 GO:0005… spindle     26/201    306/11… 1.07e-11  3.15e-9 2.89e-9 TACC3     26
#&gt; # ℹ 908 more rows</code></pre>
<p>Now each gene has its own row!</p>
</div>
