<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="Context and Problem In scRNA-seq, each cell is sequenced individually, allowing for the analysis of gene expression at the single-cell level. This provides a wealth of information about the cellular identities and states. However, the high dimensionality of the data (thousands of genes) and the technical noise in the data can lead to challenges in accurately clustering the cells. Over-clustering is one such challenge, where cells that are biologically similar are clustered into distinct clusters.">

  
  <link rel="alternate" hreflang="en-us" href="/post/fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback/">
  <meta property="og:title" content="Fine tune the best clustering resolution for scRNAseq data: trying out callback | DNA confesses Data speak">
  <meta property="og:description" content="Context and Problem In scRNA-seq, each cell is sequenced individually, allowing for the analysis of gene expression at the single-cell level. This provides a wealth of information about the cellular identities and states. However, the high dimensionality of the data (thousands of genes) and the technical noise in the data can lead to challenges in accurately clustering the cells. Over-clustering is one such challenge, where cells that are biologically similar are clustered into distinct clusters.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2024-03-20T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2024-03-20T00:00:00&#43;00:00">
  

  
  

  <title>Fine tune the best clustering resolution for scRNAseq data: trying out callback | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>Newsletter</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">Fine tune the best clustering resolution for scRNAseq data: trying out callback</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2024-03-20 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Mar 20, 2024
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    12 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioconductor/">bioconductor</a
    >, 
    
    <a href="/categories/single-cell/">single-cell</a
    >, 
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Fine%20tune%20the%20best%20clustering%20resolution%20for%20scRNAseq%20data%3a%20trying%20out%20callback&amp;url=%2fpost%2ffine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2ffine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2ffine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback%2f&amp;title=Fine%20tune%20the%20best%20clustering%20resolution%20for%20scRNAseq%20data%3a%20trying%20out%20callback"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2ffine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback%2f&amp;title=Fine%20tune%20the%20best%20clustering%20resolution%20for%20scRNAseq%20data%3a%20trying%20out%20callback"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Fine%20tune%20the%20best%20clustering%20resolution%20for%20scRNAseq%20data%3a%20trying%20out%20callback&amp;body=%2fpost%2ffine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<div id="context-and-problem" class="section level3">
<h3>Context and Problem</h3>
<p>In scRNA-seq, each cell is sequenced individually, allowing for the analysis of gene expression at the single-cell level. This provides a wealth of information about the cellular identities and states. However, the high dimensionality of the data (thousands of genes) and the technical noise in the data can lead to challenges in accurately clustering the cells. Over-clustering is one such challenge, where cells that are biologically similar are clustered into distinct clusters.</p>
<p>In a <a href="https://divingintogeneticsandgenomics.com/post/scrnaseq-clustering-significant-test-an-unsolvable-problem/">previous post</a>, I tested <a href="https://github.com/igrabski/sc-SHC"><code>scSHC</code></a> which tries to solve this problem.</p>
<p>I recently saw this paper <a href="https://www.biorxiv.org/content/10.1101/2024.03.08.584180v1.full">A knockoff calibration method to avoid over-clustering in single-cell RNA-sequencing</a>.</p>
<p>The paper introduces a “knockoff calibration method” to address the over-clustering problem. The method involves the use of knockoff features, similar to the approach described earlier, but tailored to the specific challenges of scRNA-seq data.</p>
<ol style="list-style-type: decimal">
<li><p>Knockoff Features Creation: The algorithm generates knockoff features from the original gene expression data. These knockoff features are designed to mimic the original data but are statistically independent of the cell identities or states.</p></li>
<li><p>Clustering with Knockoff Features: The algorithm then applies a clustering algorithm to these knockoff features. Since the knockoff features are independent of the cell identities, the clustering should ideally reflect the true structure of the data without being influenced by the biological identities of the cells.</p></li>
<li><p>Calibration: The algorithm uses the clustering results from the knockoff features to calibrate the clustering of the original data. This calibration process adjusts the clustering of the original data to avoid over-clustering, ensuring that the final clusters reflect the true biological identities of the cells.</p></li>
<li><p>Improved Clustering: By using knockoff features for calibration, the algorithm aims to improve the accuracy of the clustering, reducing the likelihood of over-clustering. This results in more distinct clusters that better reflect the biological diversity of the single-cell population.</p></li>
</ol>
<p>The Knockoff Calibration Method implemented in <a href="https://github.com/lcrawlab/callback"><code>callback</code></a> R package:</p>
<p>Let’s test it using the same PBMC3k datatset.</p>
<pre class="r"><code>library(dplyr)
library(Seurat)
library(scCustomize)
library(patchwork)
library(ggplot2)
library(ComplexHeatmap)
library(SeuratData)
set.seed(1234)</code></pre>
<pre class="r"><code>data(&quot;pbmc3k&quot;)

pbmc3k&lt;- UpdateSeuratObject(pbmc3k)
pbmc3k</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 13714 features across 2700 samples within 1 assay 
#&gt; Active assay: RNA (13714 features, 0 variable features)
#&gt;  2 layers present: counts, data</code></pre>
</div>
<div id="routine-processing" class="section level2">
<h2>routine processing</h2>
<pre class="r"><code>pbmc3k&lt;- pbmc3k %&gt;% 
  NormalizeData(normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000) %&gt;%
  FindVariableFeatures(selection.method = &quot;vst&quot;, nfeatures = 2000) %&gt;%
  ScaleData() %&gt;%
  RunPCA(verbose = FALSE) %&gt;%
  FindNeighbors(dims = 1:10, verbose = FALSE) %&gt;%
  FindClusters(resolution = 0.5, verbose = FALSE) %&gt;%
  RunUMAP(dims = 1:10, verbose = FALSE)


p1&lt;- DimPlot_scCustom(pbmc3k, reduction = &quot;umap&quot;, label = TRUE, group.by = 
                        &quot;RNA_snn_res.0.5&quot;)

p2&lt;- DimPlot_scCustom(pbmc3k, reduction = &quot;umap&quot;, label = TRUE, group.by = &quot;seurat_annotations&quot;, label.size = 3)

p1 + p2</code></pre>
<p><img src="/post/2024-03-20-fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback_files/figure-html/unnamed-chunk-4-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>janitor::tabyl(pbmc3k@meta.data, seurat_annotations, RNA_snn_res.0.5)</code></pre>
<pre><code>#&gt;  seurat_annotations   0   1   2   3   4   5   6  7  8
#&gt;         Naive CD4 T 608   0  66   0  23   0   0  0  0
#&gt;        Memory CD4 T  69   0 396   0  16   0   2  0  0
#&gt;          CD14+ Mono   0 472   0   0   0   3   0  4  1
#&gt;                   B   0   0   0 343   1   0   0  0  0
#&gt;               CD8 T   2   0   3   0 265   0   1  0  0
#&gt;        FCGR3A+ Mono   0   7   0   0   0 155   0  0  0
#&gt;                  NK   0   0   0   0  16   0 139  0  0
#&gt;                  DC   0   0   0   0   0   0   0 32  0
#&gt;            Platelet   0   0   0   0   0   0   0  0 14
#&gt;                &lt;NA&gt;  17  12   2   6  18   1   6  0  0</code></pre>
<p>Let’s artificially increase the resolution (to 1) to over-cluster it.</p>
<pre class="r"><code>## artificially increase the resolution
pbmc3k&lt;- pbmc3k %&gt;%
  FindNeighbors(dims = 1:10, verbose = FALSE) %&gt;%
  FindClusters(resolution = 1, verbose = FALSE) 


p3&lt;- DimPlot_scCustom(pbmc3k, reduction = &quot;umap&quot;, label = TRUE, group.by = &quot;RNA_snn_res.1&quot;)

(p1 + p3) / p2</code></pre>
<p><img src="/post/2024-03-20-fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback_files/figure-html/unnamed-chunk-5-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>CD4 naive cell cluster is split to 2 clusters (0 -&gt; 0, 5)
the CD14+ monocyte cluster is split into 2 clusters (1 -&gt; 4, 6)</p>
<pre class="r"><code>janitor::tabyl(pbmc3k@meta.data, RNA_snn_res.1, RNA_snn_res.0.5) </code></pre>
<pre><code>#&gt;  RNA_snn_res.1   0   1   2   3   4   5   6  7  8
#&gt;              0 471   0   9   0   0   0   0  0  0
#&gt;              1   4   0 455   0   0   0   0  0  0
#&gt;              2   0   0   0 349   0   0   0  0  0
#&gt;              3   0   0   2   0 303   0   0  0  0
#&gt;              4   0 268   0   0   0   1   0  0  0
#&gt;              5 221   0   1   0  36   0   0  0  0
#&gt;              6   0 223   0   0   0   0   0  0  0
#&gt;              7   0   0   0   0   0 158   0  0  0
#&gt;              8   0   0   0   0   0   0 148  0  0
#&gt;              9   0   0   0   0   0   0   0 36  0
#&gt;             10   0   0   0   0   0   0   0  0 15</code></pre>
<p>visualize it using <a href="https://github.com/crazyhottommy/scclusteval" class="uri">https://github.com/crazyhottommy/scclusteval</a></p>
<pre class="r"><code>library(scclusteval)

PairWiseJaccardSetsHeatmap(pbmc3k$RNA_snn_res.0.5,
                           pbmc3k$RNA_snn_res.1,
                           show_row_dend = F, show_column_dend = F,
                           cluster_row = F, cluster_column =F)</code></pre>
<p><img src="/post/2024-03-20-fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback_files/figure-html/unnamed-chunk-7-1.png" width="576" /></p>
<div id="test-callback" class="section level3">
<h3>Test callback</h3>
<p>Install it here <a href="https://github.com/lcrawlab/callback" class="uri">https://github.com/lcrawlab/callback</a></p>
<pre class="r"><code>#devtools::install_github(&quot;lcrawlab/callback&quot;)
library(callback)
library(tictoc)

tic()
pbmc3k_callback &lt;- FindClustersCallback(pbmc3k)
toc()</code></pre>
<pre><code>#&gt; 53.227 sec elapsed</code></pre>
<pre class="r"><code>p4&lt;- scCustomize::DimPlot_scCustom(pbmc3k_callback)

(p1 + p4) / p2</code></pre>
<p><img src="/post/2024-03-20-fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback_files/figure-html/unnamed-chunk-8-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>PairWiseJaccardSetsHeatmap(pbmc3k$seurat_annotations,
                           pbmc3k_callback$callback_clusters,
                           show_row_dend = F, show_column_dend = F,
                           cluster_row = F, cluster_column =F)</code></pre>
<p><img src="/post/2024-03-20-fine-tune-the-best-clustering-resolution-for-scrnaseq-data-trying-out-callback_files/figure-html/unnamed-chunk-9-1.png" width="576" /></p>
<p><code>callback</code> merges the CD8 T and NK cells into cluster 2, and merges the
naive CD4 T cells and the memory CD4 T cells. Of course, you do not know the
original seurat cluster annotation is 100% correct or not. Distinguishing naive
and memory CD4 is harder too. However, merging NK cells with CD8T cells do make me worry :)</p>
</div>
<div id="conclusions" class="section level3">
<h3>conclusions</h3>
<ol style="list-style-type: decimal">
<li><p>Similar to <a href="https://github.com/igrabski/sc-SHC"><code>scSHC</code></a>, while statistically attractive, we need to use it with precaution and validate the clusters with biology knowledge.</p></li>
<li><p>Always use a small dataset that you are familiar with to test a new method.</p></li>
<li><p>The cluster number will be depending on the parameter <code>dims = 1:10</code> in the <code>FindClustersCallback</code> function too.</p></li>
<li><p>I may still cluster with a bunch of different resolutions and make sense of
them using biology knowledge and manually merge or split the clusters.</p></li>
</ol>
<p>what’s your take?</p>
<pre class="r"><code>devtools::session_info()</code></pre>
<pre><code>#&gt; ─ Session info  😕  🤕  📮   ─────────────────────────────────────────────────
#&gt;  hash: confused face, face with head-bandage, postbox
#&gt; 
#&gt;  setting  value
#&gt;  version  R version 4.1.2 (2021-11-01)
#&gt;  os       macOS Big Sur 10.16
#&gt;  system   x86_64, darwin17.0
#&gt;  ui       X11
#&gt;  language (EN)
#&gt;  collate  en_US.UTF-8
#&gt;  ctype    en_US.UTF-8
#&gt;  tz       America/New_York
#&gt;  date     2024-03-22
#&gt;  pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────
#&gt;  package            * version    date (UTC) lib source
#&gt;  abind                1.4-5      2016-07-21 [1] CRAN (R 4.1.0)
#&gt;  beeswarm             0.4.0      2021-06-01 [1] CRAN (R 4.1.0)
#&gt;  BiocGenerics         0.40.0     2021-10-26 [1] Bioconductor
#&gt;  blogdown             1.7        2021-12-19 [1] CRAN (R 4.1.2)
#&gt;  bookdown             0.24       2021-09-02 [1] CRAN (R 4.1.0)
#&gt;  bslib                0.3.1      2021-10-06 [1] CRAN (R 4.1.0)
#&gt;  cachem               1.0.6      2021-08-19 [1] CRAN (R 4.1.0)
#&gt;  callback           * 0.0.0      2024-03-15 [1] Github (lcrawlab/callback@655705f)
#&gt;  callr                3.7.0      2021-04-20 [1] CRAN (R 4.1.0)
#&gt;  circlize             0.4.13     2021-06-09 [1] CRAN (R 4.1.0)
#&gt;  cli                  3.6.1      2023-03-23 [1] CRAN (R 4.1.2)
#&gt;  clue                 0.3-60     2021-10-11 [1] CRAN (R 4.1.0)
#&gt;  cluster              2.1.2      2021-04-17 [1] CRAN (R 4.1.2)
#&gt;  codetools            0.2-18     2020-11-04 [1] CRAN (R 4.1.2)
#&gt;  colorspace           2.0-2      2021-06-24 [1] CRAN (R 4.1.0)
#&gt;  ComplexHeatmap     * 2.10.0     2021-10-26 [1] Bioconductor
#&gt;  cowplot              1.1.1      2020-12-30 [1] CRAN (R 4.1.0)
#&gt;  crayon               1.4.2      2021-10-29 [1] CRAN (R 4.1.0)
#&gt;  data.table           1.14.2     2021-09-27 [1] CRAN (R 4.1.0)
#&gt;  DBI                  1.1.1      2021-01-15 [1] CRAN (R 4.1.0)
#&gt;  deldir               1.0-6      2021-10-23 [1] CRAN (R 4.1.0)
#&gt;  desc                 1.4.0      2021-09-28 [1] CRAN (R 4.1.0)
#&gt;  devtools             2.4.2      2021-06-07 [1] CRAN (R 4.1.0)
#&gt;  digest               0.6.28     2021-09-23 [1] CRAN (R 4.1.0)
#&gt;  doParallel           1.0.17     2022-02-07 [1] CRAN (R 4.1.2)
#&gt;  dotCall64            1.1-1      2023-11-28 [1] CRAN (R 4.1.2)
#&gt;  dplyr              * 1.1.2      2023-04-20 [1] CRAN (R 4.1.2)
#&gt;  ellipsis             0.3.2      2021-04-29 [1] CRAN (R 4.1.0)
#&gt;  evaluate             0.14       2019-05-28 [1] CRAN (R 4.1.0)
#&gt;  fansi                0.5.0      2021-05-25 [1] CRAN (R 4.1.0)
#&gt;  farver               2.1.0      2021-02-28 [1] CRAN (R 4.1.0)
#&gt;  fastDummies          1.7.3      2023-07-06 [1] CRAN (R 4.1.2)
#&gt;  fastmap              1.1.0      2021-01-25 [1] CRAN (R 4.1.0)
#&gt;  fitdistrplus         1.1-6      2021-09-28 [1] CRAN (R 4.1.0)
#&gt;  forcats              0.5.1      2021-01-27 [1] CRAN (R 4.1.0)
#&gt;  foreach              1.5.1      2020-10-15 [1] CRAN (R 4.1.0)
#&gt;  fs                   1.5.0      2020-07-31 [1] CRAN (R 4.1.0)
#&gt;  future               1.25.0     2022-04-24 [1] CRAN (R 4.1.2)
#&gt;  future.apply         1.8.1      2021-08-10 [1] CRAN (R 4.1.0)
#&gt;  generics             0.1.3      2022-07-05 [1] CRAN (R 4.1.2)
#&gt;  GetoptLong           1.0.5      2020-12-15 [1] CRAN (R 4.1.0)
#&gt;  ggbeeswarm           0.6.0      2017-08-07 [1] CRAN (R 4.1.0)
#&gt;  ggplot2            * 3.4.4      2023-10-12 [1] CRAN (R 4.1.2)
#&gt;  ggprism              1.0.3.9000 2021-12-07 [1] Github (csdaw/ggprism@e21c3ee)
#&gt;  ggrastr              1.0.1      2021-12-08 [1] CRAN (R 4.1.0)
#&gt;  ggrepel              0.9.3      2023-02-03 [1] CRAN (R 4.1.2)
#&gt;  ggridges             0.5.4      2022-09-26 [1] CRAN (R 4.1.2)
#&gt;  GlobalOptions        0.1.2      2020-06-10 [1] CRAN (R 4.1.0)
#&gt;  globals              0.14.0     2020-11-22 [1] CRAN (R 4.1.0)
#&gt;  glue                 1.6.2      2022-02-24 [1] CRAN (R 4.1.2)
#&gt;  goftest              1.2-3      2021-10-07 [1] CRAN (R 4.1.0)
#&gt;  gridExtra            2.3        2017-09-09 [1] CRAN (R 4.1.0)
#&gt;  gtable               0.3.0      2019-03-25 [1] CRAN (R 4.1.0)
#&gt;  highr                0.9        2021-04-16 [1] CRAN (R 4.1.0)
#&gt;  htmltools            0.5.2      2021-08-25 [1] CRAN (R 4.1.0)
#&gt;  htmlwidgets          1.5.4      2021-09-08 [1] CRAN (R 4.1.0)
#&gt;  httpuv               1.6.3      2021-09-09 [1] CRAN (R 4.1.0)
#&gt;  httr                 1.4.2      2020-07-20 [1] CRAN (R 4.1.0)
#&gt;  ica                  1.0-2      2018-05-24 [1] CRAN (R 4.1.0)
#&gt;  ifnb.SeuratData    * 3.1.0      2024-01-17 [1] local
#&gt;  igraph               1.2.7      2021-10-15 [1] CRAN (R 4.1.0)
#&gt;  IRanges              2.28.0     2021-10-26 [1] Bioconductor
#&gt;  irlba                2.3.5.1    2022-10-03 [1] CRAN (R 4.1.2)
#&gt;  iterators            1.0.13     2020-10-15 [1] CRAN (R 4.1.0)
#&gt;  janitor              2.1.0      2021-01-05 [1] CRAN (R 4.1.0)
#&gt;  jquerylib            0.1.4      2021-04-26 [1] CRAN (R 4.1.0)
#&gt;  jsonlite             1.7.2      2020-12-09 [1] CRAN (R 4.1.0)
#&gt;  KernSmooth           2.23-20    2021-05-03 [1] CRAN (R 4.1.2)
#&gt;  knitr                1.36       2021-09-29 [1] CRAN (R 4.1.0)
#&gt;  knockoff             0.3.6      2022-08-15 [1] CRAN (R 4.1.2)
#&gt;  labeling             0.4.2      2020-10-20 [1] CRAN (R 4.1.0)
#&gt;  lamW                 2.2.3      2023-12-01 [1] CRAN (R 4.1.2)
#&gt;  later                1.3.0      2021-08-18 [1] CRAN (R 4.1.0)
#&gt;  lattice              0.20-45    2021-09-22 [1] CRAN (R 4.1.2)
#&gt;  lazyeval             0.2.2      2019-03-15 [1] CRAN (R 4.1.0)
#&gt;  leiden               0.3.9      2021-07-27 [1] CRAN (R 4.1.0)
#&gt;  lifecycle            1.0.3      2022-10-07 [1] CRAN (R 4.1.2)
#&gt;  limma                3.50.0     2021-10-26 [1] Bioconductor
#&gt;  listenv              0.8.0      2019-12-05 [1] CRAN (R 4.1.0)
#&gt;  lmtest               0.9-39     2021-11-07 [1] CRAN (R 4.1.0)
#&gt;  lubridate            1.8.0      2021-10-07 [1] CRAN (R 4.1.0)
#&gt;  magick               2.7.4      2023-03-09 [1] CRAN (R 4.1.2)
#&gt;  magrittr             2.0.1      2020-11-17 [1] CRAN (R 4.1.0)
#&gt;  MASS                 7.3-54     2021-05-03 [1] CRAN (R 4.1.2)
#&gt;  Matrix               1.6-3      2023-11-14 [1] CRAN (R 4.1.2)
#&gt;  matrixStats          0.61.0     2021-09-17 [1] CRAN (R 4.1.0)
#&gt;  memoise              2.0.0      2021-01-26 [1] CRAN (R 4.1.0)
#&gt;  mime                 0.12       2021-09-28 [1] CRAN (R 4.1.0)
#&gt;  miniUI               0.1.1.1    2018-05-18 [1] CRAN (R 4.1.0)
#&gt;  munsell              0.5.0      2018-06-12 [1] CRAN (R 4.1.0)
#&gt;  nlme                 3.1-153    2021-09-07 [1] CRAN (R 4.1.2)
#&gt;  paletteer            1.4.0      2021-07-20 [1] CRAN (R 4.1.0)
#&gt;  parallelly           1.31.1     2022-04-22 [1] CRAN (R 4.1.2)
#&gt;  patchwork          * 1.1.1      2020-12-17 [1] CRAN (R 4.1.0)
#&gt;  pbapply              1.5-0      2021-09-16 [1] CRAN (R 4.1.0)
#&gt;  pbmc3k.SeuratData  * 3.1.4      2022-08-03 [1] local
#&gt;  pbmcref.SeuratData * 1.0.0      2023-09-14 [1] local
#&gt;  pillar               1.9.0      2023-03-22 [1] CRAN (R 4.1.2)
#&gt;  pkgbuild             1.2.0      2020-12-15 [1] CRAN (R 4.1.0)
#&gt;  pkgconfig            2.0.3      2019-09-22 [1] CRAN (R 4.1.0)
#&gt;  pkgload              1.2.3      2021-10-13 [1] CRAN (R 4.1.0)
#&gt;  plotly               4.10.0     2021-10-09 [1] CRAN (R 4.1.0)
#&gt;  plyr                 1.8.6      2020-03-03 [1] CRAN (R 4.1.0)
#&gt;  png                  0.1-8      2022-11-29 [1] CRAN (R 4.1.2)
#&gt;  polyclip             1.10-0     2019-03-14 [1] CRAN (R 4.1.0)
#&gt;  presto               1.0.0      2023-03-30 [1] Github (immunogenomics/presto@045390a)
#&gt;  prettyunits          1.1.1      2020-01-24 [1] CRAN (R 4.1.0)
#&gt;  prismatic            1.1.0      2021-10-17 [1] CRAN (R 4.1.0)
#&gt;  processx             3.5.2      2021-04-30 [1] CRAN (R 4.1.0)
#&gt;  progressr            0.9.0      2021-09-24 [1] CRAN (R 4.1.0)
#&gt;  promises             1.2.0.1    2021-02-11 [1] CRAN (R 4.1.0)
#&gt;  ps                   1.6.0      2021-02-28 [1] CRAN (R 4.1.0)
#&gt;  purrr                1.0.1      2023-01-10 [1] CRAN (R 4.1.2)
#&gt;  R6                   2.5.1      2021-08-19 [1] CRAN (R 4.1.0)
#&gt;  RANN                 2.6.1      2019-01-08 [1] CRAN (R 4.1.0)
#&gt;  rappdirs             0.3.3      2021-01-31 [1] CRAN (R 4.1.0)
#&gt;  RColorBrewer         1.1-2      2014-12-07 [1] CRAN (R 4.1.0)
#&gt;  Rcpp                 1.0.11     2023-07-06 [1] CRAN (R 4.1.2)
#&gt;  RcppAnnoy            0.0.19     2021-07-30 [1] CRAN (R 4.1.0)
#&gt;  RcppHNSW             0.3.0      2020-09-06 [1] CRAN (R 4.1.0)
#&gt;  RcppParallel         5.1.4      2021-05-04 [1] CRAN (R 4.1.0)
#&gt;  rematch2             2.1.2      2020-05-01 [1] CRAN (R 4.1.0)
#&gt;  remotes              2.4.1      2021-09-29 [1] CRAN (R 4.1.0)
#&gt;  reshape2             1.4.4      2020-04-09 [1] CRAN (R 4.1.0)
#&gt;  reticulate           1.28       2023-01-27 [1] CRAN (R 4.1.2)
#&gt;  rjson                0.2.20     2018-06-08 [1] CRAN (R 4.1.0)
#&gt;  rlang                1.1.3      2024-01-10 [1] CRAN (R 4.1.2)
#&gt;  rmarkdown            2.11       2021-09-14 [1] CRAN (R 4.1.0)
#&gt;  ROCR                 1.0-11     2020-05-02 [1] CRAN (R 4.1.0)
#&gt;  rprojroot            2.0.2      2020-11-15 [1] CRAN (R 4.1.0)
#&gt;  RSpectra             0.16-0     2019-12-01 [1] CRAN (R 4.1.0)
#&gt;  rstudioapi           0.13       2020-11-12 [1] CRAN (R 4.1.0)
#&gt;  Rtsne                0.15       2018-11-10 [1] CRAN (R 4.1.0)
#&gt;  S4Vectors            0.32.2     2021-11-07 [1] Bioconductor
#&gt;  sass                 0.4.0      2021-05-12 [1] CRAN (R 4.1.0)
#&gt;  scales               1.3.0      2023-11-28 [1] CRAN (R 4.1.2)
#&gt;  scattermore          1.2        2023-06-12 [1] CRAN (R 4.1.2)
#&gt;  scclusteval        * 0.0.0.9000 2022-08-05 [1] Github (crazyhottommy/scclusteval@b1b22c7)
#&gt;  scCustomize        * 2.1.2      2024-02-28 [1] CRAN (R 4.1.2)
#&gt;  sctransform          0.4.1      2023-10-19 [1] CRAN (R 4.1.2)
#&gt;  sessioninfo          1.2.1      2021-11-02 [1] CRAN (R 4.1.0)
#&gt;  Seurat             * 5.0.1      2023-11-17 [1] CRAN (R 4.1.2)
#&gt;  SeuratData         * 0.2.2      2022-08-03 [1] Github (satijalab/seurat-data@d6a8ce6)
#&gt;  SeuratObject       * 5.0.1      2023-11-17 [1] CRAN (R 4.1.2)
#&gt;  shape                1.4.6      2021-05-19 [1] CRAN (R 4.1.0)
#&gt;  shiny                1.7.1      2021-10-02 [1] CRAN (R 4.1.0)
#&gt;  snakecase            0.11.0     2019-05-25 [1] CRAN (R 4.1.0)
#&gt;  sp                 * 1.6-0      2023-01-19 [1] CRAN (R 4.1.2)
#&gt;  spam                 2.10-0     2023-10-23 [1] CRAN (R 4.1.2)
#&gt;  spatstat.data        3.0-0      2022-10-21 [1] CRAN (R 4.1.2)
#&gt;  spatstat.explore     3.0-6      2023-01-26 [1] CRAN (R 4.1.2)
#&gt;  spatstat.geom        3.0-6      2023-01-30 [1] CRAN (R 4.1.2)
#&gt;  spatstat.random      3.1-3      2023-01-25 [1] CRAN (R 4.1.2)
#&gt;  spatstat.sparse      3.0-0      2022-10-21 [1] CRAN (R 4.1.2)
#&gt;  spatstat.utils       3.0-1      2022-10-19 [1] CRAN (R 4.1.2)
#&gt;  stringi              1.7.5      2021-10-04 [1] CRAN (R 4.1.0)
#&gt;  stringr              1.5.1      2023-11-14 [1] CRAN (R 4.1.2)
#&gt;  survival             3.2-13     2021-08-24 [1] CRAN (R 4.1.2)
#&gt;  tensor               1.5        2012-05-05 [1] CRAN (R 4.1.0)
#&gt;  testthat             3.1.0      2021-10-04 [1] CRAN (R 4.1.0)
#&gt;  tibble               3.2.1      2023-03-20 [1] CRAN (R 4.1.2)
#&gt;  tictoc             * 1.0.1      2021-04-19 [1] CRAN (R 4.1.0)
#&gt;  tidyr                1.3.0      2023-01-24 [1] CRAN (R 4.1.2)
#&gt;  tidyselect           1.2.0      2022-10-10 [1] CRAN (R 4.1.2)
#&gt;  usethis              2.1.3      2021-10-27 [1] CRAN (R 4.1.0)
#&gt;  utf8                 1.2.2      2021-07-24 [1] CRAN (R 4.1.0)
#&gt;  uwot                 0.1.14     2022-08-22 [1] CRAN (R 4.1.2)
#&gt;  vctrs                0.6.2      2023-04-19 [1] CRAN (R 4.1.2)
#&gt;  vipor                0.4.5      2017-03-22 [1] CRAN (R 4.1.0)
#&gt;  viridisLite          0.4.0      2021-04-13 [1] CRAN (R 4.1.0)
#&gt;  withr                2.5.0      2022-03-03 [1] CRAN (R 4.1.2)
#&gt;  xfun                 0.36       2022-12-21 [1] CRAN (R 4.1.2)
#&gt;  xtable               1.8-4      2019-04-21 [1] CRAN (R 4.1.0)
#&gt;  yaml                 2.2.1      2020-02-01 [1] CRAN (R 4.1.0)
#&gt;  zoo                  1.8-9      2021-03-09 [1] CRAN (R 4.1.0)
#&gt; 
#&gt;  [1] /Library/Frameworks/R.framework/Versions/4.1/Resources/library
#&gt; 
#&gt; ──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/bioinformatics/">bioinformatics</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/seurat/">Seurat</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/single-cell/">single-cell</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/do-you-really-understand-log2fold-change-in-single-cell-rnaseq-data/">Do you really understand log2Fold change in single-cell RNAseq data?</a></li>
        
        <li><a href="/post/reuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets/">Reuse the single cell data! How to create a seurat object from GEO datasets</a></li>
        
        <li><a href="/post/details-in-centered-log-ratio-clr-normalization-for-cite-seq-protein-count-data/">Part 3 Centered log ratio (CLR) normalization for CITE-seq protein count data</a></li>
        
        <li><a href="/post/how-to-create-pseudobulk-from-single-cell-rnaseq-data/">How to create pseudobulk from single-cell RNAseq data </a></li>
        
        <li><a href="/post/scrnaseq-clustering-significant-test-an-unsolvable-problem/">scRNAseq clustering significance test: an unsolvable problem?</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/bioinformatics-is-not-just-statistics/" rel="next">Bioinformatics is not (just) statistics </a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/downstream-of-bulk-rnaseq-read-in-salmon-output-using-tximport-and-then-deseq2/" rel="prev">Downstream of bulk RNAseq: read in salmon output using tximport and then DESeq2</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

