<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.37" />
  <meta name="author" content="Ming Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="I asked this question on Twitter:
 what test to test if two distributions are different? I am aware of KS test. When n is large (which is common in genomic studies), the p-value is always significant. better to test against an effect size? how to do it in this context?
 In genomics studies, it is very common to have large N (e.g., the number of introns, promoters in the genome, number of cells in the single-cell studies).">

  
  <link rel="alternate" hreflang="en-us" href="/post/how-to-test-if-two-distributions-are-different/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/how-to-test-if-two-distributions-are-different/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/how-to-test-if-two-distributions-are-different/">
  <meta property="og:title" content="How to test if two distributions are different  | DNA confesses Data speak">
  <meta property="og:description" content="I asked this question on Twitter:
 what test to test if two distributions are different? I am aware of KS test. When n is large (which is common in genomic studies), the p-value is always significant. better to test against an effect size? how to do it in this context?
 In genomics studies, it is very common to have large N (e.g., the number of introns, promoters in the genome, number of cells in the single-cell studies).">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2021-08-07T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2021-08-07T00:00:00&#43;00:00">
  

  
  

  <title>How to test if two distributions are different  | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">How to test if two distributions are different </h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2021-08-07 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Aug 7, 2021
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    8 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/how-to-test-if-two-distributions-are-different/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >, 
    
    <a href="/categories/statistics/">statistics</a
    >, 
    
    <a href="/categories/rstats/">rstats</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=How%20to%20test%20if%20two%20distributions%20are%20different%20&amp;url=%2fpost%2fhow-to-test-if-two-distributions-are-different%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fhow-to-test-if-two-distributions-are-different%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fhow-to-test-if-two-distributions-are-different%2f&amp;title=How%20to%20test%20if%20two%20distributions%20are%20different%20"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fhow-to-test-if-two-distributions-are-different%2f&amp;title=How%20to%20test%20if%20two%20distributions%20are%20different%20"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=How%20to%20test%20if%20two%20distributions%20are%20different%20&amp;body=%2fpost%2fhow-to-test-if-two-distributions-are-different%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <p>I asked this question on <a href="https://twitter.com/tangming2005/status/1414431622141779971">Twitter</a>:</p>
<blockquote>
<p>what test to test if two distributions are different? I am aware of KS test. When n is large (which is common in genomic studies), the p-value is always significant. better to test against an effect size? how to do it in this context?</p>
</blockquote>
<p>In genomics studies, it is very common to have large N (e.g., the number of introns, promoters in the genome, number of cells in the single-cell studies). A more concrete example is that one have two samples: control and treatment. one can calculate the intron retention level for each intron across the genome and ask if globally the distribution of the intron retention scores is different between the control sample and treatment sample.</p>
<div id="kolmogorovsmirnov-test" class="section level3">
<h3>Kolmogorov–Smirnov test</h3>
<p>I am aware of the <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov–Smirnov test</a> (K–S test or KS test).</p>
<blockquote>
<p>The Kolmogorov–Smirnov statistic quantifies a <strong>distance</strong> between the empirical distribution function of the sample and the cumulative distribution function of the reference distribution, or between the empirical distribution functions of two samples.</p>
</blockquote>
<p>Kolmogorov–Smirnov statistic is used in the popular <a href="https://www.pnas.org/content/102/43/15545">GSEA</a> too.
It is good to see some examples.</p>
<p>We will use random sampling from normal distribution as examples.</p>
<pre class="r"><code>library(tidyverse)
library(patchwork)

test_distribution&lt;- function(N, mean1, mean2, sd1, sd2){
  set.seed(123)
  ## random sample from normal distribution
  x&lt;- rnorm(N, mean=mean1, sd = sd1)
  set.seed(234)
  y&lt;- rnorm(N, mean=mean2, sd = sd2) 
  
  df&lt;- data.frame(x=x, y =y)

  ## make it long format
  df&lt;- df %&gt;%
    pivot_longer(cols = 1:2, names_to = &quot;group&quot;, values_to = &quot;value&quot;)
  ## check the ECDF for the two distributions
  p1&lt;- ggplot(df, aes(value)) + 
    stat_ecdf(geom = &quot;step&quot;, aes(color = group)) +
    theme_classic(base_size = 14) +
    ylab(&quot;Cumulative probability&quot;) +
    ggtitle(&quot;ECDF for x and y&quot;, subtitle = glue::glue( &quot;x ~ N({mean1}, {sd1})\n y ~ N({mean2}, {sd2})&quot;)) 
    
  
  p2&lt;- ggplot(df, aes(x= group, y = value)) +
    geom_boxplot(aes(fill = group)) +
    theme_classic(base_size = 14) +
    ggtitle(paste0(&quot;N = &quot;, N))
    
  p&lt;- p1 + p2
  
  ks&lt;- ks.test(x,y)
  t&lt;- t.test(x,y)
  return(list(ks = ks, ttest =t, p = p, df = df))

}


res&lt;- test_distribution(1000, 0, 0, 1, 1)
res$p</code></pre>
<p><img src="/post/2021-08-07-how-to-test-if-two-distributions-are-different.en_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<pre class="r"><code>res$ks</code></pre>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and y
## D = 0.025, p-value = 0.9135
## alternative hypothesis: two-sided</code></pre>
<p>KS test showing the distribution is not significantly different which makes sense as we draw samples from the same normal distribution.</p>
<pre class="r"><code>res$ttest</code></pre>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  x and y
## t = 0.85389, df = 1997.9, p-value = 0.3933
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.04893966  0.12442136
## sample estimates:
##   mean of x   mean of y 
##  0.01612787 -0.02161298</code></pre>
<p>Two sample t test shows that the mean is not significantly different.</p>
<p>Let’s increase the standard deviation for the second sample to 1.2</p>
<pre class="r"><code>res&lt;- test_distribution(1000, 0, 0, 1, 1.2)
res$p</code></pre>
<p><img src="/post/2021-08-07-how-to-test-if-two-distributions-are-different.en_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<pre class="r"><code>res$ks</code></pre>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and y
## D = 0.053, p-value = 0.1205
## alternative hypothesis: two-sided</code></pre>
<pre class="r"><code>res$ttest</code></pre>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  x and y
## t = 0.86215, df = 1939.5, p-value = 0.3887
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.05362088  0.13774778
## sample estimates:
##   mean of x   mean of y 
##  0.01612787 -0.02593558</code></pre>
<p>Let’s increase the sample size</p>
<pre class="r"><code>res&lt;- test_distribution(10000, 0, 0, 1, 1.2)
res$p</code></pre>
<p><img src="/post/2021-08-07-how-to-test-if-two-distributions-are-different.en_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<pre class="r"><code>res$ks</code></pre>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and y
## D = 0.0497, p-value = 3.746e-11
## alternative hypothesis: two-sided</code></pre>
<pre class="r"><code>res$ttest</code></pre>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  x and y
## t = 0.19585, df = 19342, p-value = 0.8447
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.02758649  0.03371124
## sample estimates:
##    mean of x    mean of y 
## -0.002371702 -0.005434073</code></pre>
<p>After we increase the sample size, t-test which tests against mean is still not significant. However, the KS test becomes highly significant even the samples are drew from the same two distributions as the last comparison!!</p>
<p>Let’s try another example with very small difference (small here is subjective, one has to decide it under the experiment context) of means.</p>
<pre class="r"><code>res&lt;- test_distribution(1000, 0, 0.1, 1, 1)
res$p</code></pre>
<p><img src="/post/2021-08-07-how-to-test-if-two-distributions-are-different.en_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="r"><code>res$ks</code></pre>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and y
## D = 0.054, p-value = 0.1083
## alternative hypothesis: two-sided</code></pre>
<pre class="r"><code>res$ttest</code></pre>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  x and y
## t = -1.4086, df = 1997.9, p-value = 0.1591
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.14893966  0.02442136
## sample estimates:
##  mean of x  mean of y 
## 0.01612787 0.07838702</code></pre>
<p>Let’s increase the sample size to <code>10000</code>:</p>
<pre class="r"><code>res&lt;- test_distribution(10000, 0, 0.1, 1, 1)
res$p</code></pre>
<p><img src="/post/2021-08-07-how-to-test-if-two-distributions-are-different.en_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<pre class="r"><code>res$ks</code></pre>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and y
## D = 0.0502, p-value = 2.273e-11
## alternative hypothesis: two-sided</code></pre>
<pre class="r"><code>res$ttest</code></pre>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  x and y
## t = -6.914, df = 19998, p-value = 4.853e-12
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.12558133 -0.07010528
## sample estimates:
##    mean of x    mean of y 
## -0.002371702  0.095471606</code></pre>
<p>Suddenly, both the KS test and t-test becomes highly significant after increasing the sample size.
That’s why we should not put too much emphasis on the p-values, but also check the effect size. Many genomic papers show highly significant p values &lt; 2.22e−16 (smallest you can get from <code>R</code>) simply because the sample size is really big (I have to confess that I have it too in my papers because my PI/reviewers asked).</p>
</div>
<div id="earth-mover-distance-emd" class="section level3">
<h3>earth mover distance (EMD)</h3>
<p>Others in the tweet thread mentioned <a href="https://en.wikipedia.org/wiki/Earth_mover%27s_distance">earth mover distance</a> that can be used to measure the distance between two distributions. There is a biconductor package for calculating it. The <a href="https://www.bioconductor.org/packages/release/bioc/html/EMDomics.html">EMDomics</a> algorithm is used to perform a supervised multi-class analysis to measure the magnitude and statistical significance of observed continuous genomics data between groups.</p>
<pre class="r"><code>library(EMDomics)


calculate_EMD&lt;- function(df){
  num&lt;- 1:nrow(df)
  exp_data&lt;- df$value
  names(exp_data)&lt;- glue::glue(&quot;sample_{num}&quot;)
  labels&lt;- df$group
  names(labels)&lt;- names(exp_data)
  calculate_emd_gene(exp_data, labels, names(exp_data))
}

res&lt;- test_distribution(1000, 0, 0, 1, 1.2)
calculate_EMD(res$df)</code></pre>
<pre><code>## [1] 0.7629998</code></pre>
<pre class="r"><code>## increase the sample size 
res&lt;- test_distribution(10000, 0, 0, 1, 1.2)
calculate_EMD(res$df)</code></pre>
<pre><code>## [1] 0.8505</code></pre>
<pre class="r"><code>res&lt;- test_distribution(1000, 0, 0.1, 1, 1)
calculate_EMD(res$df)</code></pre>
<pre><code>## [1] 0.3919999</code></pre>
<pre class="r"><code>res$df %&gt;% head()</code></pre>
<pre><code>## # A tibble: 6 x 2
##   group  value
##   &lt;chr&gt;  &lt;dbl&gt;
## 1 x     -0.560
## 2 y      0.761
## 3 x     -0.230
## 4 y     -1.95 
## 5 x      1.56 
## 6 y     -1.40</code></pre>
<pre class="r"><code>res&lt;- test_distribution(10000, 0, 0.1, 1, 1)
calculate_EMD(res$df)</code></pre>
<pre><code>## [1] 0.4892001</code></pre>
<p>Two observations:</p>
<ul>
<li>EMD between <span class="math inline">\(N(0,1)\)</span> and <span class="math inline">\(N(0.1,1)\)</span> is smaller than EMD between <span class="math inline">\(N(0,1)\)</span> and <span class="math inline">\(N(0,1.2)\)</span></li>
<li>increasing the sample size will increase the EMD too!</li>
</ul>
<blockquote>
<p>The EMD score increases as the distributions become increasingly dissimilar, but we have no framework for estimating the significance of a particular EMD score. EMDomics uses a permutation-based method to calculate a q-value that is interpreted analogously to a p-value.</p>
</blockquote>
<p>The <code>EMDomics</code> packages implemented the permutation test for multiple genes. Let me do it for a single gene.</p>
<pre class="r"><code>res&lt;- test_distribution(1000, 0, 0, 1, 1.2)
res$df</code></pre>
<pre><code>## # A tibble: 2,000 x 2
##    group   value
##    &lt;chr&gt;   &lt;dbl&gt;
##  1 x     -0.560 
##  2 y      0.793 
##  3 x     -0.230 
##  4 y     -2.46  
##  5 x      1.56  
##  6 y     -1.80  
##  7 x      0.0705
##  8 y      1.77  
##  9 x      0.129 
## 10 y      1.75  
## # … with 1,990 more rows</code></pre>
<pre class="r"><code>calculate_EMD(res$df)</code></pre>
<pre><code>## [1] 0.7629998</code></pre>
<pre class="r"><code>## random shuffle the x, y group designation and calculate the EMD score
permutation_EMD&lt;- function(d){
  set.seed(NULL)
  d$group&lt;- sample(d$group)
  calculate_EMD(d)
}

permutation_EMD(res$df)</code></pre>
<pre><code>## [1] 0.165</code></pre>
<pre class="r"><code>## a different shuffle gives a different EMD
permutation_EMD(res$df)</code></pre>
<pre><code>## [1] 0.511</code></pre>
<pre class="r"><code>## permutate N times and calculate how many times the permutated EMD is bigger than the EMD
## for the original data and that&#39;s the p-value

permutation_EMD_pvalue&lt;- function(d, N_permutation = 1000){
  permutation_EMDs&lt;- replicate(N_permutation, permutation_EMD(d))
  ### p-value
  mean(calculate_EMD(d) &lt; permutation_EMDs)
}

permutation_EMD_pvalue(res$df)</code></pre>
<pre><code>## [1] 0.003</code></pre>
<p>The p-value is &lt; 0.05 which suggests that the two distributions are significantly different.</p>
<p>If we sample from the same <span class="math inline">\(N(0,1)\)</span> distribution:</p>
<pre class="r"><code>res&lt;- test_distribution(1000, 0, 0, 1, 1)
permutation_EMD_pvalue(res$df)</code></pre>
<pre><code>## [1] 0.602</code></pre>
<p>p-value is not significant.</p>
<p>In fact, many people suggested using a permutation based method to evaluate the significance of the distribution differences using the KS test. The below code snippet is from <a href="https://twitter.com/jma1991">James</a>:</p>
<p><img src="/img/permutation_ks.jpeg" /></p>
</div>
<div id="comparing-groups-of-distributions" class="section level3">
<h3>Comparing groups of distributions</h3>
<p>For scRNAseq or CyTOF data, if one has treatment vs control groups, 3 samples each, each sample contains 1000 cells. How to test if the treatment changes the expression of a certain gene? This is an example of multi-sample, multi-group single-cell differential analysis. Pseudo-bulk method can be used as in the <a href="https://www.nature.com/articles/s41467-020-19894-4">muscat</a>. Note that it is different from marker gene identification where differential gene expression is carried out between clusters of cells.</p>
<p>Papers to read:</p>
<ul>
<li><p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0679-0">quantro: a data-driven approach to guide the choice of an appropriate normalization method</a></p></li>
<li><p><a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btab226/6207964">Fast identification of differential distributions in single-cell RNA-sequencing data with waddR</a></p></li>
<li><p><a href="https://www.biorxiv.org/content/10.1101/2020.11.24.394213v3.full">distinct: a novel approach to differential distribution analyses</a></p></li>
</ul>
<p>Rafa, our department chair asked a good question:</p>
<blockquote class="twitter-tweet">
<p lang="en" dir="ltr">
First I would ask why is testing for a difference in mean and a difference in SD not enough? Can you construct examples of two distributions with the same mean and same SD but different in a meaningful way in your context? Seeing these examples will help motivate a solution.
</p>
— Rafael Irizarry (<span class="citation">@rafalab</span>) <a href="https://twitter.com/rafalab/status/1415110253549690882?ref_src=twsrc%5Etfw">July 14, 2021</a>
</blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
<p>Figure 7 b from the <code>distinct</code> paper shows some genes have different distributions but have the same mean across groups.</p>
<p><img src="/img/distinct_example.png" /></p>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/statistics/">statistics</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/r/">R</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/modeling-single-cell-rnaseq-data-with-multinomial-distribution/">Modeling single cell RNAseq data with multinomial distribution </a></li>
        
        <li><a href="/post/compare-slopes-in-linear-regression/">compare slopes in linear regression</a></li>
        
        <li><a href="/post/dplyr-count-misses-factor-levels-a-case-in-comparing-scrnaseq-cell-type-abundance/">dplyr::count misses factor levels: a case in comparing scRNAseq cell type abundance </a></li>
        
        <li><a href="/post/customize-featureplot-in-seurat-for-multi-condition-comparisons-using-patchwork/">customize FeaturePlot in Seurat for multi-condition comparisons using patchwork</a></li>
        
        <li><a href="/post/monty-hall-problem-a-peek-through-simulation/">Monty Hall problem- a peek through simulation</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/matrix-factorization-for-single-cell-rnaseq-data/" rel="next">Matrix Factorization for single-cell RNAseq data</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/clustered-dotplot-for-single-cell-rnaseq/" rel="prev">clustered dotplot for single-cell RNAseq</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

