<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="Download the data https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116256
cd data/GSE116256 wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE116nnn/GSE116256/suppl/GSE116256_RAW.tar tar xvf GSE116256_RAW.tar rm GSE116256_RAW.tar Depending on how the authors upload their data.
Some authors may just upload the merged count matrix file. This is the easiest situation.
In this dataset, each sample has a separate set of matrix (*dem.txt.gz), features and barcodes.
Total, there are 43 samples. For each sample, it has an associated metadata file (*anno.txt.gz) too.
You can inspect the files in command line:">

  
  <link rel="alternate" hreflang="en-us" href="/post/reuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Chatomics">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Chatomics">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/reuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="Chatomics">
  <meta property="og:url" content="/post/reuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets/">
  <meta property="og:title" content="Reuse the single cell data! How to create a seurat object from GEO datasets | Chatomics">
  <meta property="og:description" content="Download the data https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116256
cd data/GSE116256 wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE116nnn/GSE116256/suppl/GSE116256_RAW.tar tar xvf GSE116256_RAW.tar rm GSE116256_RAW.tar Depending on how the authors upload their data.
Some authors may just upload the merged count matrix file. This is the easiest situation.
In this dataset, each sample has a separate set of matrix (*dem.txt.gz), features and barcodes.
Total, there are 43 samples. For each sample, it has an associated metadata file (*anno.txt.gz) too.
You can inspect the files in command line:">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2023-07-14T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2023-07-14T00:00:00&#43;00:00">
  

  
  

  <title>Reuse the single cell data! How to create a seurat object from GEO datasets | Chatomics</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">Chatomics</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>Newsletter</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">Reuse the single cell data! How to create a seurat object from GEO datasets</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2023-07-14 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Jul 14, 2023
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    2 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/reuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >, 
    
    <a href="/categories/single-cell/">single-cell</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Reuse%20the%20single%20cell%20data%21%20How%20to%20create%20a%20seurat%20object%20from%20GEO%20datasets&amp;url=%2fpost%2freuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2freuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2freuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets%2f&amp;title=Reuse%20the%20single%20cell%20data%21%20How%20to%20create%20a%20seurat%20object%20from%20GEO%20datasets"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2freuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets%2f&amp;title=Reuse%20the%20single%20cell%20data%21%20How%20to%20create%20a%20seurat%20object%20from%20GEO%20datasets"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Reuse%20the%20single%20cell%20data%21%20How%20to%20create%20a%20seurat%20object%20from%20GEO%20datasets&amp;body=%2fpost%2freuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<link href="/rmarkdown-libs/vembedr/css/vembedr.css" rel="stylesheet" />


<div id="download-the-data" class="section level3">
<h3>Download the data</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116256" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116256</a></p>
<pre class="bash"><code>cd data/GSE116256
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE116nnn/GSE116256/suppl/GSE116256_RAW.tar
tar xvf GSE116256_RAW.tar
rm  GSE116256_RAW.tar</code></pre>
<p>Depending on how the authors upload their data.</p>
<p>Some authors may just upload the merged count matrix file. This is the easiest situation.</p>
<p>In this dataset, each sample has a separate set of matrix (*dem.txt.gz), features and barcodes.</p>
<p>Total, there are 43 samples. For each sample, it has an associated metadata file (*anno.txt.gz) too.</p>
<p><img src="/img/geo_data.png" />
You can inspect the files in command line:</p>
<pre class="bash"><code>zless -S GSM3587923_AML1012-D0.dem.txt.gz</code></pre>
<p>Go back to <code>R</code>.
Load the libraries:</p>
<pre class="r"><code>library(here)
library(stringr)
library(dplyr)
library(ggplot2)
library(Seurat)
library(purrr)
library(readr)
library(harmony)
library(scCustomize)
library(SeuratDisk)</code></pre>
<p>read in the count matrix</p>
<pre class="r"><code># a small function to read the counts
read_counts&lt;- function(file){
  x&lt;- read_tsv(file)
  x&lt;- as.data.frame(x)
  genes&lt;- x$Gene
  x&lt;- x[, -1]
  rownames(x)&lt;- genes
  return(as.matrix(x))
}


counts_files&lt;- list.files(here(&quot;data/GSE116256&quot;), full.names = TRUE, pattern = &quot;*dem.txt.gz&quot;)

samples&lt;- map_chr(counts_files, basename) 

samples&lt;- str_replace(samples, &quot;(GSM[0-9]+_.+).dem.txt.gz&quot;, &quot;\\1&quot;)

names(counts_files)&lt;- samples

# for demonstration, only read in the first 4 samples
counts&lt;- purrr::map(counts_files[1:4], read_counts)</code></pre>
<p>read in the meta data</p>
<pre class="r"><code>read_meta&lt;- function(file){
  y&lt;- read_tsv(file)
  y&lt;- as.data.frame(y)
  cells&lt;- y$Cell
  y&lt;- y[,-1]
  rownames(y)&lt;- cells
  return(y)
}


meta_files&lt;- list.files(here(&quot;data/GSE116256&quot;), full.names = TRUE, pattern = &quot;*anno.txt.gz&quot;)
meta_names&lt;- map_chr(meta_files, basename)

meta_names&lt;- str_replace(meta_names, &quot;(GSM[0-9]+_.+).anno.txt.gz&quot;, &quot;\\1&quot;)
names(meta_files)&lt;- meta_names

meta&lt;- purrr::map(meta_files[1:4], read_meta)</code></pre>
</div>
<div id="create-a-seurat-object" class="section level3">
<h3>create a seurat object</h3>
<pre class="r"><code>library(Matrix) #for sparse matrix
objs&lt;- purrr::map2(counts, meta,  
                   ~CreateSeuratObject(counts = as(.x, &quot;sparseMatrix&quot;), 
                                       meta.data = .y))


# merge to a single object 
merged_seurat&lt;- purrr::reduce(objs, function(x,y) {merge(x,y)})

## free memory
rm(counts)
rm(objs)
rm(meta)
gc()</code></pre>
<pre><code>#&gt;            used  (Mb) gc trigger  (Mb) limit (Mb)  max used  (Mb)
#&gt; Ncells  3689626 197.1    5497720 293.7         NA   5497720 293.7
#&gt; Vcells 13347305 101.9   99405338 758.5      32768 118504523 904.2</code></pre>
</div>
<div id="standard-preprocess-of-the-data" class="section level3">
<h3>standard preprocess of the data</h3>
<pre class="r"><code>merged_seurat&lt;- merged_seurat %&gt;%
  NormalizeData(normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000) %&gt;%
  FindVariableFeatures( selection.method = &quot;vst&quot;, nfeatures = 2000) %&gt;%
  ScaleData() %&gt;%
  RunPCA() %&gt;%
  RunHarmony(group.by.vars = &quot;orig.ident&quot;, dims.use = 1:30) %&gt;%
  RunUMAP(reduction = &quot;harmony&quot;, dims = 1:30) %&gt;%
  FindNeighbors(reduction = &quot;harmony&quot;, dims = 1:30) %&gt;% 
  FindClusters(resolution = 0.6)</code></pre>
<pre><code>#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
#&gt; 
#&gt; Number of nodes: 2392
#&gt; Number of edges: 128252
#&gt; 
#&gt; Running Louvain algorithm...
#&gt; Maximum modularity in 10 random starts: 0.8706
#&gt; Number of communities: 11
#&gt; Elapsed time: 0 seconds</code></pre>
</div>
<div id="visualization" class="section level3">
<h3>visualization</h3>
<pre class="r"><code>DimPlot_scCustom(seurat_object = merged_seurat)</code></pre>
<p><img src="/post/2023-07-14-reuse-the-single-cell-data-how-to-create-a-seurat-object-from-geo-datasets_files/figure-html/unnamed-chunk-9-1.png" width="576" /></p>
<p>I have a youtube video on this too! subscribe to chatomics!</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/5e_8wr5Xx_Q" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/single-cell/">single-cell</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/seurat/">Seurat</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/bioinformatics/">bioinformatics</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/ten-single-cell-data-benchmarking-papers/">10 single-cell data benchmarking papers</a></li>
        
        <li><a href="/post/how-to-add-boxplots-or-density-plots-side-by-side-a-scatterplot-a-single-cell-case-study/">How to add boxplots or density plots side-by-side a scatterplot: a single cell case study</a></li>
        
        <li><a href="/post/how-to-construct-a-spatial-object-in-seurat/">How to construct a spatial object in Seurat</a></li>
        
        <li><a href="/post/use-random-forest-and-boost-trees-to-find-marker-genes-in-scrnaseq-data/">use random forest and boost trees to find marker genes in scRNAseq data</a></li>
        
        <li><a href="/post/partial-least-square-regression-for-marker-gene-identification-in-scrnaseq-data/">Partial least square regression for marker gene identification in scRNAseq data</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/navigating-variant-calling-for-disease-causing-mutations-the-state-of-art-process/" rel="next">Navigating Variant Calling for Disease-Causing Mutations: The state-of-art process</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/ten-single-cell-data-benchmarking-papers/" rel="prev">10 single-cell data benchmarking papers</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

