<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="In my last post, I tried to include transgenes to the cellranger reference and want to get the counts for the transgenes. However, even after I extended the Tdtomato and Cre with the potential 3’UTR, I still get very few cells express them. This is confusing to me.
My next thought is: maybe the STAR aligner is doing something weird that excluded those reads? At this point, I want to give kb-python, a python wrapper on kallisto and bustools a try.">

  
  <link rel="alternate" hreflang="en-us" href="/post/compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data/">
  <meta property="og:title" content="compare kallisto-bustools and cellranger for single nuclei sequencing data | DNA confesses Data speak">
  <meta property="og:description" content="In my last post, I tried to include transgenes to the cellranger reference and want to get the counts for the transgenes. However, even after I extended the Tdtomato and Cre with the potential 3’UTR, I still get very few cells express them. This is confusing to me.
My next thought is: maybe the STAR aligner is doing something weird that excluded those reads? At this point, I want to give kb-python, a python wrapper on kallisto and bustools a try.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2020-02-21T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2020-02-21T00:00:00&#43;00:00">
  

  
  

  <title>compare kallisto-bustools and cellranger for single nuclei sequencing data | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">compare kallisto-bustools and cellranger for single nuclei sequencing data</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2020-02-21 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Feb 21, 2020
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    11 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/scrnaseq/">scRNAseq</a
    >, 
    
    <a href="/categories/r/">R</a
    >, 
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=compare%20kallisto-bustools%20and%20cellranger%20for%20single%20nuclei%20sequencing%20data&amp;url=%2fpost%2fcompare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fcompare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fcompare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data%2f&amp;title=compare%20kallisto-bustools%20and%20cellranger%20for%20single%20nuclei%20sequencing%20data"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fcompare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data%2f&amp;title=compare%20kallisto-bustools%20and%20cellranger%20for%20single%20nuclei%20sequencing%20data"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=compare%20kallisto-bustools%20and%20cellranger%20for%20single%20nuclei%20sequencing%20data&amp;body=%2fpost%2fcompare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <p>In my last <a href="https://divingintogeneticsandgenomics.rbind.io/post/cellranger-mk-reference-with-transgenes/">post</a>, I tried to include transgenes to the cellranger reference and want to
get the counts for the transgenes. However, even after I extended the <code>Tdtomato</code> and <code>Cre</code> with
the potential 3’UTR, I still get very few cells express them. This is confusing to me.</p>
<p>My next thought is: maybe the STAR aligner is doing something weird that excluded those reads?
At this point, I want to give <a href="https://github.com/pachterlab/kb_python"><code>kb-python</code></a>, a python wrapper on <code>kallisto</code> and <code>bustools</code> a try.</p>
<p>Before <code>kb-python</code>, the workflow for processing single-nuclei data using <code>kallisto</code> and <code>bustools</code> is cumbersome. see this <a href="https://github.com/BUStools/getting_started/blob/master/kallisto_bus_mouse_nuclei_tutorial.ipynb">tutorial</a> and <a href="https://www.kallistobus.tools/velocity_index_tutorial.html">Building a cDNA and intron index</a>. I was unwilling to try it out until <code>kb-python</code> supports single-nuclei data as well. <code>kb-python</code> automates all the steps and greatly simplify the processing.</p>
<p>kb-python uses the gtf file and genome fasta file for indexing, and it will create the cDNA and intron fasta and the transcript to gene mapping file on the fly.</p>
<p>It requires the entries with exons should also have a corresponding entry with transcript in the third column of the gtf file.</p>
<p>Just duplicated the rows below and concatenate with the <code>genes.gtf</code> from the cellranger website.</p>
<pre class="bash"><code>tdtomato        custom  transcript      1       1880    .       +       .       gene_id &quot;ENSMUSGtdtomato&quot;; gene_version &quot;1&quot;; transcript_id &quot;tdtomato1&quot;; gene_name &quot;Tdtomato&quot;
tdtomato        custom  exon    1       1880    .       +       .       gene_id &quot;ENSMUSGtdtomato&quot;; gene_version &quot;1&quot;; transcript_id &quot;tdtomato1&quot;; gene_name &quot;Tdtomato&quot;
cre     custom  transcript      1       1067    .       +       .       gene_id &quot;ENSMUSGcre&quot;; gene_version &quot;1&quot;; transcript_id &quot;cre1&quot;; gene_name &quot;Cre&quot;
cre     custom  exon    1       1067    .       +       .       gene_id &quot;ENSMUSGcre&quot;; gene_version &quot;1&quot;; transcript_id &quot;cre1&quot;; gene_name &quot;Cre&quot;</code></pre>
<p>The developers of <code>kb-python</code> included a <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_single_nucleus.ipynb">tutorial</a> for pre-processing single-nuclei data.</p>
<p>Following it and make a reference:</p>
<pre class="bash"><code>## install kbpython
conda create -n kbpython pip
conda activate kbpython
pip install git+https://github.com/pachterlab/kb_python@count-kite

# make a reference
# you can specify -n 8 to split the index to 8 files to reduce the memory usage.
time kb ref -i index.idx -g t2g.txt -f1 cdna.fa -f2 intron.fa -c1 cdna_t2c.txt -c2 intron_t2c.txt --workflow nucleus -n 1 genome.fa genes.gtf &gt; log.txt  2&gt;&amp;1

real    266m53.243s
user    229m5.505s
sys     37m7.056s

## count
ref_dir=&quot;/reference_genome_by_tommy/kallisto_bus_ref/mm10_nuclei_single&quot;

fastq_dir=&quot;novaseq/outs/fastq_path/HJF3WDMXX/Sample1&quot;

time kb count -i ${ref_dir}/index.idx \
-g ${ref_dir}/t2g.txt -c1 ${ref_dir}/cdna_t2c.txt -c2 ${ref_dir}/intron_t2c.txt -x 10xv2 -o sample1_kb_h5ad -t 15 --workflow nucleus --h5ad \
${fastq_dir}/Sample1_S1_L001_R1_001.fastq.gz \
${fastq_dir}/Sample1_S1_L001_R2_001.fastq.gz \
${fastq_dir}/Sample1_S1_L002_R1_001.fastq.gz \
${fastq_dir}/Sample1_S1_L002_R2_001.fastq.gz

real    101m39.986s
user    899m28.925s
sys     114m7.979s
</code></pre>
<p>Inside the <code>sample1_kb_h5ad</code> output folder, there is a <code>counts_unfiltered</code> folder which contains the files we are going to work with.</p>
<pre class="bash"><code>ls counts_unfiltered/
adata.h5ad  spliced.barcodes.txt  spliced.genes.txt  spliced.mtx  unspliced.barcodes.txt  unspliced.genes.txt  unspliced.mtx
</code></pre>
<p>There are two matrices, spliced and unspliced. We need to sum up them together to get the final counts. The <code>adata.h5ad</code> is H5AD file contains the summed up matrix.</p>
<p>Some tips after playing with <code>kb-python</code> for a bit:</p>
<ul>
<li><p>Specify the wrong protocol will give you errors. <strong>If you specify <code>10xv2</code>, go and check the raw fastq and make sure it is 16 bp cell barcode + 10 bp UMI. If you specify <code>10xv3</code>, make sure it is 16 bp cell barcode + 12 bp UMI.</strong> Ideally, <code>kb-python</code> should check the input.</p></li>
<li><p>Sometimes, if you specify <code>--h5ad</code>, when combining the two spliced and unspliced sparse matrix, it gives error: &quot;in _get_arrayXarray csr_sample_values(M, N, self.indptr, self.indices, self.data, ValueError: could not convert integer scalar&quot;</p></li>
<li><p>If you specify whitelist by <code>-w</code>, use the unzipped txt file. Otherwise, you may get “died with &lt;Signals.SIGSEGV: 11&gt;” error.</p></li>
<li><p><code>kb-python</code> is strict with your gtf file. You may get an error when making references. I had some non-model gff3 file downloaded from NCBI and then converted to gtf using <code>gffread</code>, but <code>kb-python</code> complains about it.</p></li>
</ul>
<div id="downstream-analysis-in-r" class="section level3">
<h3>Downstream analysis in R</h3>
<p>Now, let’s import the data into R.</p>
<pre class="r"><code>library(Seurat)

Sample1&lt;- ReadH5AD(&quot;~/github_repos/blogdown_data/counts_unfiltered/adata.h5ad&quot;)</code></pre>
<p>I got this error:
“Pulling expression matrices and metadata
Data is unscaled
Error in file[[&quot;obs&quot;]][] :
object of type ‘environment’ is not subsettable”</p>
<p>I have to work with the <code>.mtx</code> files.</p>
<pre class="r"><code>library(Matrix)
library(tidyverse)

# a function to read in the kallisto count matrix
read_kallisto_sparse&lt;- function(cells, regions, mtx){
  mtx&lt;- Matrix::readMM(mtx)
  # the sparse matrix with rows are cells and columns are peaks/features
  mtx&lt;- t(mtx)
  regions&lt;- read_tsv(regions, col_names = FALSE)
  cells&lt;- read_tsv(cells, col_names = FALSE)
  rownames(mtx)&lt;- regions$X1
  # cellranger add -1 to the cell barcode, I add it for later compare with cellranger output
  colnames(mtx)&lt;- paste0(cells$X1, &quot;-1&quot;)
  return(mtx)
}

spliced&lt;- read_kallisto_sparse(cells = &quot;~/github_repos/blogdown_data/counts_unfiltered/spliced.barcodes.txt&quot;,
                               regions = &quot;~/github_repos/blogdown_data/counts_unfiltered/spliced.genes.txt&quot;,
                               mtx = &quot;~/github_repos/blogdown_data/counts_unfiltered/spliced.mtx&quot;)

unspliced&lt;- read_kallisto_sparse(cells = &quot;~/github_repos/blogdown_data/counts_unfiltered/unspliced.barcodes.txt&quot;,
                               regions = &quot;~/github_repos/blogdown_data/counts_unfiltered/unspliced.genes.txt&quot;,
                               mtx = &quot;~/github_repos/blogdown_data/counts_unfiltered/unspliced.mtx&quot;)

## common index
common_cells&lt;- intersect(colnames(spliced), colnames(unspliced))
spliced&lt;- spliced[, colnames(spliced) %in% common_cells]
unspliced&lt;- unspliced[, colnames(unspliced) %in% common_cells]

# make sure the cells and genes are lined up
all.equal(colnames(spliced), colnames(unspliced))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>all.equal(rownames(spliced), rownames(unspliced))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>## add up the counts
Sample1_kb&lt;- spliced + unspliced

# the rowname and colnames are lost, put them back
rownames(Sample1_kb)&lt;- rownames(spliced)
colnames(Sample1_kb)&lt;- colnames(spliced)</code></pre>
<p>The matrices are unfiltered, we can filter out some cells using the knee-plot. There are several nice posts on how to
by the <a href="https://github.com/CGATOxford/UMI-tools"><code>UMI-tools</code></a> developers:</p>
<ul>
<li><p><a href="https://cgatoxford.wordpress.com/2017/05/18/estimating-the-number-of-true-cell-barcodes-in-single-cell-rna-seq/" class="uri">https://cgatoxford.wordpress.com/2017/05/18/estimating-the-number-of-true-cell-barcodes-in-single-cell-rna-seq/</a></p></li>
<li><p><a href="https://cgatoxford.wordpress.com/2017/05/23/estimating-the-number-of-true-cell-barcodes-in-single-cell-rna-seq-part-2/" class="uri">https://cgatoxford.wordpress.com/2017/05/23/estimating-the-number-of-true-cell-barcodes-in-single-cell-rna-seq-part-2/</a></p></li>
</ul>
<pre class="r"><code>library(DropletUtils)

tot_counts &lt;- Matrix::colSums(Sample1_kb)

## many of them have very low counts per cell
summary(tot_counts)</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##     0.0     2.0     3.0   287.2    16.0 96681.0</code></pre>
<pre class="r"><code># Compute barcode rank from Dropletutils
bc_rank &lt;- barcodeRanks(Sample1_kb)

qplot(bc_rank$total, bc_rank$rank, geom = &quot;line&quot;) +
  geom_vline(xintercept = bc_rank$knee, color = &quot;blue&quot;, linetype = 2) +
  geom_vline(xintercept = bc_rank$inflection, color = &quot;green&quot;, linetype = 2) +
  annotate(&quot;text&quot;, y = 1000, x = 1.5 * c(bc_rank$knee, bc_rank$inflection),
           label = c(&quot;knee&quot;, &quot;inflection&quot;), color = c(&quot;blue&quot;, &quot;green&quot;)) +
  scale_x_log10() +
  scale_y_log10() +
  labs(y = &quot;Barcode rank&quot;, x = &quot;Total UMI count&quot;)</code></pre>
<p><img src="/post/2020-02-21-compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<pre class="r"><code># Filter the matrix using this cutoff
Sample1_kb &lt;- Sample1_kb[, tot_counts &gt; bc_rank$inflection]

## 73676 cells are left
dim(Sample1_kb)</code></pre>
<pre><code>## [1] 28694 73676</code></pre>
<p>This is way more than the cells we have in this experiment. As I will show later, cellranger gives ~10,000 cells which is about the right number of cells we have.</p>
</div>
<div id="kallisto-bustools-and-cellranger-correlation" class="section level3">
<h3>kallisto + bustools and cellranger correlation</h3>
<p>cellranger output</p>
<pre class="r"><code>library(Seurat)
# this is the cellranger output. read in the sparse matrix
Sample1_cr&lt;- Read10X_h5(filename = &quot;~/github_repos/blogdown_data/filtered_feature_bc_matrix.h5&quot;)</code></pre>
<pre class="r"><code>## cells pass cellranger and Seurat filter
colnames(Sample1_cr) %&gt;% 
  length()</code></pre>
<pre><code>## [1] 10937</code></pre>
<pre class="r"><code>## how many cells from the kb-python are in the cellranger output
(colnames(Sample1_kb) %in% colnames(Sample1_cr)) %&gt;% table()</code></pre>
<pre><code>## .
## FALSE  TRUE 
## 62739 10937</code></pre>
<p>All the cells in <code>kb-python</code> output are in <code>cellranger</code> output.</p>
<p>subset the <code>kb-python</code> matrix and rearrange the rows and columns to match each other.</p>
<pre class="r"><code># kb-python uses the ENSMBEL id
rownames(Sample1_kb) %&gt;% head()</code></pre>
<pre><code>## [1] &quot;ENSMUSG00000026535.9&quot;  &quot;ENSMUSG00000026315.13&quot; &quot;ENSMUSG00000000817.10&quot;
## [4] &quot;ENSMUSG00000063558.4&quot;  &quot;ENSMUSG00000001138.13&quot; &quot;ENSMUSG00000001143.13&quot;</code></pre>
<pre class="r"><code># cellranger uses the gene symbol
rownames(Sample1_cr) %&gt;% head()</code></pre>
<pre><code>## [1] &quot;Xkr4&quot;    &quot;Gm1992&quot;  &quot;Gm37381&quot; &quot;Rp1&quot;     &quot;Rp1.1&quot;   &quot;Sox17&quot;</code></pre>
<pre class="r"><code>## read in the transcript to gene map file t2g.txt was created when making kb-python index.
t2g&lt;- read_tsv(&quot;~/github_repos/blogdown_data/t2g.txt&quot;, col_names = FALSE, col_types = cols(.default = col_character()))

head(t2g)</code></pre>
<pre><code>## # A tibble: 6 x 8
##   X1             X2             X3     X4       X5    X6      X7      X8   
##   &lt;chr&gt;          &lt;chr&gt;          &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;
## 1 ENSMUST000000… ENSMUSG000000… Ifi20… Ifi202b… 1     173962… 173982… -    
## 2 ENSMUST000000… ENSMUSG000000… Serpi… Serpinb… 1     107590… 107608… +    
## 3 ENSMUST000000… ENSMUSG000000… Fasl   Fasl-001 1     161780… 161788… -    
## 4 ENSMUST000000… ENSMUSG000000… Aox1   Aox1-001 1     580299… 581064… +    
## 5 ENSMUST000000… ENSMUSG000000… Cnnm3  Cnnm3-0… 1     365118… 365282… +    
## 6 ENSMUST000000… ENSMUSG000000… Lman2l Lman2l-… 1     364231… 364452… -</code></pre>
<pre class="r"><code>ensemble2symbol&lt;- t2g %&gt;% 
  dplyr::select(X2,X3) %&gt;% distinct()

head(ensemble2symbol)</code></pre>
<pre><code>## # A tibble: 6 x 2
##   X2                    X3      
##   &lt;chr&gt;                 &lt;chr&gt;   
## 1 ENSMUSG00000026535.9  Ifi202b 
## 2 ENSMUSG00000026315.13 Serpinb8
## 3 ENSMUSG00000000817.10 Fasl    
## 4 ENSMUSG00000063558.4  Aox1    
## 5 ENSMUSG00000001138.13 Cnnm3   
## 6 ENSMUSG00000001143.13 Lman2l</code></pre>
<pre class="r"><code># not all genes in cellranger matrix are in this mapping file...
table(rownames(Sample1_cr) %in% ensemble2symbol$X3)</code></pre>
<pre><code>## 
## FALSE  TRUE 
##    67 28627</code></pre>
<pre class="r"><code>## what are the genes?
problematic_genes&lt;- rownames(Sample1_cr)[!(rownames(Sample1_cr) %in% ensemble2symbol$X3)]
problematic_genes</code></pre>
<pre><code>##  [1] &quot;Rp1.1&quot;           &quot;Gm15853.1&quot;       &quot;Gm16701.1&quot;      
##  [4] &quot;Olfr1284.1&quot;      &quot;Olfr1309.1&quot;      &quot;Olfr1316.1&quot;     
##  [7] &quot;Gm2464.1&quot;        &quot;Schip1.1&quot;        &quot;Flg.1&quot;          
## [10] &quot;Flg.2&quot;           &quot;Flg.3&quot;           &quot;Flg.4&quot;          
## [13] &quot;Flg.5&quot;           &quot;Flg.6&quot;           &quot;Hist2h2bb.1&quot;    
## [16] &quot;Smim20.1&quot;        &quot;Dancr.1&quot;         &quot;Gbp6.1&quot;         
## [19] &quot;D130017N08Rik.1&quot; &quot;Umad1.1&quot;         &quot;Ccdc142.1&quot;      
## [22] &quot;Zfand4.1&quot;        &quot;C1s2.1&quot;          &quot;Atn1.1&quot;         
## [25] &quot;Pik3c2g.1&quot;       &quot;Nova2.1&quot;         &quot;Apoc2.1&quot;        
## [28] &quot;Ltbp4.1&quot;         &quot;U2af1l4.1&quot;       &quot;Tead2.1&quot;        
## [31] &quot;Cd37.1&quot;          &quot;Tulp2.1&quot;         &quot;Tulp2.2&quot;        
## [34] &quot;Ntn5.1&quot;          &quot;Ntn5.2&quot;          &quot;Syngr4.1&quot;       
## [37] &quot;l7Rn6.1&quot;         &quot;Itgam.1&quot;         &quot;Tgfb1i1.1&quot;      
## [40] &quot;Olfr790.1&quot;       &quot;Olfr809.1&quot;       &quot;Map2k7.1&quot;       
## [43] &quot;Olfr730.1&quot;       &quot;Fbxw14.1&quot;        &quot;Olfr1396.1&quot;     
## [46] &quot;Olfr1366.1&quot;      &quot;3110039M20Rik.1&quot; &quot;Ighv5-8.1&quot;      
## [49] &quot;Ighv1-13.1&quot;      &quot;4930556M19Rik.1&quot; &quot;Sgsm3.1&quot;        
## [52] &quot;Olfr170.1&quot;       &quot;Olfr108.1&quot;       &quot;Olfr126.1&quot;      
## [55] &quot;Pcdha11.1&quot;       &quot;Pcdhga8.1&quot;       &quot;Olfr1496.1&quot;     
## [58] &quot;Fam205a2.1&quot;      &quot;Ccl21b.1&quot;        &quot;Il11ra2.1&quot;      
## [61] &quot;Ccl27a.1&quot;        &quot;Ccl21c.1&quot;        &quot;Gm3286.1&quot;       
## [64] &quot;Ccl27a.2&quot;        &quot;Il11ra2.2&quot;       &quot;Ccl19.1&quot;        
## [67] &quot;Ccl21a.1&quot;</code></pre>
<pre class="r"><code>## how about we remove the .1 and .2
problematic_genes %&gt;% str_replace(&quot;\\.[1-9]$&quot;, &quot;&quot;) %in% ensemble2symbol$X3 %&gt;% table()</code></pre>
<pre><code>## .
## TRUE 
##   67</code></pre>
<p>They all are in the <code>ensemble2symbol</code> file now if remove the version number.</p>
<pre class="r"><code># there are other gene symbols ends with .1 and .2... but has a corresponding name in ensemble2symbol...
# I can not use str_replace(&quot;\\.1$&quot;, &quot;&quot;)
# rownames(Sample1_cr) [rownames(Sample1_cr) %&gt;% str_detect(&quot;\\.[0-9]$&quot;) ] 

# find the index and replace with the symbols without version number
problematic_indx&lt;- which(rownames(Sample1_cr) %in% problematic_genes)

rownames(Sample1_cr)[problematic_indx]&lt;- problematic_genes %&gt;% 
  str_replace(&quot;\\.[1-9]$&quot;, &quot;&quot;)

rownames(Sample1_kb) %in% ensemble2symbol$X2 %&gt;% table()</code></pre>
<pre><code>## .
##  TRUE 
## 28694</code></pre>
<pre class="r"><code>rownames(Sample1_cr) %in% ensemble2symbol$X3 %&gt;% table()</code></pre>
<pre><code>## .
##  TRUE 
## 28694</code></pre>
<pre class="r"><code># a dictionary like vector with names are ensemble id and values are gene symbol
gene_map&lt;- ensemble2symbol %&gt;% tibble::deframe() 

## change the ensembel id with gene symbol
rownames(Sample1_kb)&lt;- gene_map[rownames(Sample1_kb)] %&gt;% unname()

#rearrange the columns and rows
Sample1_kb&lt;- Sample1_kb[rownames(Sample1_cr),colnames(Sample1_cr)]

## final check the rows and columns are lined up
all.equal(colnames(Sample1_kb), colnames(Sample1_cr))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>all.equal(rownames(Sample1_kb), rownames(Sample1_cr))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<p>Now, let me check the transgene expression in both pipelines.</p>
<pre class="r"><code># from kb-python
table(Sample1_kb[&quot;Cre&quot;, ] !=0)</code></pre>
<pre><code>## 
## FALSE  TRUE 
## 10728   209</code></pre>
<pre class="r"><code>table(Sample1_kb[&quot;Tdtomato&quot;, ]!=0)</code></pre>
<pre><code>## 
## FALSE  TRUE 
## 10715   222</code></pre>
<pre class="r"><code># from cellranger
table(Sample1_cr[&quot;Cre&quot;, ] !=0)</code></pre>
<pre><code>## 
## FALSE  TRUE 
## 10742   195</code></pre>
<pre class="r"><code>table(Sample1_cr[&quot;Tdtomato&quot;, ]!=0)</code></pre>
<pre><code>## 
## FALSE  TRUE 
## 10783   154</code></pre>
<p><code>kb-python</code> detected more cells express the transgens, but still the number is very low. I will
need to keep investigating the reason.</p>
</div>
<div id="calculate-the-correlation-between-the-two-pipelines" class="section level3">
<h3>Calculate the correlation between the two pipelines</h3>
<p><code>cor(matrixA, matrixB)</code> calculates the pair-wise correlation, one can use diag() to extract the
column correlations, but for big data matrix, it is not efficient.</p>
<p>googled and found <a href="https://stackoverflow.com/questions/6713973/how-do-i-calculate-correlation-between-corresponding-columns-of-two-matrices-and" class="uri">https://stackoverflow.com/questions/6713973/how-do-i-calculate-correlation-between-corresponding-columns-of-two-matrices-and</a></p>
<pre class="r"><code># from arrayMagic Bioconductor package
colCors = function(x, y) { 
   sqr = function(x) x*x
   if(!is.matrix(x)||!is.matrix(y)||any(dim(x)!=dim(y)))
     stop(&quot;Please supply two matrices of equal size.&quot;)
   x   = sweep(x, 2, colMeans(x))
   y   = sweep(y, 2, colMeans(y))
   cor = colSums(x*y) /  sqrt(colSums(sqr(x))*colSums(sqr(y)))
   return(cor)
}

cors&lt;- colCors(as.matrix(Sample1_cr), as.matrix(Sample1_kb))
head(cors)</code></pre>
<pre><code>## AAACCCAAGCAAGTGC-1 AAACCCAAGCGTGAGT-1 AAACCCAAGCTAGAAT-1 
##          0.9667430          0.8787159          0.9046970 
## AAACCCAAGCTCGTGC-1 AAACCCAAGCTGAAGC-1 AAACCCAAGCTGTCCG-1 
##          0.7184020          0.9463552          0.9077403</code></pre>
<p>I will make figures similar to the ones in the original <code>kallisto</code> + <code>bustool</code> paper <a href="https://www.biorxiv.org/content/10.1101/673285v1" class="uri">https://www.biorxiv.org/content/10.1101/673285v1</a></p>
<p>Fig2B plots a scatter plot of the total number of UMIs for each cell from the two pipelines.</p>
<pre class="r"><code># set up the theme for all figures
# take suggestions for theme https://twitter.com/ChenxinLi2/status/1228958667686338560
theme_set(theme_minimal() +
  theme(axis.line = element_line(colour = &quot;black&quot;, 
                      size = 1, linetype = &quot;solid&quot;),
        text = element_text(size = 18, color = &quot;black&quot;),  #face = &quot;bold&quot;
        axis.text.x = element_text(size = 18, color = &quot;black&quot;),
        axis.text.y = element_text(size = 18, color = &quot;black&quot;)))


Matrix::colSums(Sample1_cr) %&gt;%
  enframe(name = &quot;cell&quot;, value = &quot;CR_totalUMI&quot;) %&gt;% 
  bind_cols(KB_totalUMI= Matrix::colSums(Sample1_kb)) %&gt;% 
  ggplot(aes(x = KB_totalUMI, y = CR_totalUMI)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
  coord_equal() + 
  geom_abline(slope =1, intercept = 0, linetype=2, color = &quot;red&quot;) +
  ggtitle(&quot;total UMI per cell\nkb-python vs cellranger&quot;)</code></pre>
<p><img src="/post/2020-02-21-compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p><strong>kb-python always have more counts than cellranger for single-nuclei data</strong></p>
<p>Fig2C plots the correlation of counts from two pipelines for the same cell.</p>
<pre class="r"><code># install.packages(&quot;ggpointdensity&quot;)
library(ggpointdensity)
library(viridis)
df&lt;- Matrix::colSums(Sample1_kb) %&gt;%
  enframe(name = &quot;cell&quot;, value = &quot;KB_totalUMI&quot;) %&gt;%
  bind_cols(cors = cors) 


ggplot(df, aes(x= KB_totalUMI, y = cors)) +
  geom_pointdensity(adjust = .2) +
  scale_x_log10() + 
  scale_color_viridis() +
  ggtitle(&quot;pearson correlation per cell as a \nfunction of total UMI by kb-python&quot;)</code></pre>
<p><img src="/post/2020-02-21-compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p><strong>most of the correlations are higher than 0.8.</strong>, but not as good as for the single-cell data showed in the biorxiv paper.</p>
<pre class="r"><code>table(df$cors &lt; 0.8)</code></pre>
<pre><code>## 
## FALSE  TRUE 
##  9016  1921</code></pre>
<p>In addition, too many zeros can inflate the correlation.
Let’s remove the genes if the counts from two pipelines are both 0s.</p>
<pre class="r"><code># mapply(cor, as.data.frame(x), as.data.frame(y))

# map2 takes columns of df1 and df2 as argument and apply cor function to each pair of columns
cors2&lt;- map2_dbl( as.data.frame(Sample1_cr), as.data.frame(Sample1_kb), cor)
all.equal(cors, cors2)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>cor_remove_zero&lt;- function(x,y){
  indx&lt;- (x==0 &amp; y==0)
  return(cor(x[!indx],y[!indx]))
}

cors_remove_zeros&lt;- map2_dbl( as.data.frame(Sample1_cr), as.data.frame(Sample1_kb), cor_remove_zero)

df2&lt;- Matrix::colSums(Sample1_kb) %&gt;%
  enframe(name = &quot;cell&quot;, value = &quot;KB_totalUMI&quot;) %&gt;%
  bind_cols(cors_remove_zeros = cors_remove_zeros) 

# boxplot for correlations before and after removing 0s 
inner_join(df, df2) %&gt;%
  gather(3:4, key = &quot;group&quot;, value = &quot;correlation&quot;) %&gt;% 
  ggplot(aes(x=group, y = correlation)) +
  geom_boxplot() +
  xlab(&quot;&quot;)</code></pre>
<p><img src="/post/2020-02-21-compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>We see removing zeros decreases the correlation a bit.</p>
<p>Let’s plot side by side.</p>
<pre class="r"><code>p1&lt;- ggplot(df, aes(x= KB_totalUMI, y = cors)) +
  geom_pointdensity(adjust = .2) +
  scale_x_log10() + 
  scale_color_viridis() +
  ggtitle(&quot;correlation&quot;)


p2&lt;- ggplot(df2, aes(x= KB_totalUMI, y = cors)) +
  geom_pointdensity(adjust = .2) +
  scale_x_log10() + 
  scale_color_viridis() +
  ggtitle(&quot;correlation removing 0s&quot;)

#install.packages(&quot;patchwork&quot;)
# use patchwork to combine the legends from multiple plots
library(patchwork)
p1 / p2 + plot_layout(guides = &quot;collect&quot;)</code></pre>
<p><img src="/post/2020-02-21-compare-kallisto-bustools-and-cellranger-for-single-nuclei-seqencing-data_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<p>I have not seen many posts comparing cellranger with <code>kallisto + bustools</code> for single nuclei data. I hope this post opens the discussion for the single-cell RNAseq community.</p>
<ul>
<li><code>kallisto + bustools</code> always gives more counts for single-nuclei data, why is that?</li>
<li>Why the correlation between cellranger and <code>kallisto + bustools</code> is not as good for single-nuclei data?</li>
<li>For those cells with bad correlation, what’s going on?</li>
</ul>
<p>Comment below!</p>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/scrnaseq/">scRNAseq</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/cellranger-mk-reference-with-transgenes/">cellranger mk reference with transgenes</a></li>
        
        <li><a href="/post/add-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column/">add pct_in for each cluster for scRNAseq result table using list column</a></li>
        
        <li><a href="/post/mixing-mouse-and-human-10x-single-cell-rnaseq-data/">Mixing mouse and human 10x single cell RNAseq data</a></li>
        
        <li><a href="/post/negative-bionomial-distribution-in-single-cell-rnaseq/">negative bionomial distribution in (single-cell) RNAseq </a></li>
        
        <li><a href="/post/run-rstudio-server-with-singularity-on-hpc/">Run Rstudio server with singularity on HPC</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/monty-hall-problem-a-peek-through-simulation/" rel="next">Monty Hall problem- a peek through simulation</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/cellranger-mk-reference-with-transgenes/" rel="prev">cellranger mk reference with transgenes</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

