<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="Install fastq-dl To easily download fastq from GEO or ENA, use fastq-dl
Assume you already have conda installed, do the following:
conda config --add channels conda-forge conda config --add channels bioconda conda create -n fastq_download -c conda-forge -c bioconda fastq-dl conda activate fastq_download  Tip: use mamba if conda is too slow for you. They are all big snakes!!
We will use bulk RNAseq data from this GEO accession ID: https://www.">

  
  <link rel="alternate" hreflang="en-us" href="/post/how-to-preprocess-geo-bulk-rnaseq-data-with-salmon/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Chatomics">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Chatomics">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/how-to-preprocess-geo-bulk-rnaseq-data-with-salmon/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="Chatomics">
  <meta property="og:url" content="/post/how-to-preprocess-geo-bulk-rnaseq-data-with-salmon/">
  <meta property="og:title" content="How to preprocess GEO bulk RNAseq data with salmon  | Chatomics">
  <meta property="og:description" content="Install fastq-dl To easily download fastq from GEO or ENA, use fastq-dl
Assume you already have conda installed, do the following:
conda config --add channels conda-forge conda config --add channels bioconda conda create -n fastq_download -c conda-forge -c bioconda fastq-dl conda activate fastq_download  Tip: use mamba if conda is too slow for you. They are all big snakes!!
We will use bulk RNAseq data from this GEO accession ID: https://www.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2024-03-13T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2024-03-13T00:00:00&#43;00:00">
  

  
  

  <title>How to preprocess GEO bulk RNAseq data with salmon  | Chatomics</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">Chatomics</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>Newsletter</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">How to preprocess GEO bulk RNAseq data with salmon </h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2024-03-13 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Mar 13, 2024
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    4 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/how-to-preprocess-geo-bulk-rnaseq-data-with-salmon/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >, 
    
    <a href="/categories/rnaseq/">RNAseq</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=How%20to%20preprocess%20GEO%20bulk%20RNAseq%20data%20with%20salmon%20&amp;url=%2fpost%2fhow-to-preprocess-geo-bulk-rnaseq-data-with-salmon%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fhow-to-preprocess-geo-bulk-rnaseq-data-with-salmon%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fhow-to-preprocess-geo-bulk-rnaseq-data-with-salmon%2f&amp;title=How%20to%20preprocess%20GEO%20bulk%20RNAseq%20data%20with%20salmon%20"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fhow-to-preprocess-geo-bulk-rnaseq-data-with-salmon%2f&amp;title=How%20to%20preprocess%20GEO%20bulk%20RNAseq%20data%20with%20salmon%20"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=How%20to%20preprocess%20GEO%20bulk%20RNAseq%20data%20with%20salmon%20&amp;body=%2fpost%2fhow-to-preprocess-geo-bulk-rnaseq-data-with-salmon%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<link href="/rmarkdown-libs/vembedr/css/vembedr.css" rel="stylesheet" />


<div id="install-fastq-dl" class="section level3">
<h3>Install fastq-dl</h3>
<p>To easily download fastq from GEO or ENA, use <a href="https://github.com/rpetit3/fastq-dl">fastq-dl</a></p>
<p>Assume you already have <a href="https://conda.io/projects/conda/en/latest/user-guide/install/index.html"><code>conda</code></a> installed, do the following:</p>
<pre class="bash"><code>conda config --add channels conda-forge
conda config --add channels bioconda
conda create -n fastq_download -c conda-forge -c bioconda fastq-dl
conda activate fastq_download </code></pre>
<p><strong>Tip</strong>: use <a href="https://github.com/mamba-org/mamba"><code>mamba</code></a> if <code>conda</code> is too slow for you.
They are all big snakes!!</p>
<p>We will use bulk RNAseq data from this GEO accession ID:
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576</a></p>
<pre class="bash"><code>mkdir hypoxia_RNAseq
cd hypoxia_RNAseq

mkdir data
cd data
## some times the same experiment may have multiple runs (SRR)
## use the --group-by-experiment to merge them, this example only has one SRR
## per sample   

## normoxia sgCTRL
fastq-dl --accession SRX14311105  --group-by-experiment
fastq-dl --accession SRX14311106  --group-by-experiment

## hypoxia sgCTRL
fastq-dl --accession SRX14311111  --group-by-experiment
fastq-dl --accession SRX14311112  --group-by-experiment</code></pre>
<p>We only have 4 samples here and we typed in the command one by one which is fine.
What if we want to process all of them and you do not want to do it manually.</p>
<p>Get the metadata using <a href="https://github.com/NCBI-Hackathons/ncbi-cloud-tutorials/blob/master/SRA%20tutorials/tutorial_SRA_run_selector.md">SRA Run selector</a> or use <a href="https://www.youtube.com/watch?v=NSoU58IqA3Y">GEOquery</a> or <a href="https://www.bioconductor.org/packages/release/bioc/html/GEOmetadb.html">GEOmetadb</a>.</p>
<p>You may want to do a <code>fastqc</code> for quality control of the reads and trimming with
<code>fastp</code> for the sequencing adapters. We will skip it in this tutorial.</p>
</div>
<div id="install-salmon" class="section level3">
<h3>install salmon</h3>
<pre class="bash"><code>conda create -n salmon salmon</code></pre>
</div>
<div id="build-an-index" class="section level3">
<h3>build an index</h3>
<p>We need a transcriptome fasta file to build the index, we also need the gtf file that we can
use to map gene ids.</p>
<p>Let’s go to gencode <a href="https://www.gencodegenes.org/human/" class="uri">https://www.gencodegenes.org/human/</a></p>
<p>There are multiple sources to get the reference files. If you are confused,
watch this video.</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/eIVlSG11umQ" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
<p>Download the reference fasta:</p>
<pre class="bash"><code>cd hypoxia_RNAseq
mkdir reference
cd reference

wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_45/gencode.v45.annotation.gtf.gz


wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_45/gencode.v45.basic.annotation.gtf.gz</code></pre>
<p>A quick look at the gtf file:</p>
<pre class="bash"><code>zless -S gencode.v45.annotation.gtf.gz | grep -v &quot;#&quot; | \
awk &#39;$3==&quot;gene&quot;&#39; | cut -f9 | head -3

gene_id &quot;ENSG00000290825.1&quot;; gene_type &quot;lncRNA&quot;; gene_name &quot;DDX11L2&quot;; level 2; tag &quot;overlaps_pseudogene&quot;;
gene_id &quot;ENSG00000223972.6&quot;; gene_type &quot;transcribed_unprocessed_pseudogene&quot;; gene_name &quot;DDX11L1&quot;; level 2; hgnc_id &quot;HGNC:37102&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;;
gene_id &quot;ENSG00000227232.6&quot;; gene_type &quot;unprocessed_pseudogene&quot;; gene_name &quot;WASH7P&quot;; level 2; hgnc_id &quot;HGNC:38034&quot;; havana_gene &quot;OTTHUMG00000000958.1&quot;;</code></pre>
<p>Let’s tally the gene_type</p>
<pre class="bash"><code>zless -S gencode.v45.annotation.gtf.gz | grep -v &quot;#&quot; | \
awk &#39;$3==&quot;gene&quot;&#39; | cut -f9 | \
cut -f2 -d &quot;;&quot; | sort | uniq -c | sort -k1,1nr

  20049  gene_type &quot;protein_coding&quot;
  19370  gene_type &quot;lncRNA&quot;
  10143  gene_type &quot;processed_pseudogene&quot;
   2604  gene_type &quot;unprocessed_pseudogene&quot;
   2208  gene_type &quot;misc_RNA&quot;
   1901  gene_type &quot;snRNA&quot;
   1879  gene_type &quot;miRNA&quot;
   1054  gene_type &quot;TEC&quot;
    962  gene_type &quot;transcribed_unprocessed_pseudogene&quot;
    942  gene_type &quot;snoRNA&quot;
    513  gene_type &quot;transcribed_processed_pseudogene&quot;
    497  gene_type &quot;rRNA_pseudogene&quot;
    187  gene_type &quot;IG_V_pseudogene&quot;
    158  gene_type &quot;transcribed_unitary_pseudogene&quot;
    145  gene_type &quot;IG_V_gene&quot;
    107  gene_type &quot;TR_V_gene&quot;
    100  gene_type &quot;unitary_pseudogene&quot;
     79  gene_type &quot;TR_J_gene&quot;
     49  gene_type &quot;scaRNA&quot;
     47  gene_type &quot;rRNA&quot;
     37  gene_type &quot;IG_D_gene&quot;
     33  gene_type &quot;TR_V_pseudogene&quot;
     22  gene_type &quot;Mt_tRNA&quot;
     19  gene_type &quot;artifact&quot;
     18  gene_type &quot;IG_J_gene&quot;
     14  gene_type &quot;IG_C_gene&quot;
      9  gene_type &quot;IG_C_pseudogene&quot;
      8  gene_type &quot;ribozyme&quot;
      6  gene_type &quot;TR_C_gene&quot;
      5  gene_type &quot;TR_D_gene&quot;
      5  gene_type &quot;sRNA&quot;
      4  gene_type &quot;TR_J_pseudogene&quot;
      4  gene_type &quot;vault_RNA&quot;
      3  gene_type &quot;IG_J_pseudogene&quot;
      2  gene_type &quot;Mt_rRNA&quot;
      2  gene_type &quot;translated_processed_pseudogene&quot;
      1  gene_type &quot;IG_pseudogene&quot;
      1  gene_type &quot;scRNA&quot;</code></pre>
<p>We will use the <code>gtf</code> file for downstream analysis.</p>
<p>Download the transcript fasta file</p>
<pre class="bash"><code>wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_45/gencode.v45.transcripts.fa.gz
</code></pre>
<p>index the transcriptome</p>
<pre class="bash"><code># list all the conda env
conda env list

# activate the salmon env
conda activate salmon

salmon index -t gencode.v45.transcripts.fa.gz -i gencode.v45_human_index -k 31 --gencode</code></pre>
</div>
<div id="quantification" class="section level3">
<h3>quantification</h3>
<p>This is a single end read dataset, so I used <code>-r</code> to feed in the fastq. If you
have paired-end reads, you need to use <code>-1</code> and <code>-2</code> for the <code>R1</code> and <code>R2</code> reads.</p>
<p>There are also different library types for paired end reads. read <a href="https://salmon.readthedocs.io/en/latest/library_type.html">the documentation</a> for details.</p>
<p><img src="/img/salmon_lib.png" /></p>
<p>Make sure you know how the library is constructed. I do not know if this library
is stranded or unstranded, so I used <code>-l A</code> to automatically let
salmon to detect the library type. salmon detected it as <code>U</code> unstranded.</p>
<pre class="bash"><code>cd ..

salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311105.fastq.gz -o SRX14311105_quant  --validateMappings</code></pre>
<p>This finished in a couple of mins, so fast!</p>
<pre class="bash"><code>cat SRX14311105_quant/logs/salmon_quant.log | grep &quot;Mapping rate&quot;
[jointLog] [info] Mapping rate = 93.0329%</code></pre>
<p>93% mapping rate!</p>
<p>Let’s do the same for the other 3 samples</p>
<pre class="bash"><code>salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311106.fastq.gz -o SRX14311106_quant  --validateMappings

salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311111.fastq.gz -o SRX14311111_quant  --validateMappings

salmon quant -i reference/gencode.v45_human_index/ -l A -r SRX14311112.fastq.gz -o SRX14311112_quant  --validateMappings</code></pre>
<p>Now we have the salmon quantification output:</p>
<pre class="bash"><code> find . -name &quot;*sf&quot;
./SRX14311111_quant/quant.sf
./SRX14311105_quant/quant.sf
./SRX14311106_quant/quant.sf
./SRX14311112_quant/quant.sf</code></pre>
<p>and the mapping rates are all pretty high:</p>
<pre class="bash"><code>find . -name &quot;salmon_quant.log&quot; | xargs grep &quot;Mapping rate&quot;
./SRX14311111_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 91.5163%
./SRX14311105_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 93.0329%
./SRX14311106_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 92.8808%
./SRX14311112_quant/logs/salmon_quant.log:[jointLog] [info] Mapping rate = 91.4357%</code></pre>
<p>we can load it into R and use <code>DESeq2</code>for downstream analysis. That’s for another post!
Happy Learning!</p>
<p>Watch the video here:</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/_Q8fYokTCTs" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/bioinformatics/">bioinformatics</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/rnaseq/">RNAseq</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/how-to-convert-raw-counts-to-tpm-for-tcga-data-and-make-a-heatmap-across-cancer-types/">How to convert raw counts to TPM for TCGA data and make a heatmap across cancer types</a></li>
        
        <li><a href="/post/do-you-really-understand-log2fold-change-in-single-cell-rnaseq-data/">Do you really understand log2Fold change in single-cell RNAseq data?</a></li>
        
        <li><a href="/post/hidden-skills-beyond-programming-for-computational-biology/">Hidden skills beyond programming for computational biology</a></li>
        
        <li><a href="/post/great-things-take-time-how-decades-of-effort-led-to-my-dream-career/">Great Things Take Time: How Decades of Effort Led to My Dream Career</a></li>
        
        <li><a href="/post/review-2023/">Review 2023</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/downstream-of-bulk-rnaseq-read-in-salmon-output-using-tximport-and-then-deseq2/" rel="next">Downstream of bulk RNAseq: read in salmon output using tximport and then DESeq2</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/do-you-really-understand-log2fold-change-in-single-cell-rnaseq-data/" rel="prev">Do you really understand log2Fold change in single-cell RNAseq data?</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

