<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="Googling is a required technique for programmers. Once I have a programming problem in mind, the first thing I do is to google to see if other people have encountered the same problem and maybe they already have a solution. Do not re-invent the wheels. Actually, reading other people’s code and mimicing their code is a great way of learning. Today, I am going to show you how to compute binned averages/sums along a genome or any genomic regions of interest.">

  
  <link rel="alternate" hreflang="en-us" href="/post/compute-averages-sums-on-granges-or-equal-length-bins/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/compute-averages-sums-on-granges-or-equal-length-bins/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/compute-averages-sums-on-granges-or-equal-length-bins/">
  <meta property="og:title" content="Compute averages/sums on GRanges or equal length bins | DNA confesses Data speak">
  <meta property="og:description" content="Googling is a required technique for programmers. Once I have a programming problem in mind, the first thing I do is to google to see if other people have encountered the same problem and maybe they already have a solution. Do not re-invent the wheels. Actually, reading other people’s code and mimicing their code is a great way of learning. Today, I am going to show you how to compute binned averages/sums along a genome or any genomic regions of interest.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2018-09-07T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2018-09-07T00:00:00&#43;00:00">
  

  
  

  <title>Compute averages/sums on GRanges or equal length bins | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>Newsletter</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">Compute averages/sums on GRanges or equal length bins</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2018-09-07 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Sep 7, 2018
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    10 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/compute-averages-sums-on-granges-or-equal-length-bins/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/r/">R</a
    >, 
    
    <a href="/categories/genomics/">genomics</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Compute%20averages%2fsums%20on%20GRanges%20or%20equal%20length%20bins&amp;url=%2fpost%2fcompute-averages-sums-on-granges-or-equal-length-bins%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fcompute-averages-sums-on-granges-or-equal-length-bins%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fcompute-averages-sums-on-granges-or-equal-length-bins%2f&amp;title=Compute%20averages%2fsums%20on%20GRanges%20or%20equal%20length%20bins"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fcompute-averages-sums-on-granges-or-equal-length-bins%2f&amp;title=Compute%20averages%2fsums%20on%20GRanges%20or%20equal%20length%20bins"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Compute%20averages%2fsums%20on%20GRanges%20or%20equal%20length%20bins&amp;body=%2fpost%2fcompute-averages-sums-on-granges-or-equal-length-bins%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <p>Googling is a required technique for programmers. Once I have a programming problem in mind, the first thing I do is to google to see if other people have encountered the same problem and maybe they already have a solution. Do not re-invent the wheels. Actually, reading other people’s code and mimicing their code is a great way of learning. Today, I am going to show you how to compute binned averages/sums along a genome or any genomic regions of interest. All the codes I am going to show I found them online.</p>
<div id="how-to-compute-binned-averages-along-a-genome" class="section level4">
<h4>How to compute binned averages along a genome</h4>
<p>I found it in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesHOWTOs.pdf">How to tutorial</a> of the GRanges package.</p>
<pre class="r"><code># using yeast with smaller genome
library(BSgenome.Scerevisiae.UCSC.sacCer2)
library(GenomicRanges)
set.seed(55)
my_var &lt;- RleList(
        lapply(seqlengths(Scerevisiae),
        function(seqlen) {
        tmp &lt;- sample(50L, seqlen, replace=TRUE) %/% 50L
        Rle(cumsum(tmp - rev(tmp)))
        }
        ),
        compress=FALSE)

my_var</code></pre>
<pre><code>## RleList of length 18
## $chrI
## integer-Rle of length 230208 with 8693 runs
##   Lengths:  32  33  51  21  22   9   1  36 ...   1   9  22  21  51  33  33
##   Values :   0   1   2   1   0  -1   0   1 ...   0  -1   0   1   2   1   0
## 
## $chrII
## integer-Rle of length 813178 with 31959 runs
##   Lengths: 56 10 52 12 69  4 48 35 11  1 ...  1 11 35 48  4 69 12 52 10 57
##   Values :  0  1  0  1  0  1  2  1  2  3 ...  3  2  1  2  1  0  1  0  1  0
## 
## $chrIII
## integer-Rle of length 316617 with 12209 runs
##   Lengths:  20 116   9   2  21  16  43   3 ...  43  16  21   2   9 116  21
##   Values :   0  -1   0   1   0  -1  -2  -3 ...  -2  -1   0   1   0  -1   0
## 
## $chrIV
## integer-Rle of length 1531919 with 60091 runs
##   Lengths: 39 80 67 22 48 77 19 45 13  3 ...  3 13 45 19 77 48 22 67 80 40
##   Values :  0 -1  0  1  2  3  2  1  0  1 ...  1  0  1  2  3  2  1  0 -1  0
## 
## $chrV
## integer-Rle of length 576869 with 22903 runs
##   Lengths:  11  29   7   1  10  29  63  32 ...  63  29  10   1   7  29  12
##   Values :   0  -1  -2  -3  -4  -3  -4  -5 ...  -4  -3  -4  -3  -2  -1   0
## 
## ...
## &lt;13 more elements&gt;</code></pre>
<p>tile the whole genome to 100 bp bins</p>
<pre class="r"><code>bins &lt;- tileGenome(seqinfo(Scerevisiae), tilewidth=100,cut.last.tile.in.chrom=TRUE)</code></pre>
<p>compute the binned average for my_var</p>
<pre class="r"><code>binnedAverage(bins, my_var, &quot;binned_var&quot;)</code></pre>
<pre><code>## GRanges object with 121639 ranges and 1 metadata column:
##            seqnames       ranges strand | binned_var
##               &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt;
##        [1]     chrI   [  1, 100]      * |       1.03
##        [2]     chrI   [101, 200]      * |       0.75
##        [3]     chrI   [201, 300]      * |       0.92
##        [4]     chrI   [301, 400]      * |       2.75
##        [5]     chrI   [401, 500]      * |       6.06
##        ...      ...          ...    ... .        ...
##   [121635]  2micron [5901, 6000]      * |       4.76
##   [121636]  2micron [6001, 6100]      * |       2.62
##   [121637]  2micron [6101, 6200]      * |       0.87
##   [121638]  2micron [6201, 6300]      * |       0.03
##   [121639]  2micron [6301, 6318]      * |          0
##   -------
##   seqinfo: 18 sequences (2 circular) from sacCer2 genome</code></pre>
<p>The key is to convert any values (sequencing depth across the genome, RNA-seq counts etc) into a RleList object, then one can use the <code>binnedAverage</code> to calculate the average across each small bin of the genome.</p>
</div>
<div id="transformation-of-grange-object-to-density-per-bin" class="section level4">
<h4>Transformation of GRange object to density per bin</h4>
<p>see <a href="https://stat.ethz.ch/pipermail/bioconductor/2013-January/050445.html">post</a></p>
<pre class="r"><code>### &#39;x&#39;: a GenomicRanges objects with non-NA seqlengths.
### &#39;binsize&#39;: a single positive integer.
### &#39;mcolnames&#39;: names of numeric metadata columns in &#39;x&#39; to &quot;average&quot;
###              i.e. to propagate to the result after averaging them
###              on each bin.
### Returns a GRanges object with: (a) the same seqinfo as &#39;x&#39;,
### (b) ranges of width &#39;binsize&#39; covering all the sequences in
### &#39;seqinfo(x)&#39;, and (c) the &quot;averaged&quot; metadata columns specified
### in &#39;mcolnames&#39;.

averagePerBin &lt;- function(x, binsize, mcolnames=NULL)
{
     if (!is(x, &quot;GenomicRanges&quot;))
         stop(&quot;&#39;x&#39; must be a GenomicRanges object&quot;)
     if (any(is.na(seqlengths(x))))
         stop(&quot;&#39;seqlengths(x)&#39; contains NAs&quot;)
     bins &lt;- IRangesList(lapply(seqlengths(x),
                                function(seqlen)
                                  IRanges(breakInChunks(seqlen, binsize))))
     ans &lt;- as(bins, &quot;GRanges&quot;)
     seqinfo(ans) &lt;- seqinfo(x)
     if (is.null(mcolnames))
         return(ans)
     averageMCol &lt;- function(colname)
     {
         cvg &lt;- coverage(x, weight=colname)
         views_list &lt;- RleViewsList(
                           lapply(names(cvg),
                               function(seqname)
                                   Views(cvg[[seqname]], bins[[seqname]])))
         unlist(viewMeans(views_list), use.names=FALSE)
     }
     mcols(ans) &lt;- DataFrame(lapply(mcols(x)[mcolnames], averageMCol))
     ans
}</code></pre>
<pre class="r"><code>library(BSgenome.Hsapiens.UCSC.hg19)

testset.gr&lt;- GRanges(&quot;chr1&quot;, IRanges(start=seq( 50000, 55000,by = 100), width=50), strand = &quot;+&quot;)

## Set the sequence lengths.
seqlengths(testset.gr) &lt;- seqlengths(Hsapiens)[seqlevels(testset.gr)]

## Add the density metadata col.
mcols(testset.gr)$density &lt;- 100

## Compute the average per bin for the specified metadata cols.
avg_per_bin &lt;- averagePerBin(testset.gr, 100, mcolnames=&quot;density&quot;)

avg_per_bin[avg_per_bin$density &gt; 0]</code></pre>
<pre><code>## GRanges object with 52 ranges and 1 metadata column:
##        seqnames         ranges strand |   density
##           &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##    [1]     chr1 [49901, 50000]      * |         1
##    [2]     chr1 [50001, 50100]      * |        50
##    [3]     chr1 [50101, 50200]      * |        50
##    [4]     chr1 [50201, 50300]      * |        50
##    [5]     chr1 [50301, 50400]      * |        50
##    ...      ...            ...    ... .       ...
##   [48]     chr1 [54601, 54700]      * |        50
##   [49]     chr1 [54701, 54800]      * |        50
##   [50]     chr1 [54801, 54900]      * |        50
##   [51]     chr1 [54901, 55000]      * |        50
##   [52]     chr1 [55001, 55100]      * |        49
##   -------
##   seqinfo: 1 sequence from an unspecified genome</code></pre>
<p>Note that calling <code>averagePerBin()</code> without specifying ‘mcolnames’ is a convenient way to generate genomic bins covering the entire genome (and in that case the supplied GRanges doesn’t even need to contain ranges). similar to <code>tileGenome</code>.</p>
<pre class="r"><code>empty_gr &lt;- GRanges(seqinfo=seqinfo(Hsapiens))
empty_gr</code></pre>
<pre><code>## GRanges object with 0 ranges and 0 metadata columns:
##    seqnames    ranges strand
##       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<pre class="r"><code>averagePerBin(empty_gr, 25000000)</code></pre>
<pre><code>## GRanges object with 205 ranges and 0 metadata columns:
##               seqnames                 ranges strand
##                  &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt;
##     [1]           chr1 [        1,  25000000]      *
##     [2]           chr1 [ 25000001,  50000000]      *
##     [3]           chr1 [ 50000001,  75000000]      *
##     [4]           chr1 [ 75000001, 100000000]      *
##     [5]           chr1 [100000001, 125000000]      *
##     ...            ...                    ...    ...
##   [201] chrUn_gl000245             [1, 36651]      *
##   [202] chrUn_gl000246             [1, 38154]      *
##   [203] chrUn_gl000247             [1, 36422]      *
##   [204] chrUn_gl000248             [1, 39786]      *
##   [205] chrUn_gl000249             [1, 38502]      *
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="how-to-compute-averages-of-a-meta-column-from-one-granges-on-the-other-granges" class="section level4">
<h4>How to compute averages of a meta column from one GRanges on the other GRanges</h4>
<p>see a <a href="https://support.bioconductor.org/p/57956/">post</a></p>
<pre class="r"><code>size &lt;- 50000
windowSize &lt;- 10

dat &lt;- GRanges(&quot;chr1&quot;, IRanges(start=1:size, width=2), strand=&quot;+&quot;,score=sample(1:size, size))

# windows
GRwin &lt;- GRanges(&quot;chr1&quot;, IRanges(start=(0:(size/windowSize))*windowSize, width=windowSize), strand=&quot;+&quot;)

## make a RleList object from the data
score &lt;- coverage(dat, weight=&quot;score&quot;)</code></pre>
<p>Then to summarize ‘score’ on your fixed-size tiling windows, you need a summarizing function like the <code>binnedAverage()</code> function shown in <code>?tileGenome</code>. <code>binnedAverage()</code> computes the average on each window but it’s easy to write a summarizing function that computes the sum:</p>
<pre class="r"><code>binnedSum &lt;- function(bins, numvar, mcolname)
{
  stopifnot(is(bins, &quot;GRanges&quot;))
  stopifnot(is(numvar, &quot;RleList&quot;))
  stopifnot(identical(seqlevels(bins), names(numvar)))
  bins_per_chrom &lt;- split(ranges(bins), seqnames(bins))
  sums_list &lt;- lapply(names(numvar),
      function(seqname) {
          views &lt;- Views(numvar[[seqname]],
                         bins_per_chrom[[seqname]])
          viewSums(views)
      })
  new_mcol &lt;- unsplit(sums_list, as.factor(seqnames(bins)))
  mcols(bins)[[mcolname]] &lt;- new_mcol
  bins
}


GRwin2 &lt;- binnedSum(GRwin, score, &quot;binned_score&quot;)

GRwin2</code></pre>
<pre><code>## GRanges object with 5001 ranges and 1 metadata column:
##          seqnames         ranges strand | binned_score
##             &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;integer&gt;
##      [1]     chr1       [ 0,  9]      + |       304897
##      [2]     chr1       [10, 19]      + |       517317
##      [3]     chr1       [20, 29]      + |       377486
##      [4]     chr1       [30, 39]      + |       274838
##      [5]     chr1       [40, 49]      + |       513542
##      ...      ...            ...    ... .          ...
##   [4997]     chr1 [49960, 49969]      + |       515986
##   [4998]     chr1 [49970, 49979]      + |       521740
##   [4999]     chr1 [49980, 49989]      + |       424913
##   [5000]     chr1 [49990, 49999]      + |       514258
##   [5001]     chr1 [50000, 50009]      + |        11963
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="turning-a-granges-metadata-column-into-rlelist-object." class="section level4">
<h4>turning a GRanges metadata column into RleList object.</h4>
<p>see a <a href="https://support.bioconductor.org/p/50014/">post</a></p>
<pre class="r"><code>gr&lt;- GRanges(c(&quot;chr1&quot;, &quot;chr2&quot;), IRanges(c(10, 50), c(16, 55)), scores= c(20, 10))
seqlengths(gr) &lt;- c(100, 100)

coverage(gr, weight=gr$scores)</code></pre>
<pre><code>## RleList of length 2
## $chr1
## numeric-Rle of length 100 with 3 runs
##   Lengths:  9  7 84
##   Values :  0 20  0
## 
## $chr2
## numeric-Rle of length 100 with 3 runs
##   Lengths: 49  6 45
##   Values :  0 10  0</code></pre>
<p>Depending on your needs, the ranges which aren’t present in the GRanges object may effectively have missing scores and need to be NA, and 0 is a valid score for the ranges which are present. One hack would be to add 1 to all of the scores, replace the zeros in the <code>coverage()</code> result with NAs and subtract 1:</p>
<pre class="r"><code>gr$scores &lt;- gr$scores + 1L
cov &lt;- coverage(gr, weight  = &quot;scores&quot;)
cov[cov == 0L] &lt;- NA
cov &lt;- cov - 1L</code></pre>
<p>Alternatively you could call <code>coverage()</code> a 2nd time with no weights to find the regions with no coverage, and set them to NA:</p>
<pre class="r"><code>cvg &lt;- coverage(gr, weight=gr$scores)
cvg[coverage(gr) == 0] &lt;- NA</code></pre>
<p>It turns out that there are functions to convert between meta data column and RleList. Just be careful with the different behaviors of different functions.</p>
<pre class="r"><code># ?bindAsGRanges
# ?mcolAsRleList

mcolAsRleList(gr, varname = &quot;scores&quot;)</code></pre>
<pre><code>## RleList of length 2
## $chr1
## numeric-Rle of length 100 with 3 runs
##   Lengths:  9  7 84
##   Values : NA 21 NA
## 
## $chr2
## numeric-Rle of length 100 with 3 runs
##   Lengths: 49  6 45
##   Values : NA 11 NA</code></pre>
<pre class="r"><code>bindAsGRanges(cvg)</code></pre>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |        V1
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr1  [10, 16]      * |        21
##   [2]     chr2  [50, 55]      * |        11
##   -------
##   seqinfo: 2 sequences from an unspecified genome</code></pre>
<pre class="r"><code>bindAsGRanges(coverage(gr,weight=gr$scores))</code></pre>
<pre><code>## GRanges object with 6 ranges and 1 metadata column:
##       seqnames    ranges strand |        V1
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr1 [ 1,   9]      * |         0
##   [2]     chr1 [10,  16]      * |        21
##   [3]     chr1 [17, 100]      * |         0
##   [4]     chr2 [ 1,  49]      * |         0
##   [5]     chr2 [50,  55]      * |        11
##   [6]     chr2 [56, 100]      * |         0
##   -------
##   seqinfo: 2 sequences from an unspecified genome</code></pre>
<pre class="r"><code># or coerce using as
as(cvg, &quot;GRanges&quot;)</code></pre>
<pre><code>## GRanges object with 6 ranges and 1 metadata column:
##       seqnames    ranges strand |     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr1 [ 1,   9]      * |      &lt;NA&gt;
##   [2]     chr1 [10,  16]      * |        21
##   [3]     chr1 [17, 100]      * |      &lt;NA&gt;
##   [4]     chr2 [ 1,  49]      * |      &lt;NA&gt;
##   [5]     chr2 [50,  55]      * |        11
##   [6]     chr2 [56, 100]      * |      &lt;NA&gt;
##   -------
##   seqinfo: 2 sequences from an unspecified genome</code></pre>
<pre class="r"><code>as(coverage(gr, weight = gr$scores), &quot;GRanges&quot;)</code></pre>
<pre><code>## GRanges object with 6 ranges and 1 metadata column:
##       seqnames    ranges strand |     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr1 [ 1,   9]      * |         0
##   [2]     chr1 [10,  16]      * |        21
##   [3]     chr1 [17, 100]      * |         0
##   [4]     chr2 [ 1,  49]      * |         0
##   [5]     chr2 [50,  55]      * |        11
##   [6]     chr2 [56, 100]      * |         0
##   -------
##   seqinfo: 2 sequences from an unspecified genome</code></pre>
</div>
<div id="subset-rlelist-with-granges" class="section level4">
<h4>subset RleList with GRanges</h4>
<pre class="r"><code>cov[gr]</code></pre>
<pre><code>## RleList of length 2
## $chr1
## numeric-Rle of length 7 with 1 run
##   Lengths:  7
##   Values : 20
## 
## $chr2
## numeric-Rle of length 6 with 1 run
##   Lengths:  6
##   Values : 10</code></pre>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/granges/">GRanges</a>
  
</div>




    
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/set-up-my-new-mac-laptop/" rel="next">set up my new mac laptop</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/my-first-try-on-rmarkdown-using-blogdown/" rel="prev">my first try on Rmarkdown using blogdown</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

